// openclaw-sdk/dist/esm/utils/timeoutManager.js
/**
 * Timeout Manager
 *
 * Provides centralized timeout management with automatic cleanup,
 * promise-based delays, and prevents common timeout-related bugs.
 */
// ============================================================================
// Timeout Manager
// ============================================================================
/**
 * Centralized timeout management utility.
 *
 * Provides:
 * - Automatic cleanup on manager destruction
 * - Promise-based delays
 * - Named timeouts for debugging
 * - Batch timeout management
 */
export class TimeoutManager {
    constructor(config = {}) {
        this.timeouts = new Map();
        this.idCounter = 0;
        this.config = {
            defaultTimeoutMs: config.defaultTimeoutMs ?? 30000,
            warnOnClear: config.warnOnClear ?? false,
        };
    }
    /**
     * Set a named timeout.
     *
     * @param callback - Function to execute on timeout
     * @param ms - Timeout in milliseconds
     * @param name - Optional name for debugging
     * @returns TimeoutHandle for clearing the timeout
     */
    set(callback, ms, name) {
        const id = name ?? `timeout_${++this.idCounter}`;
        // Clear any existing timeout with same name
        if (this.timeouts.has(id)) {
            if (this.config.warnOnClear) {
                console.warn(`TimeoutManager: Overwriting existing timeout "${id}"`);
            }
            this.clear(id);
        }
        const timeoutId = setTimeout(() => {
            this.timeouts.delete(id);
            callback();
        }, ms);
        this.timeouts.set(id, timeoutId);
        return {
            id,
            cleared: false,
            clear: () => this.clear(id),
        };
    }
    /**
     * Set a timeout with a Promise-based API.
     *
     * @param ms - Delay in milliseconds
     * @returns Promise that resolves after the delay
     */
    delay(ms) {
        return new Promise(resolve => {
            this.set(resolve, ms, `delay_${++this.idCounter}`);
        });
    }
    /**
     * Set a timeout that can be awaited.
     *
     * @param ms - Delay in milliseconds
     * @returns Object with promise and cancel function
     */
    wait(ms) {
        let cancelFn;
        const promise = new Promise(resolve => {
            const handle = this.set(() => resolve(), ms);
            cancelFn = () => handle.clear();
        });
        return { promise, cancel: cancelFn };
    }
    /**
     * Clear a specific timeout by ID.
     *
     * @param id - The timeout ID to clear
     * @returns true if timeout was found and cleared
     */
    clear(id) {
        const timeoutId = this.timeouts.get(id);
        if (timeoutId) {
            clearTimeout(timeoutId);
            this.timeouts.delete(id);
            return true;
        }
        return false;
    }
    /**
     * Clear all timeouts.
     */
    clearAll() {
        for (const timeoutId of this.timeouts.values()) {
            clearTimeout(timeoutId);
        }
        this.timeouts.clear();
    }
    /**
     * Check if a specific timeout is active.
     *
     * @param id - The timeout ID to check
     * @returns true if the timeout is active
     */
    has(id) {
        return this.timeouts.has(id);
    }
    /**
     * Get the number of active timeouts.
     */
    get size() {
        return this.timeouts.size;
    }
    /**
     * Get all active timeout IDs.
     */
    get activeIds() {
        return Array.from(this.timeouts.keys());
    }
    /**
     * Destroy the manager and clear all timeouts.
     * Should be called when the manager is no longer needed.
     */
    destroy() {
        this.clearAll();
    }
}
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a new TimeoutManager instance.
 *
 * @param config - Optional configuration
 * @returns A new TimeoutManager instance
 *
 * @example
 * ```ts
 * import { createTimeoutManager } from './utils/timeoutManager.js';
 *
 * const timeoutMgr = createTimeoutManager();
 *
 * // Delay helper
 * await timeoutMgr.delay(5000);
 *
 * // Set custom timeout
 * const id = timeoutMgr.setTimeout(() => console.log("Done!"), 3000);
 * timeoutMgr.clearTimeout(id);
 * ```
 */
export function createTimeoutManager(config) {
    return new TimeoutManager(config);
}
// ============================================================================
// Re-exports
// ============================================================================
export default TimeoutManager;
