// openclaw-sdk/dist/cjs/managers/event.js
"use strict";
/**
 * Event Manager
 *
 * Event subscription system with wildcard support for server-sent events.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventManager = exports.MAX_EVENT_NAME_LENGTH = void 0;
exports.createEventManager = createEventManager;
// ============================================================================
// Types
// ============================================================================
/** Maximum event name length to prevent ReDoS */
exports.MAX_EVENT_NAME_LENGTH = 128;
// ============================================================================
// Event Manager
// ============================================================================
/**
 * Event manager for client-side event handling.
 *
 * Supports:
 * - Exact match: 'tick' matches only 'tick'
 * - Prefix wildcard: 'agent:*' matches 'agent.event', 'agent.created'
 * - Global wildcard: '*' matches all events
 */
class EventManager {
    constructor() {
        this.subscriptions = new Map();
        this.wildcardSubscriptions = [];
        this.defaultNamespace = '__default__';
        this.listenerErrorHandler = null;
    }
    /**
     * Set a callback for handling errors thrown by event listeners.
     *
     * By default, listener errors are logged to console.error. Use this method
     * to customize error handling (e.g., send to error tracking service).
     *
     * @param handler - Callback function to handle listener errors
     *
     * @example
     * ```ts
     * events.onListenerError(({ error, eventName, pattern }) => {
     *   console.error(`Handler for pattern "${pattern}" failed on event "${eventName}":`, error);
     *   // Send to error tracking service
     *   Sentry.captureException(error);
     * });
     * ```
     */
    onListenerError(handler) {
        this.listenerErrorHandler = handler;
    }
    /**
     * Subscribe to events.
     *
     * @param pattern Event pattern ('*', 'agent:*', 'event.name')
     * @param handler Event handler function
     * @param namespace Optional namespace for isolation
     * @returns Unsubscribe function
     *
     * @example
     * ```ts
     * const unsub1 = events.on('*', (e) => console.log(e.event));
     * const unsub2 = events.on('agent:*', (e) => console.log(e.event));
     * const unsub3 = events.on('tick', (e) => console.log(e.payload));
     * ```
     */
    on(pattern, handler, namespace) {
        // Validate event name
        if (pattern.length > exports.MAX_EVENT_NAME_LENGTH) {
            throw new Error(`Event pattern exceeds max length of ${exports.MAX_EVENT_NAME_LENGTH}`);
        }
        const entry = {
            pattern,
            handler: handler,
            namespace: namespace ?? this.defaultNamespace,
            priority: 0,
        };
        if (pattern === '*') {
            this.wildcardSubscriptions.push(entry);
        }
        else if (pattern.endsWith(':*')) {
            // Prefix wildcard - store without the trailing *
            const prefix = pattern.slice(0, -2);
            const key = `wildcard:${prefix}`;
            const existing = this.subscriptions.get(key) ?? [];
            existing.push(entry);
            this.subscriptions.set(key, existing);
        }
        else {
            // Exact match
            const existing = this.subscriptions.get(pattern) ?? [];
            existing.push(entry);
            this.subscriptions.set(pattern, existing);
        }
        // Return unsubscribe function
        return () => this.off(pattern, handler, namespace);
    }
    /**
     * Subscribe to event once.
     */
    once(pattern, handler, namespace) {
        const wrapped = event => {
            handler(event);
            this.off(pattern, wrapped, namespace);
        };
        return this.on(pattern, wrapped, namespace);
    }
    /**
     * Unsubscribe from events.
     */
    off(pattern, handler, namespace) {
        if (!pattern) {
            // Clear all
            if (namespace) {
                this.clearNamespace(namespace);
            }
            else {
                this.subscriptions.clear();
                this.wildcardSubscriptions = [];
            }
            return;
        }
        // Determine target namespace: use provided namespace or default to '__default__'
        // This ensures entries stored with default namespace can be removed when no namespace specified
        const targetNs = namespace ?? this.defaultNamespace;
        const filterByNs = (e) => {
            // Keep entries NOT in the target namespace (remove entries in target namespace)
            return e.namespace !== targetNs;
        };
        const filterByHandler = handler
            ? (e) => {
                // Keep entries where handler OR namespace doesn't match (remove matching entries)
                return !(e.handler === handler && e.namespace === targetNs);
            }
            : filterByNs;
        if (pattern === '*') {
            this.wildcardSubscriptions = this.wildcardSubscriptions.filter(filterByHandler);
        }
        else if (pattern.endsWith(':*')) {
            const prefix = pattern.slice(0, -2);
            const key = `wildcard:${prefix}`;
            const existing = this.subscriptions.get(key);
            if (existing) {
                this.subscriptions.set(key, existing.filter(filterByHandler));
            }
        }
        else {
            const existing = this.subscriptions.get(pattern);
            if (existing) {
                this.subscriptions.set(pattern, existing.filter(filterByHandler));
            }
        }
    }
    /**
     * Emit an event to all matching handlers.
     */
    emit(eventName, payload) {
        // Validate incoming event name length
        if (eventName.length > exports.MAX_EVENT_NAME_LENGTH) {
            console.warn(`Received event name exceeds max length (${eventName.length}), dropping`);
            return;
        }
        const called = new Set();
        // 1. Exact match
        const exactHandlers = this.subscriptions.get(eventName);
        if (exactHandlers) {
            for (const entry of exactHandlers) {
                called.add(entry.handler);
                this.safeCall(entry.handler, payload, eventName, entry.pattern);
            }
        }
        // 2. Prefix wildcard matches
        // Find all wildcard subscriptions where eventName starts with prefix
        for (const [key, entries] of this.subscriptions) {
            if (key.startsWith('wildcard:')) {
                const prefix = key.slice(9); // Remove 'wildcard:'
                if (eventName.startsWith(prefix + ':')) {
                    for (const entry of entries) {
                        if (!called.has(entry.handler)) {
                            called.add(entry.handler);
                            this.safeCall(entry.handler, payload, eventName, entry.pattern);
                        }
                    }
                }
            }
        }
        // 3. Global wildcard
        for (const entry of this.wildcardSubscriptions) {
            if (!called.has(entry.handler)) {
                this.safeCall(entry.handler, payload, eventName, entry.pattern);
            }
        }
    }
    /**
     * Emit a protocol EventFrame.
     */
    emitFrame(frame) {
        this.emit(frame.event, frame);
    }
    /**
     * Get handler count.
     */
    handlerCount(pattern, namespace) {
        if (!pattern) {
            let count = this.wildcardSubscriptions.length;
            for (const handlers of this.subscriptions.values()) {
                count += handlers.length;
            }
            if (namespace) {
                count -= this.subscriptions.size; // Approximate
            }
            return count;
        }
        if (pattern === '*') {
            const handlers = this.wildcardSubscriptions;
            return namespace ? handlers.filter(h => h.namespace === namespace).length : handlers.length;
        }
        const key = pattern.endsWith(':*') ? `wildcard:${pattern.slice(0, -2)}` : pattern;
        const handlers = this.subscriptions.get(key);
        if (!handlers)
            return 0;
        return namespace ? handlers.filter(h => h.namespace === namespace).length : handlers.length;
    }
    /**
     * Clear all handlers for a namespace.
     */
    clearNamespace(namespace) {
        for (const [key, handlers] of this.subscriptions) {
            this.subscriptions.set(key, handlers.filter(h => h.namespace !== namespace));
        }
        this.wildcardSubscriptions = this.wildcardSubscriptions.filter(h => h.namespace !== namespace);
    }
    /**
     * Safely call handler, catching errors.
     */
    safeCall(handler, payload, eventName, pattern) {
        try {
            handler(payload);
        }
        catch (error) {
            if (this.listenerErrorHandler) {
                this.listenerErrorHandler({ error, eventName, pattern });
            }
            else {
                // Default: log with context for debugging
                console.error(`Error in event handler for pattern "${pattern}" on event "${eventName}":`, error);
            }
        }
    }
}
exports.EventManager = EventManager;
// ============================================================================
// Factory
// ============================================================================
/**
 * Create an event manager.
 *
 * @returns New EventManager instance
 *
 * @example
 * ```ts
 * const events = createEventManager();
 *
 * // Subscribe to events
 * const unsub = events.on('tick', (payload) => {
 *   console.log("Tick:", payload);
 * });
 *
 * // Emit events
 * events.emit('tick', { timestamp: Date.now() });
 *
 * // Unsubscribe
 * unsub();
 * ```
 */
function createEventManager() {
    return new EventManager();
}
