// openclaw-sdk/dist/cjs/managers/request.js
"use strict";
/**
 * OpenClaw Request Manager
 *
 * This module provides a request manager that tracks pending requests and correlates
 * responses with their corresponding requests. It handles timeouts and cleanup of
 * completed requests.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestManager = void 0;
exports.createRequestManager = createRequestManager;
// ============================================================================
// Request Manager
// ============================================================================
/**
 * Request Manager
 *
 * Tracks pending requests and correlates responses with their corresponding requests.
 * Handles timeouts and cleanup of completed requests.
 */
class RequestManager {
    constructor() {
        /** Map of pending requests by ID */
        this.pendingRequests = new Map();
    }
    /**
     * Add a new pending request
     *
     * Creates a promise that will resolve when a response is received or reject
     * if the request times out.
     *
     * @param id - The request ID
     * @param options - Request options
     * @returns Promise that resolves with the response or rejects on timeout/error
     */
    addRequest(id, options) {
        // Check if a request with this ID already exists
        if (this.pendingRequests.has(id)) {
            return Promise.reject(new Error(`Request with ID "${id}" already exists`));
        }
        return new Promise((resolve, reject) => {
            // Set up timeout
            const timeoutId = setTimeout(() => {
                // Remove from pending requests
                this.pendingRequests.delete(id);
                // Reject with timeout error
                reject(new Error(`Request "${id}" timed out after ${options.timeout}ms`));
            }, options.timeout);
            // Store the request entry
            const entry = {
                id,
                resolve,
                reject,
                timeoutId,
            };
            this.pendingRequests.set(id, entry);
        });
    }
    /**
     * Resolve a pending request with a response
     *
     * @param id - The request ID
     * @param response - The response frame
     * @throws Error if no pending request with the given ID exists
     */
    resolveRequest(id, response) {
        const entry = this.pendingRequests.get(id);
        if (!entry) {
            throw new Error(`No pending request with ID "${id}"`);
        }
        // Clean up the timeout timer
        if (entry.timeoutId !== null) {
            clearTimeout(entry.timeoutId);
        }
        // Remove from pending requests
        this.pendingRequests.delete(id);
        // Resolve the promise
        entry.resolve(response);
    }
    /**
     * Reject a pending request with an error
     *
     * @param id - The request ID
     * @param error - The error to reject with
     * @throws Error if no pending request with the given ID exists
     */
    rejectRequest(id, error) {
        const entry = this.pendingRequests.get(id);
        if (!entry) {
            throw new Error(`No pending request with ID "${id}"`);
        }
        // Clean up the timeout timer
        if (entry.timeoutId !== null) {
            clearTimeout(entry.timeoutId);
        }
        // Remove from pending requests
        this.pendingRequests.delete(id);
        // Reject the promise
        entry.reject(error);
    }
    /**
     * Abort a pending request
     *
     * Similar to rejectRequest but uses a standard abort error.
     *
     * @param id - The request ID
     * @throws Error if no pending request with the given ID exists
     */
    abortRequest(id) {
        const entry = this.pendingRequests.get(id);
        if (!entry) {
            throw new Error(`No pending request with ID "${id}"`);
        }
        // Clean up the timeout timer
        if (entry.timeoutId !== null) {
            clearTimeout(entry.timeoutId);
        }
        // Remove from pending requests
        this.pendingRequests.delete(id);
        // Reject with abort error
        entry.reject(new Error(`Request "${id}" was aborted`));
    }
    /**
     * Clear all pending requests
     *
     * Aborts all pending requests and clears the pending requests map.
     * Useful when closing a connection or cleaning up.
     */
    clear() {
        // Abort all pending requests
        for (const [id, entry] of this.pendingRequests) {
            // Clean up the timeout timer
            if (entry.timeoutId !== null) {
                clearTimeout(entry.timeoutId);
            }
            // Reject with abort error
            entry.reject(new Error(`Request "${id}" was aborted due to clear()`));
        }
        // Clear the map
        this.pendingRequests.clear();
    }
    /**
     * Get the number of pending requests
     *
     * @returns The number of pending requests
     */
    get pendingCount() {
        return this.pendingRequests.size;
    }
    /**
     * Check if a request with the given ID is pending
     *
     * @param id - The request ID
     * @returns True if the request is pending, false otherwise
     */
    isPending(id) {
        return this.pendingRequests.has(id);
    }
}
exports.RequestManager = RequestManager;
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a request manager instance
 *
 * @returns A new request manager instance
 *
 * @example
 * ```ts
 * import { createRequestManager } from './managers/request.js';
 *
 * const requestMgr = createRequestManager();
 *
 * // Add a pending request
 * const responsePromise = requestMgr.addRequest("req-1", { timeout: 30000 });
 *
 * // Later: resolve the request
 * requestMgr.resolveRequest("req-1", { type: "res", id: "req-1", ok: true, payload: {} });
 * ```
 */
function createRequestManager() {
    return new RequestManager();
}
// ============================================================================
// Re-exports
// ============================================================================
exports.default = RequestManager;
