// openclaw-sdk/dist/cjs/events/gap.js
"use strict";
/**
 * Gap Detector
 *
 * Detects gaps in event sequence numbers.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GapDetector = void 0;
exports.createGapDetector = createGapDetector;
const events_1 = require("events");
// Default max gaps to track
const DEFAULT_MAX_GAPS = 100;
// ============================================================================
// Gap Detector
// ============================================================================
/**
 * Gap detector for sequence number tracking.
 *
 * Detects gaps in event sequence numbers and can trigger
 * recovery actions based on configured mode.
 */
class GapDetector extends events_1.EventEmitter {
    /**
     * Create a gap detector.
     *
     * @param config Configuration
     */
    constructor(config) {
        super();
        this.lastSequence = null;
        this.gaps = [];
        this.recovery = config.recovery;
        this.maxGaps = config.maxGaps ?? DEFAULT_MAX_GAPS;
    }
    /**
     * Record an incoming sequence number.
     *
     * @param seq Sequence number from event
     */
    recordSequence(seq) {
        // Check for gap if we have a previous sequence
        if (this.lastSequence !== null) {
            const expected = this.lastSequence + 1;
            // Only detect gaps for sequences after the last one
            // Duplicate or old sequences are ignored
            if (seq > this.lastSequence && seq > expected) {
                const gap = {
                    expected,
                    received: seq,
                    detectedAt: Date.now(),
                };
                this.gaps.push(gap);
                // Trim if exceeds max
                if (this.gaps.length > this.maxGaps) {
                    this.gaps = this.gaps.slice(-this.maxGaps);
                }
                // Emit callback
                if (this.recovery.onGap) {
                    this.recovery.onGap(this.gaps);
                }
                // Emit event
                this.emit('gap', this.gaps);
            }
        }
        this.lastSequence = seq;
    }
    /**
     * Check if gaps exist.
     *
     * @returns true if gaps detected
     */
    hasGap() {
        return this.gaps.length > 0;
    }
    /**
     * Get all detected gaps.
     *
     * @returns Array of gap information
     */
    getGaps() {
        return [...this.gaps];
    }
    /**
     * Get the last recorded sequence number.
     *
     * @returns Last sequence or null
     */
    getLastSequence() {
        return this.lastSequence;
    }
    /**
     * Reset sequence tracking.
     */
    reset() {
        this.lastSequence = null;
        this.gaps = [];
    }
}
exports.GapDetector = GapDetector;
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a gap detector.
 *
 * @param config - Gap detector configuration
 * @returns A new GapDetector instance
 *
 * @example
 * ```ts
 * import { createGapDetector } from './events/gap.js';
 *
 * const gapDetector = createGapDetector({
 *   expectedSequence: 1,
 *   onGap: (gapInfo) => {
 *     console.log(`Gap detected: expected ${gapInfo.expected}, got ${gapInfo.received}`);
 *   }
 * });
 *
 * // Process incoming event frames
 * gapDetector.processSequence(1); // ok
 * gapDetector.processSequence(3); // gap! (missed 2)
 * ```
 */
function createGapDetector(config) {
    return new GapDetector(config);
}
