// openclaw/dist/api-DylG1XX8.js
import { _ as uniqueStrings } from "./string-normalization-B8G0vlWE.js";
import "./string-coerce-runtime-Ccm67Pxa.js";
import "./provider-catalog-CpiJJWRZ.js";
import "./models-vSpoc_mc.js";
//#region extensions/volcengine/api.ts
const VOLCENGINE_UNSUPPORTED_TOOL_SCHEMA_KEYWORDS = [
	"minLength",
	"maxLength",
	"minItems",
	"maxItems",
	"minContains",
	"maxContains"
];
function mergeUnsupportedToolSchemaKeywords(existing) {
	return uniqueStrings([...existing ?? [], ...VOLCENGINE_UNSUPPORTED_TOOL_SCHEMA_KEYWORDS]);
}
function resolveVolcengineToolSchemaCompatPatch(compat) {
	return { unsupportedToolSchemaKeywords: mergeUnsupportedToolSchemaKeywords(compat?.unsupportedToolSchemaKeywords) };
}
function applyVolcengineToolSchemaCompat(model) {
	const unsupportedToolSchemaKeywords = mergeUnsupportedToolSchemaKeywords(model.compat?.unsupportedToolSchemaKeywords);
	if (model.compat?.unsupportedToolSchemaKeywords?.length === unsupportedToolSchemaKeywords.length && unsupportedToolSchemaKeywords.every((keyword, index) => model.compat?.unsupportedToolSchemaKeywords?.[index] === keyword)) return model;
	return {
		...model,
		compat: {
			...model.compat,
			unsupportedToolSchemaKeywords
		}
	};
}
//#endregion
export { applyVolcengineToolSchemaCompat as n, resolveVolcengineToolSchemaCompatPatch as r, VOLCENGINE_UNSUPPORTED_TOOL_SCHEMA_KEYWORDS as t };
