// openclaw/dist/browser-cli-shared-Bqqxa2QZ.js
import { c as normalizeOptionalString } from "./string-coerce-DKw2K5wM.js";
import { m as parseStrictPositiveInteger, p as parseStrictNonNegativeInteger } from "./number-coercion-DjYijOzx.js";
import "./string-coerce-runtime-Ccm67Pxa.js";
import "./number-runtime-FvL6JHZB.js";
import { n as callGatewayFromCli } from "./gateway-rpc-BHQouYas.js";
import { r as BROWSER_REQUEST_GATEWAY_SCOPES, t as BROWSER_REQUEST_GATEWAY_METHOD } from "./browser-gateway-contract-yfVqwwwt.js";
import { f as normalizeBrowserTimerDelayMs } from "./cdp.helpers-WyU0BhTo.js";
import "./core-api-iEfXC4Xf.js";
//#region extensions/browser/src/cli/browser-cli-shared.ts
function normalizeQuery(query) {
	if (!query) return;
	const out = {};
	for (const [key, value] of Object.entries(query)) {
		if (value === void 0) continue;
		out[key] = String(value);
	}
	return Object.keys(out).length ? out : void 0;
}
function parseBrowserPositiveIntegerValue(value) {
	return parseStrictPositiveInteger(value);
}
function parseBrowserNonNegativeIntegerValue(value) {
	return parseStrictNonNegativeInteger(value);
}
function parseBrowserPositiveIntegerOption(raw, flag) {
	const parsed = parseBrowserPositiveIntegerValue(raw);
	if (parsed === void 0) throw new Error(`${flag} must be a positive integer.`);
	return parsed;
}
function parseBrowserNonNegativeIntegerOption(raw, flag) {
	const parsed = parseBrowserNonNegativeIntegerValue(raw);
	if (parsed === void 0) throw new Error(`${flag} must be a non-negative integer.`);
	return parsed;
}
async function callBrowserRequest(opts, params, extra) {
	const resolvedTimeoutMs = typeof extra?.timeoutMs === "number" && Number.isFinite(extra.timeoutMs) ? normalizeBrowserTimerDelayMs(extra.timeoutMs) : typeof opts.timeout === "string" ? normalizeBrowserTimerDelayMs(parseBrowserPositiveIntegerOption(opts.timeout, "--timeout")) : void 0;
	const resolvedTimeout = typeof resolvedTimeoutMs === "number" && Number.isFinite(resolvedTimeoutMs) ? resolvedTimeoutMs : void 0;
	const timeout = typeof resolvedTimeout === "number" ? String(resolvedTimeout) : opts.timeout;
	const payload = await callGatewayFromCli(BROWSER_REQUEST_GATEWAY_METHOD, {
		...opts,
		timeout
	}, {
		method: params.method,
		path: params.path,
		query: normalizeQuery(params.query),
		body: params.body,
		timeoutMs: resolvedTimeout
	}, {
		progress: extra?.progress,
		scopes: [...BROWSER_REQUEST_GATEWAY_SCOPES]
	});
	if (payload === void 0) throw new Error("Unexpected browser.request response");
	return payload;
}
async function callBrowserResize(opts, params, extra) {
	return callBrowserRequest(opts, {
		method: "POST",
		path: "/act",
		query: params.profile ? { profile: params.profile } : void 0,
		body: {
			kind: "resize",
			width: params.width,
			height: params.height,
			targetId: normalizeOptionalString(params.targetId)
		}
	}, extra);
}
//#endregion
export { parseBrowserPositiveIntegerOption as a, parseBrowserNonNegativeIntegerValue as i, callBrowserResize as n, parseBrowserPositiveIntegerValue as o, parseBrowserNonNegativeIntegerOption as r, callBrowserRequest as t };
