// openclaw/dist/bundled-health-checks-jg1Yn_fk.js
import { r as asOptionalObjectRecord } from "./record-coerce-Btbek4uV.js";
import { s as normalizePluginsConfig } from "./config-state-Cz5agD6j.js";
import { t as loadBundledPluginPublicArtifactModuleSync } from "./public-surface-loader-DhFvfS89.js";
import { i as passesManifestOwnerBasePolicy } from "./manifest-owner-policy-uC_3_nv-.js";
import { i as registerHealthCheck } from "./health-check-registry-DxAtOKZD.js";
//#region src/flows/bundled-health-checks.ts
function registerBundledHealthChecks(params) {
	if (!shouldRegisterPolicyHealth(params)) return;
	loadBundledPluginPublicArtifactModuleSync({
		dirName: "policy",
		artifactBasename: "api.js"
	}).registerPolicyDoctorChecks?.({ registerHealthCheck });
}
function shouldRegisterPolicyHealth(params) {
	const entry = params.cfg.plugins?.entries?.policy;
	const config = asOptionalObjectRecord(entry?.config) ?? {};
	if (entry === void 0 || entry.enabled === false || config.enabled === false) return false;
	if (!passesManifestOwnerBasePolicy({
		plugin: { id: "policy" },
		normalizedConfig: normalizePluginsConfig(params.cfg.plugins)
	})) return false;
	return entry.enabled === true || config.enabled === true;
}
//#endregion
export { registerBundledHealthChecks as t };
