// openclaw/dist/copilot-runtime-plugin-install-Bk-Uof2s.js
import { p as resolveUserPath } from "./utils-DlHt_n1Q.js";
import { n as loadInstalledPluginIndexInstallRecords } from "./installed-plugin-index-record-reader-B_NZbZil.js";
import "./installed-plugin-index-records-B68pYawc.js";
import { t as enablePluginInConfig } from "./enable-DcBk9n3x.js";
import { c as parseModelRefProvider } from "./openai-codex-routing-Cc71DT_j.js";
import { t as resolveModelRuntimePolicy } from "./model-runtime-policy-ZCInBBPn.js";
import { existsSync } from "node:fs";
import path from "node:path";
//#region src/agents/copilot-routing.ts
const GITHUB_COPILOT_PROVIDER_ID = "github-copilot";
/**
* Canonical id of the Copilot agent runtime plugin.
*/
const COPILOT_RUNTIME_ID = "copilot";
function parseModelRefId(model) {
	if (typeof model !== "string") return;
	const trimmed = model.trim();
	const slash = trimmed.indexOf("/");
	if (slash <= 0 || slash === trimmed.length - 1) return;
	return trimmed.slice(slash + 1);
}
/**
* Returns true when the selected model should trigger the external
* `@openclaw/copilot` runtime plugin install.
*
* Gating contract (review #2, P1):
*   - Model ref must use the `github-copilot/*` provider prefix.
*   - The user's config must explicitly opt in by setting
*     `agentRuntime.id: "copilot"` at the provider, model, or agent scope
*     (resolved via `resolveModelRuntimePolicy`).
*
* Without the explicit opt-in we fall through to the built-in GitHub
* Copilot provider, which has shipped support for `github-copilot/*`
* models for a long time and must not install the runtime plugin for
* users who never asked for it.
*/
function modelSelectionShouldEnsureCopilotRuntimePlugin(params) {
	if (parseModelRefProvider(params.model) !== "github-copilot") return false;
	const modelId = parseModelRefId(params.model);
	return resolveModelRuntimePolicy({
		config: params.config,
		provider: GITHUB_COPILOT_PROVIDER_ID,
		modelId
	}).policy?.id?.trim().toLowerCase() === COPILOT_RUNTIME_ID;
}
//#endregion
//#region src/commands/copilot-runtime-plugin-install.ts
const COPILOT_RUNTIME_PLUGIN_ID = "copilot";
const COPILOT_RUNTIME_PLUGIN_LABEL = "GitHub Copilot agent runtime";
const COPILOT_RUNTIME_PLUGIN_NPM_SPEC = "@openclaw/copilot";
function isInstalledRecordPresentOnDisk(record, env) {
	const installPath = record?.installPath?.trim();
	if (!installPath) return false;
	return existsSync(path.join(resolveUserPath(installPath, env), "package.json"));
}
function selectedModelShouldEnsureCopilotRuntimePlugin(params) {
	return modelSelectionShouldEnsureCopilotRuntimePlugin({
		config: params.cfg,
		model: params.model
	});
}
async function ensureCopilotRuntimePluginForModelSelection(params) {
	if (!selectedModelShouldEnsureCopilotRuntimePlugin({
		cfg: params.cfg,
		model: params.model
	})) return {
		cfg: params.cfg,
		required: false,
		installed: false
	};
	if (isInstalledRecordPresentOnDisk((await loadInstalledPluginIndexInstallRecords({ env: process.env }))["copilot"], process.env)) {
		const repair = await repairCopilotRuntimePluginInstallForModelSelection({
			cfg: params.cfg,
			model: params.model,
			env: process.env
		});
		for (const change of repair.changes) params.runtime.log?.(change);
		for (const warning of repair.warnings) params.runtime.log?.(`GitHub Copilot update warning: ${warning}`);
		const enableResult = enablePluginInConfig(params.cfg, COPILOT_RUNTIME_PLUGIN_ID);
		return {
			cfg: enableResult.enabled ? enableResult.config : params.cfg,
			required: true,
			installed: true,
			status: "installed"
		};
	}
	const { ensureOnboardingPluginInstalled } = await import("./onboarding-plugin-install-RDvgSSVl.js");
	const result = await ensureOnboardingPluginInstalled({
		cfg: params.cfg,
		entry: {
			pluginId: COPILOT_RUNTIME_PLUGIN_ID,
			label: COPILOT_RUNTIME_PLUGIN_LABEL,
			install: {
				npmSpec: COPILOT_RUNTIME_PLUGIN_NPM_SPEC,
				defaultChoice: "npm"
			},
			trustedSourceLinkedOfficialInstall: true,
			preferRemoteInstall: true
		},
		prompter: params.prompter,
		runtime: params.runtime,
		...params.workspaceDir !== void 0 ? { workspaceDir: params.workspaceDir } : {},
		promptInstall: false,
		autoConfirmSingleSource: true
	});
	return {
		cfg: result.cfg,
		required: true,
		installed: result.installed,
		status: result.status
	};
}
async function repairCopilotRuntimePluginInstallForModelSelection(params) {
	if (!selectedModelShouldEnsureCopilotRuntimePlugin({
		cfg: params.cfg,
		model: params.model
	})) return {
		required: false,
		changes: [],
		warnings: []
	};
	const { repairMissingPluginInstallsForIds } = await import("./missing-configured-plugin-install-CXsLtFKu.js");
	const result = await repairMissingPluginInstallsForIds({
		cfg: params.cfg,
		pluginIds: [COPILOT_RUNTIME_PLUGIN_ID],
		...params.env !== void 0 ? { env: params.env } : {}
	});
	return {
		required: true,
		changes: result.changes,
		warnings: result.warnings
	};
}
//#endregion
export { selectedModelShouldEnsureCopilotRuntimePlugin as i, ensureCopilotRuntimePluginForModelSelection as n, repairCopilotRuntimePluginInstallForModelSelection as r, COPILOT_RUNTIME_PLUGIN_ID as t };
