// openclaw/dist/dispatch-acp-transcript.runtime-D3m6fbQY.js
import { _ as resolveSessionAgentId } from "./agent-scope-BeI1SY1H.js";
import { u as resolveStorePath } from "./paths-CBJPLq30.js";
import { t as loadSessionStore, z as resolveSessionStoreEntry } from "./store-load-BHiTqlRK.js";
import "./sessions-BN2mnV-J.js";
import { n as resolveAcpSessionCwd } from "./session-identifiers-BgkN7cWT.js";
import { c as persistAcpTurnTranscript } from "./attempt-execution-3yrKLFR9.js";
//#region src/auto-reply/reply/dispatch-acp-transcript.runtime.ts
async function persistAcpDispatchTranscript(params) {
	const promptText = params.promptText.trim();
	const finalText = params.finalText.trim();
	if (!promptText && !finalText) return;
	const sessionAgentId = resolveSessionAgentId({
		sessionKey: params.sessionKey,
		config: params.cfg
	});
	const storePath = resolveStorePath(params.cfg.session?.store, { agentId: sessionAgentId });
	const sessionStore = loadSessionStore(storePath, { skipCache: true });
	const sessionEntry = resolveSessionStoreEntry({
		store: sessionStore,
		sessionKey: params.sessionKey
	}).existing;
	const sessionId = sessionEntry?.sessionId;
	if (!sessionId) throw new Error(`unknown ACP session key: ${params.sessionKey}`);
	await persistAcpTurnTranscript({
		body: promptText,
		transcriptBody: promptText,
		finalText,
		sessionId,
		sessionKey: params.sessionKey,
		sessionEntry,
		sessionStore,
		storePath,
		sessionAgentId,
		threadId: params.threadId,
		sessionCwd: resolveAcpSessionCwd(params.meta) ?? process.cwd(),
		config: params.cfg
	});
}
//#endregion
export { persistAcpDispatchTranscript };
