// openclaw/dist/model-picker-visibility-DzVys-Rr.js
import { r as normalizeProviderId } from "./provider-id-DhcncFyL.js";
import { r as listCliRuntimeProviderIds } from "./cli-backends-rYsy36w0.js";
import { i as isCliRuntimeProvider } from "./model-runtime-aliases-DdKzTF-r.js";
//#region src/agents/model-picker-visibility.ts
const RETIRED_MODEL_PICKER_PROVIDERS = new Set(["codex", "codex-cli"]);
function createModelPickerVisibleProviderPredicate(params = {}) {
	const cliRuntimeProviders = new Set(listCliRuntimeProviderIds({
		config: params.config,
		env: params.env,
		includeSetupRegistry: params.includeSetupRegistry ?? false
	}));
	return (provider) => {
		const normalized = normalizeProviderId(provider);
		return !RETIRED_MODEL_PICKER_PROVIDERS.has(normalized) && !cliRuntimeProviders.has(normalized);
	};
}
function isModelPickerVisibleProvider(provider) {
	const normalized = normalizeProviderId(provider);
	return !RETIRED_MODEL_PICKER_PROVIDERS.has(normalized) && !isCliRuntimeProvider(normalized, { includeSetupRegistry: true });
}
//#endregion
export { isModelPickerVisibleProvider as n, createModelPickerVisibleProviderPredicate as t };
