// openclaw/dist/number-coercion-DjYijOzx.js
//#region src/shared/number-coercion.ts
function asFiniteNumber(value) {
	return typeof value === "number" && Number.isFinite(value) ? value : void 0;
}
function asFiniteNumberInRange(value, range) {
	const number = asFiniteNumber(value);
	if (number === void 0) return;
	if (range.min !== void 0) {
		if (range.minExclusive ? number <= range.min : number < range.min) return;
	}
	if (range.max !== void 0) {
		if (range.maxExclusive ? number >= range.max : number > range.max) return;
	}
	return number;
}
function asSafeIntegerInRange(value, range) {
	if (typeof value !== "number" || !Number.isSafeInteger(value)) return;
	if (range.min !== void 0 && value < range.min) return;
	if (range.max !== void 0 && value > range.max) return;
	return value;
}
function normalizeNumericString(value) {
	const trimmed = value.trim();
	return trimmed ? trimmed : void 0;
}
function parseFiniteNumber(value) {
	if (typeof value === "number") return Number.isFinite(value) ? value : void 0;
	return parseStrictFiniteNumber(value);
}
function parseStrictInteger(value) {
	if (typeof value === "number") return Number.isSafeInteger(value) ? value : void 0;
	if (typeof value !== "string") return;
	const normalized = normalizeNumericString(value);
	if (!normalized || !/^[+-]?\d+$/.test(normalized)) return;
	const parsed = Number(normalized);
	return Number.isSafeInteger(parsed) ? parsed : void 0;
}
function parseStrictFiniteNumber(value) {
	if (typeof value === "number") return Number.isFinite(value) ? value : void 0;
	if (typeof value !== "string") return;
	const normalized = normalizeNumericString(value);
	if (!normalized || !/^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$/i.test(normalized)) return;
	const parsed = Number(normalized);
	return Number.isFinite(parsed) ? parsed : void 0;
}
function asPositiveSafeInteger(value) {
	return typeof value === "number" && Number.isSafeInteger(value) && value > 0 ? value : void 0;
}
const MAX_TIMER_TIMEOUT_MS = 2147e6;
const MAX_TIMER_TIMEOUT_SECONDS = Math.floor(MAX_TIMER_TIMEOUT_MS / 1e3);
function clampTimerTimeoutMs(valueMs, minMs = 1) {
	const value = asFiniteNumber(valueMs);
	if (value === void 0) return;
	return Math.min(Math.max(Math.floor(value), Math.max(1, Math.floor(minMs))), MAX_TIMER_TIMEOUT_MS);
}
function resolveTimerTimeoutMs(valueMs, fallbackMs, minMs = 1) {
	const value = asFiniteNumber(valueMs) ?? asFiniteNumber(fallbackMs);
	const min = Math.max(0, Math.floor(minMs));
	if (value === void 0) return min;
	return Math.min(Math.max(Math.floor(value), min), MAX_TIMER_TIMEOUT_MS);
}
function finiteSecondsToTimerSafeMilliseconds(value, opts = {}) {
	const seconds = asFiniteNumber(value);
	if (seconds === void 0 || seconds <= 0) return;
	const boundedSeconds = opts.floorSeconds ? Math.floor(seconds) : seconds;
	const milliseconds = Math.floor(boundedSeconds * 1e3);
	if (!Number.isFinite(milliseconds) || milliseconds <= 0) return;
	return Math.min(milliseconds, MAX_TIMER_TIMEOUT_MS);
}
function resolveIntegerOption(value, fallback, range = {}) {
	const floored = Math.floor(typeof value === "number" && Number.isFinite(value) ? value : fallback);
	const minBounded = range.min === void 0 ? floored : Math.max(range.min, floored);
	return range.max === void 0 ? minBounded : Math.min(range.max, minBounded);
}
function resolveOptionalIntegerOption(value, range = {}) {
	if (typeof value !== "number" || !Number.isFinite(value)) return;
	return resolveIntegerOption(value, value, range);
}
function resolveNonNegativeIntegerOption(value, fallback) {
	return resolveIntegerOption(value, fallback, { min: 0 });
}
function parseStrictPositiveInteger(value) {
	const parsed = parseStrictInteger(value);
	return parsed !== void 0 && parsed > 0 ? parsed : void 0;
}
function parseStrictNonNegativeInteger(value) {
	const parsed = parseStrictInteger(value);
	return parsed !== void 0 && parsed >= 0 ? parsed : void 0;
}
function positiveSecondsToSafeMilliseconds(value) {
	const seconds = parseStrictPositiveInteger(value);
	if (seconds === void 0) return;
	const milliseconds = seconds * 1e3;
	return Number.isSafeInteger(milliseconds) ? milliseconds : void 0;
}
function nonNegativeSecondsToSafeMilliseconds(value) {
	const seconds = parseStrictNonNegativeInteger(value);
	if (seconds === void 0) return;
	const milliseconds = seconds * 1e3;
	return Number.isSafeInteger(milliseconds) ? milliseconds : void 0;
}
function resolveExpiresAtMsFromDurationSeconds(value, opts = {}) {
	const durationMs = positiveSecondsToSafeMilliseconds(value);
	if (durationMs === void 0) return;
	const nowMs = opts.nowMs ?? Date.now();
	const expiresAt = nowMs + durationMs - (opts.bufferMs ?? 0);
	if (!Number.isSafeInteger(expiresAt)) return;
	const minRemainingMs = opts.minRemainingMs;
	return minRemainingMs === void 0 ? expiresAt : Math.max(expiresAt, nowMs + minRemainingMs);
}
function resolveExpiresAtMsFromEpochSeconds(value, opts = {}) {
	const epochMs = positiveSecondsToSafeMilliseconds(value);
	if (epochMs === void 0) return;
	const expiresAt = epochMs - (opts.bufferMs ?? 0);
	return Number.isSafeInteger(expiresAt) ? expiresAt : void 0;
}
function resolveExpiresAtMsFromDurationOrEpoch(value, opts = {}) {
	const parsed = parseStrictPositiveInteger(value);
	if (parsed === void 0) return;
	if (parsed < (opts.relativeSecondsThreshold ?? 1e9)) return resolveExpiresAtMsFromDurationSeconds(parsed, { nowMs: opts.nowMs });
	if (parsed < (opts.absoluteMillisecondsThreshold ?? 0xe8d4a51000)) return positiveSecondsToSafeMilliseconds(parsed);
	return parsed;
}
//#endregion
export { resolveTimerTimeoutMs as S, resolveExpiresAtMsFromDurationSeconds as _, asPositiveSafeInteger as a, resolveNonNegativeIntegerOption as b, finiteSecondsToTimerSafeMilliseconds as c, parseStrictFiniteNumber as d, parseStrictInteger as f, resolveExpiresAtMsFromDurationOrEpoch as g, positiveSecondsToSafeMilliseconds as h, asFiniteNumberInRange as i, nonNegativeSecondsToSafeMilliseconds as l, parseStrictPositiveInteger as m, MAX_TIMER_TIMEOUT_SECONDS as n, asSafeIntegerInRange as o, parseStrictNonNegativeInteger as p, asFiniteNumber as r, clampTimerTimeoutMs as s, MAX_TIMER_TIMEOUT_MS as t, parseFiniteNumber as u, resolveExpiresAtMsFromEpochSeconds as v, resolveOptionalIntegerOption as x, resolveIntegerOption as y };
