// openclaw/dist/openai-codex-routing-Cc71DT_j.js
import { s as normalizeOptionalLowercaseString } from "./string-coerce-DKw2K5wM.js";
import { n as findNormalizedProviderValue, r as normalizeProviderId } from "./provider-id-DhcncFyL.js";
import { r as resolveProviderIdForAuth } from "./provider-auth-aliases-DakmN5Fw.js";
//#region src/agents/agent-runtime-id.ts
const OPENCLAW_AGENT_RUNTIME_ID = "openclaw";
const AUTO_AGENT_RUNTIME_ID = "auto";
function normalizeEmbeddedAgentRuntime(raw) {
	const value = raw?.trim();
	if (!value) return OPENCLAW_AGENT_RUNTIME_ID;
	if (value === "openclaw" || value === "pi") return OPENCLAW_AGENT_RUNTIME_ID;
	if (value === "auto") return AUTO_AGENT_RUNTIME_ID;
	if (value === "codex-app-server") return "codex";
	return value;
}
function normalizeOptionalAgentRuntimeId(raw) {
	if (typeof raw !== "string") return;
	const value = raw.trim().toLowerCase();
	return value ? normalizeEmbeddedAgentRuntime(value) : void 0;
}
/**
* @deprecated Whole-agent runtime environment selection is retired. Use
* provider/model runtime policy or a registered agent harness instead.
*/
function resolveEmbeddedAgentRuntime(_env = process.env) {
	return OPENCLAW_AGENT_RUNTIME_ID;
}
function isDefaultAgentRuntimeId(runtime) {
	return runtime === void 0 || runtime === "auto" || runtime === "default";
}
//#endregion
//#region src/agents/openai-codex-routing.ts
const OPENAI_PROVIDER_ID = "openai";
const OPENAI_CODEX_PROVIDER_ID = "openai-codex";
function isOfficialOpenAIBaseUrl(baseUrl) {
	if (typeof baseUrl !== "string" || !baseUrl.trim()) return true;
	try {
		const url = new URL(baseUrl.trim());
		return url.protocol === "https:" && url.hostname.toLowerCase() === "api.openai.com" && (url.pathname === "" || url.pathname === "/" || url.pathname === "/v1" || url.pathname === "/v1/");
	} catch {
		return false;
	}
}
function openAIProviderUsesCustomBaseUrl(config) {
	return !isOfficialOpenAIBaseUrl(config?.models?.providers?.openai?.baseUrl);
}
function isOpenAIProvider(provider) {
	return normalizeProviderId(provider ?? "") === OPENAI_PROVIDER_ID;
}
function isOpenAICodexProvider(provider) {
	return normalizeProviderId(provider ?? "") === OPENAI_CODEX_PROVIDER_ID;
}
function openAIProviderUsesCodexRuntimeByDefault(params) {
	return isOpenAIProvider(params.provider) && !openAIProviderUsesCustomBaseUrl(params.config);
}
function parseModelRefProvider(value) {
	if (typeof value !== "string") return;
	const slashIndex = value.trim().indexOf("/");
	if (slashIndex <= 0) return;
	return normalizeProviderId(value.trim().slice(0, slashIndex));
}
function modelSelectionShouldEnsureCodexPlugin(params) {
	const provider = parseModelRefProvider(params.model);
	if (provider === "openai-codex") return true;
	return provider === "openai" && !openAIProviderUsesCustomBaseUrl(params.config);
}
function hasOpenAICodexAuthProfileOverride(value) {
	return typeof value === "string" && normalizeOptionalLowercaseString(value)?.startsWith(`openai-codex:`) === true;
}
function configuredOpenAIAuthOrderStartsWithCodexProfile(config) {
	if (!openAIProviderUsesCodexRuntimeByDefault({
		provider: "openai",
		config
	})) return false;
	const firstProfile = findNormalizedProviderValue(config?.auth?.order, OPENAI_PROVIDER_ID)?.find((profileId) => typeof profileId === "string" && profileId.trim().length > 0);
	return hasOpenAICodexAuthProfileOverride(firstProfile);
}
function shouldRouteOpenAIThroughCodexAuthProvider(params) {
	if (!isOpenAIProvider(params.provider)) return false;
	if ((normalizeOptionalAgentRuntimeId(params.agentHarnessId ?? params.harnessRuntime) ?? "openclaw") !== "openclaw") return false;
	if (!hasOpenAICodexAuthProfileOverride(params.authProfileId)) return false;
	const aliasLookupParams = {
		config: params.config,
		workspaceDir: params.workspaceDir
	};
	return resolveProviderIdForAuth(params.authProfileProvider ?? params.authProfileId?.split(":", 1)[0] ?? "", aliasLookupParams) === OPENAI_CODEX_PROVIDER_ID;
}
function listOpenAIAuthProfileProvidersForAgentRuntime(params) {
	if (!isOpenAIProvider(params.provider)) return [params.provider];
	const runtime = normalizeOptionalAgentRuntimeId(normalizeExplicitRuntimePin(params.agentHarnessId) ?? params.harnessRuntime) ?? "openclaw";
	if (runtime === "codex") return [OPENAI_CODEX_PROVIDER_ID];
	if (runtime === "openclaw") {
		if (configuredOpenAIAuthOrderStartsWithCodexProfile(params.config)) return [OPENAI_CODEX_PROVIDER_ID, OPENAI_PROVIDER_ID];
		return [OPENAI_PROVIDER_ID, OPENAI_CODEX_PROVIDER_ID];
	}
	return [params.provider];
}
function normalizeExplicitRuntimePin(value) {
	const runtime = normalizeOptionalAgentRuntimeId(value);
	return runtime === "auto" || runtime === "default" ? void 0 : runtime;
}
function resolveOpenAIRuntimeProvider(params) {
	return shouldRouteOpenAIThroughCodexAuthProvider(params) ? OPENAI_CODEX_PROVIDER_ID : params.provider;
}
function resolveSelectedOpenAIRuntimeProvider(params) {
	if (shouldRouteOpenAIThroughCodexAuthProvider(params)) return OPENAI_CODEX_PROVIDER_ID;
	const runtime = normalizeOptionalAgentRuntimeId(params.agentHarnessId ?? params.harnessRuntime) ?? "openclaw";
	if (!isOpenAIProvider(params.provider)) return params.provider;
	if (runtime === "codex") return OPENAI_CODEX_PROVIDER_ID;
	return runtime === "openclaw" && !params.authProfileId?.trim() && configuredOpenAIAuthOrderStartsWithCodexProfile(params.config) ? OPENAI_CODEX_PROVIDER_ID : params.provider;
}
function resolveContextConfigProviderForRuntime(params) {
	const provider = normalizeProviderId(params.provider);
	const runtimeId = normalizeOptionalAgentRuntimeId(params.runtimeId) ?? "openclaw";
	if (provider === "openai" && runtimeId === "codex") return OPENAI_CODEX_PROVIDER_ID;
	return params.provider;
}
//#endregion
export { normalizeOptionalAgentRuntimeId as _, listOpenAIAuthProfileProvidersForAgentRuntime as a, parseModelRefProvider as c, resolveSelectedOpenAIRuntimeProvider as d, shouldRouteOpenAIThroughCodexAuthProvider as f, normalizeEmbeddedAgentRuntime as g, isDefaultAgentRuntimeId as h, isOpenAIProvider as i, resolveContextConfigProviderForRuntime as l, OPENCLAW_AGENT_RUNTIME_ID as m, OPENAI_PROVIDER_ID as n, modelSelectionShouldEnsureCodexPlugin as o, AUTO_AGENT_RUNTIME_ID as p, isOpenAICodexProvider as r, openAIProviderUsesCodexRuntimeByDefault as s, OPENAI_CODEX_PROVIDER_ID as t, resolveOpenAIRuntimeProvider as u, resolveEmbeddedAgentRuntime as v };
