// openclaw/dist/path-prepend-C1I3sJQT.js
import { l as normalizeStringEntries, p as normalizeUniqueStringEntries } from "./string-normalization-B8G0vlWE.js";
import path from "node:path";
//#region src/infra/path-prepend.ts
/**
* Find the actual key used for PATH in the env object.
* On Windows, `process.env` stores it as `Path` (not `PATH`),
* and after copying to a plain object the original casing is preserved.
*/
function findPathKey(env) {
	if ("PATH" in env) return "PATH";
	for (const key of Object.keys(env)) if (key.toUpperCase() === "PATH") return key;
	return "PATH";
}
function normalizePathPrepend(entries) {
	if (!Array.isArray(entries)) return [];
	const seen = /* @__PURE__ */ new Set();
	const normalized = [];
	for (const entry of entries) {
		if (typeof entry !== "string") continue;
		const trimmed = entry.trim();
		if (!trimmed || seen.has(trimmed)) continue;
		seen.add(trimmed);
		normalized.push(trimmed);
	}
	return normalized;
}
function mergePathPrepend(existing, prepend) {
	if (prepend.length === 0) return existing;
	return normalizeUniqueStringEntries([...prepend, ...(existing ?? "").split(path.delimiter)]).join(path.delimiter);
}
function removePathPrepend(existing, prepend) {
	if (!existing || prepend.length === 0) return existing;
	const prependEntries = new Set(normalizeStringEntries(prepend));
	return normalizeStringEntries((existing ?? "").split(path.delimiter)).filter((part) => !prependEntries.has(part)).join(path.delimiter);
}
function applyPathPrepend(env, prepend, options) {
	if (!Array.isArray(prepend) || prepend.length === 0) return;
	const pathKey = findPathKey(env);
	if (options?.requireExisting && !env[pathKey]) return;
	const merged = mergePathPrepend(env[pathKey], prepend);
	if (merged) env[pathKey] = merged;
}
//#endregion
export { removePathPrepend as a, normalizePathPrepend as i, findPathKey as n, mergePathPrepend as r, applyPathPrepend as t };
