// openclaw/dist/policy-CWPpOk4M.js
import { _ as normalizeOptionalAgentRuntimeId, p as AUTO_AGENT_RUNTIME_ID, r as isOpenAICodexProvider, s as openAIProviderUsesCodexRuntimeByDefault } from "./openai-codex-routing-Cc71DT_j.js";
import { t as resolveModelRuntimePolicy } from "./model-runtime-policy-ZCInBBPn.js";
//#region src/agents/harness/policy.ts
function resolveAgentHarnessPolicy(params) {
	const configured = resolveModelRuntimePolicy({
		config: params.config,
		provider: params.provider,
		modelId: params.modelId,
		agentId: params.agentId,
		sessionKey: params.sessionKey
	});
	const configuredRuntime = normalizeOptionalAgentRuntimeId(configured.policy?.id);
	const runtimeSource = configured.source ?? "implicit";
	const runtime = configuredRuntime && configuredRuntime !== "default" ? configuredRuntime : AUTO_AGENT_RUNTIME_ID;
	if (openAIProviderUsesCodexRuntimeByDefault({
		provider: params.provider,
		config: params.config
	})) {
		if (runtime === "auto") return {
			runtime: "codex",
			runtimeSource
		};
		return {
			runtime,
			runtimeSource
		};
	}
	if (isOpenAICodexProvider(params.provider)) {
		if (runtime === "auto") return {
			runtime: "codex",
			runtimeSource
		};
		return {
			runtime,
			runtimeSource
		};
	}
	return {
		runtime,
		runtimeSource
	};
}
//#endregion
export { resolveAgentHarnessPolicy as t };
