// openclaw/dist/runtime-api-CShYPGPx.js
import { i as formatErrorMessage } from "./errors-AAwmFBlx.js";
import "./error-runtime-Dn_nurEc.js";
import { randomUUID } from "node:crypto";
//#region extensions/workboard/src/types.ts
const WORKBOARD_STATUSES = [
	"backlog",
	"todo",
	"running",
	"review",
	"blocked",
	"done"
];
const WORKBOARD_PRIORITIES = [
	"low",
	"normal",
	"high",
	"urgent"
];
const WORKBOARD_EXECUTION_ENGINES = ["codex", "claude"];
const WORKBOARD_EXECUTION_MODES = ["autonomous", "manual"];
const WORKBOARD_EXECUTION_STATUSES = [
	"idle",
	"running",
	"review",
	"blocked",
	"done"
];
const WORKBOARD_EVENT_KINDS = [
	"created",
	"edited",
	"moved",
	"linked",
	"claimed",
	"heartbeat",
	"execution_updated",
	"attempt_started",
	"attempt_updated",
	"comment_added",
	"link_added",
	"proof_added",
	"artifact_added",
	"diagnostic",
	"notification",
	"archived",
	"unarchived",
	"stale"
];
const WORKBOARD_ATTEMPT_STATUSES = [
	"running",
	"succeeded",
	"failed",
	"blocked",
	"stopped"
];
const WORKBOARD_LINK_TYPES = [
	"blocks",
	"blocked_by",
	"relates_to"
];
const WORKBOARD_PROOF_STATUSES = [
	"passed",
	"failed",
	"skipped",
	"unknown"
];
const WORKBOARD_TEMPLATE_IDS = [
	"bugfix",
	"docs",
	"release",
	"pr_review",
	"plugin"
];
const WORKBOARD_DIAGNOSTIC_KINDS = [
	"stranded_ready",
	"running_without_heartbeat",
	"blocked_too_long",
	"repeated_failures",
	"missing_proof",
	"orphaned_session"
];
const WORKBOARD_DIAGNOSTIC_SEVERITIES = [
	"warning",
	"error",
	"critical"
];
const WORKBOARD_NOTIFICATION_KINDS = [
	"completed",
	"failed",
	"stale"
];
//#endregion
//#region extensions/workboard/src/store.ts
const POSITION_STEP = 1e3;
const MAX_CARDS = 2e3;
const MAX_CARD_METADATA_BYTES = 24 * 1024;
const DEFAULT_CLAIM_TTL_MS = 1800 * 1e3;
const READY_STRANDED_MS = 3600 * 1e3;
const RUNNING_HEARTBEAT_STALE_MS = 1200 * 1e3;
const BLOCKED_TOO_LONG_MS = 1440 * 60 * 1e3;
function normalizeOptionalString(value) {
	return typeof value === "string" && value.trim() ? value.trim() : void 0;
}
function normalizeTitle(value) {
	const title = normalizeOptionalString(value);
	if (!title) throw new Error("title is required.");
	if (title.length > 180) throw new Error("title must be 180 characters or fewer.");
	return title;
}
function normalizeNotes(value) {
	const notes = normalizeOptionalString(value);
	if (!notes) return;
	if (notes.length > 4e3) throw new Error("notes must be 4000 characters or fewer.");
	return notes;
}
function normalizeBoundedString(value, fallback, maxLength, fieldName) {
	const normalized = normalizeOptionalString(value);
	if (!normalized) return fallback;
	if (normalized.length > maxLength) throw new Error(`${fieldName} must be ${maxLength} characters or fewer.`);
	return normalized;
}
function normalizeStatus(value, fallback) {
	if (typeof value !== "string" || !value.trim()) return fallback;
	if (WORKBOARD_STATUSES.includes(value)) return value;
	throw new Error(`status must be one of: ${WORKBOARD_STATUSES.join(", ")}.`);
}
function normalizePriority(value, fallback) {
	if (typeof value !== "string" || !value.trim()) return fallback;
	if (WORKBOARD_PRIORITIES.includes(value)) return value;
	throw new Error(`priority must be one of: ${WORKBOARD_PRIORITIES.join(", ")}.`);
}
function normalizeLabels(value, fallback = []) {
	if (value == null) return fallback;
	const entries = typeof value === "string" ? value.split(",") : Array.isArray(value) ? value : void 0;
	if (!entries) throw new Error("labels must be an array or comma-separated string.");
	const labels = [];
	for (const entry of entries) {
		const label = normalizeOptionalString(entry);
		if (!label || labels.includes(label)) continue;
		if (label.length > 40) throw new Error("labels must be 40 characters or fewer.");
		labels.push(label);
		if (labels.length >= 12) break;
	}
	return labels;
}
function normalizePosition(value, fallback) {
	if (typeof value !== "number" || !Number.isFinite(value)) return fallback;
	return Math.max(0, Math.trunc(value));
}
function normalizeExecutionEngine(value, fallback) {
	if (typeof value === "string" && WORKBOARD_EXECUTION_ENGINES.includes(value)) return value;
	return fallback;
}
function normalizeExecutionMode(value, fallback) {
	if (typeof value === "string" && WORKBOARD_EXECUTION_MODES.includes(value)) return value;
	return fallback;
}
function normalizeExecutionStatus(value, fallback) {
	if (typeof value === "string" && WORKBOARD_EXECUTION_STATUSES.includes(value)) return value;
	return fallback;
}
function normalizeAttemptStatus(value, fallback) {
	if (typeof value === "string" && WORKBOARD_ATTEMPT_STATUSES.includes(value)) return value;
	return fallback;
}
function normalizeLinkType(value, fallback) {
	if (typeof value === "string" && WORKBOARD_LINK_TYPES.includes(value)) return value;
	return fallback;
}
function normalizeProofStatus(value, fallback) {
	if (typeof value === "string" && WORKBOARD_PROOF_STATUSES.includes(value)) return value;
	return fallback;
}
function normalizeTemplateId(value) {
	return typeof value === "string" && WORKBOARD_TEMPLATE_IDS.includes(value) ? value : void 0;
}
function normalizeTimestamp(value, fallback) {
	return typeof value === "number" && Number.isFinite(value) ? Math.max(0, Math.trunc(value)) : fallback;
}
function normalizeEvent(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id);
	const kind = WORKBOARD_EVENT_KINDS.includes(record.kind) ? record.kind : null;
	const at = normalizeTimestamp(record.at, 0);
	if (!id || !kind || !at) return null;
	const fromStatus = typeof record.fromStatus === "string" && WORKBOARD_STATUSES.includes(record.fromStatus) ? record.fromStatus : void 0;
	const toStatus = typeof record.toStatus === "string" && WORKBOARD_STATUSES.includes(record.toStatus) ? record.toStatus : void 0;
	const sessionKey = normalizeOptionalString(record.sessionKey);
	const runId = normalizeOptionalString(record.runId);
	return {
		id,
		kind,
		at,
		...fromStatus ? { fromStatus } : {},
		...toStatus ? { toStatus } : {},
		...sessionKey ? { sessionKey } : {},
		...runId ? { runId } : {}
	};
}
function normalizeEvents(value) {
	if (!Array.isArray(value)) return [];
	return value.map(normalizeEvent).filter((event) => event !== null).slice(-50);
}
function normalizeAttempt(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id);
	const startedAt = normalizeTimestamp(record.startedAt, 0);
	if (!id || !startedAt) return null;
	const endedAt = normalizeTimestamp(record.endedAt, 0);
	const sessionKey = normalizeOptionalString(record.sessionKey);
	const runId = normalizeOptionalString(record.runId);
	const error = normalizeBoundedString(record.error, void 0, 800, "attempt error");
	const model = normalizeBoundedString(record.model, void 0, 160, "attempt model");
	return {
		id,
		status: normalizeAttemptStatus(record.status, "running"),
		startedAt,
		...endedAt ? { endedAt } : {},
		...typeof record.engine === "string" && WORKBOARD_EXECUTION_ENGINES.includes(record.engine) ? { engine: record.engine } : {},
		...typeof record.mode === "string" && WORKBOARD_EXECUTION_MODES.includes(record.mode) ? { mode: record.mode } : {},
		...model ? { model } : {},
		...sessionKey ? { sessionKey } : {},
		...runId ? { runId } : {},
		...error ? { error } : {}
	};
}
function normalizeComment(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id);
	const body = normalizeBoundedString(record.body, void 0, 2e3, "comment body");
	const createdAt = normalizeTimestamp(record.createdAt, 0);
	if (!id || !body || !createdAt) return null;
	const updatedAt = normalizeTimestamp(record.updatedAt, 0);
	return {
		id,
		body,
		createdAt,
		...updatedAt ? { updatedAt } : {}
	};
}
function normalizeLink(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id);
	const createdAt = normalizeTimestamp(record.createdAt, 0);
	if (!id || !createdAt) return null;
	const targetCardId = normalizeBoundedString(record.targetCardId, void 0, 120, "link target");
	const title = normalizeBoundedString(record.title, void 0, 180, "link title");
	const url = normalizeBoundedString(record.url, void 0, 2e3, "link URL");
	if (!targetCardId && !url) return null;
	return {
		id,
		type: normalizeLinkType(record.type, "relates_to"),
		createdAt,
		...targetCardId ? { targetCardId } : {},
		...title ? { title } : {},
		...url ? { url } : {}
	};
}
function normalizeProof(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id);
	const createdAt = normalizeTimestamp(record.createdAt, 0);
	if (!id || !createdAt) return null;
	const label = normalizeBoundedString(record.label, void 0, 160, "proof label");
	const command = normalizeBoundedString(record.command, void 0, 1e3, "proof command");
	const url = normalizeBoundedString(record.url, void 0, 2e3, "proof URL");
	const note = normalizeBoundedString(record.note, void 0, 2e3, "proof note");
	return {
		id,
		status: normalizeProofStatus(record.status, "unknown"),
		createdAt,
		...label ? { label } : {},
		...command ? { command } : {},
		...url ? { url } : {},
		...note ? { note } : {}
	};
}
function normalizeArtifact(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id) ?? randomUUID();
	const createdAt = normalizeTimestamp(record.createdAt, Date.now());
	const label = normalizeBoundedString(record.label, void 0, 160, "artifact label");
	const url = normalizeBoundedString(record.url, void 0, 2e3, "artifact URL");
	const artifactPath = normalizeBoundedString(record.path, void 0, 2e3, "artifact path");
	const mimeType = normalizeBoundedString(record.mimeType, void 0, 160, "artifact MIME type");
	if (!url && !artifactPath) return null;
	return {
		id,
		createdAt,
		...label ? { label } : {},
		...url ? { url } : {},
		...artifactPath ? { path: artifactPath } : {},
		...mimeType ? { mimeType } : {}
	};
}
function normalizeClaim(value, fallback) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return fallback;
	const record = value;
	const ownerId = normalizeBoundedString(record.ownerId, fallback?.ownerId, 120, "claim owner");
	const token = normalizeBoundedString(record.token, fallback?.token, 160, "claim token");
	const claimedAt = normalizeTimestamp(record.claimedAt, fallback?.claimedAt ?? Date.now());
	const lastHeartbeatAt = normalizeTimestamp(record.lastHeartbeatAt, fallback?.lastHeartbeatAt ?? claimedAt);
	const expiresAt = normalizeTimestamp(record.expiresAt, fallback?.expiresAt ?? 0);
	if (!ownerId || !token || !claimedAt || !lastHeartbeatAt) return;
	return {
		ownerId,
		token,
		claimedAt,
		lastHeartbeatAt,
		...expiresAt ? { expiresAt } : {}
	};
}
function normalizeDiagnosticAction(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const kind = record.kind === "claim" || record.kind === "unblock" || record.kind === "reassign" || record.kind === "add_proof" || record.kind === "open_session" ? record.kind : void 0;
	const label = normalizeBoundedString(record.label, void 0, 120, "diagnostic action label");
	return kind && label ? {
		kind,
		label
	} : null;
}
function normalizeDiagnostic(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const kind = WORKBOARD_DIAGNOSTIC_KINDS.includes(record.kind) ? record.kind : void 0;
	const severity = WORKBOARD_DIAGNOSTIC_SEVERITIES.includes(record.severity) ? record.severity : "warning";
	const title = normalizeBoundedString(record.title, void 0, 160, "diagnostic title");
	const detail = normalizeBoundedString(record.detail, void 0, 800, "diagnostic detail");
	const firstSeenAt = normalizeTimestamp(record.firstSeenAt, Date.now());
	const lastSeenAt = normalizeTimestamp(record.lastSeenAt, firstSeenAt);
	if (!kind || !title || !detail) return null;
	return {
		kind,
		severity,
		title,
		detail,
		firstSeenAt,
		lastSeenAt,
		count: typeof record.count === "number" && Number.isFinite(record.count) ? Math.max(1, Math.trunc(record.count)) : 1,
		actions: Array.isArray(record.actions) ? record.actions.map(normalizeDiagnosticAction).filter((action) => action !== null).slice(0, 4) : []
	};
}
function normalizeNotification(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return null;
	const record = value;
	const id = normalizeOptionalString(record.id) ?? randomUUID();
	const kind = WORKBOARD_NOTIFICATION_KINDS.includes(record.kind) ? record.kind : void 0;
	const createdAt = normalizeTimestamp(record.createdAt, Date.now());
	const message = normalizeBoundedString(record.message, void 0, 240, "notification message");
	if (!kind || !message) return null;
	const sessionKey = normalizeBoundedString(record.sessionKey, void 0, 240, "session key");
	const runId = normalizeBoundedString(record.runId, void 0, 120, "run id");
	return {
		id,
		kind,
		createdAt,
		message,
		...sessionKey ? { sessionKey } : {},
		...runId ? { runId } : {}
	};
}
function normalizeProofInput(input, now) {
	const label = normalizeBoundedString(input.label, void 0, 160, "proof label");
	const command = normalizeBoundedString(input.command, void 0, 1e3, "proof command");
	const url = normalizeBoundedString(input.url, void 0, 2e3, "proof URL");
	const note = normalizeBoundedString(input.note, void 0, 2e3, "proof note");
	return {
		id: randomUUID(),
		status: normalizeProofStatus(input.status, "unknown"),
		createdAt: now,
		...label ? { label } : {},
		...command ? { command } : {},
		...url ? { url } : {},
		...note ? { note } : {}
	};
}
function normalizeMetadata(value, fallback = {}) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return trimMetadataToBudget(fallback);
	const record = value;
	const stale = record.stale && typeof record.stale === "object" && !Array.isArray(record.stale) ? record.stale : null;
	const hasArchivedAt = Object.hasOwn(record, "archivedAt");
	const hasStale = Object.hasOwn(record, "stale");
	return trimMetadataToBudget({
		attempts: Array.isArray(record.attempts) ? record.attempts.map(normalizeAttempt).filter((attempt) => attempt !== null).slice(-30) : fallback.attempts,
		comments: Array.isArray(record.comments) ? record.comments.map(normalizeComment).filter((comment) => comment !== null).slice(-50) : fallback.comments,
		links: Array.isArray(record.links) ? record.links.map(normalizeLink).filter((link) => link !== null).slice(-50) : fallback.links,
		proof: Array.isArray(record.proof) ? record.proof.map(normalizeProof).filter((proof) => proof !== null).slice(-40) : fallback.proof,
		artifacts: Array.isArray(record.artifacts) ? record.artifacts.map(normalizeArtifact).filter((artifact) => artifact !== null).slice(-40) : fallback.artifacts,
		claim: Object.hasOwn(record, "claim") ? record.claim ? normalizeClaim(record.claim, fallback.claim) : void 0 : fallback.claim,
		diagnostics: Array.isArray(record.diagnostics) ? record.diagnostics.map(normalizeDiagnostic).filter((diagnostic) => diagnostic !== null).slice(-12) : fallback.diagnostics,
		notifications: Array.isArray(record.notifications) ? record.notifications.map(normalizeNotification).filter((notification) => notification !== null).slice(-20) : fallback.notifications,
		templateId: normalizeTemplateId(record.templateId) ?? fallback.templateId,
		archivedAt: hasArchivedAt ? normalizeTimestamp(record.archivedAt, 0) || void 0 : fallback.archivedAt,
		stale: hasStale ? stale ? {
			detectedAt: normalizeTimestamp(stale.detectedAt, Date.now()),
			lastSessionUpdatedAt: normalizeTimestamp(stale.lastSessionUpdatedAt, 0) || void 0,
			reason: normalizeBoundedString(stale.reason, fallback.stale?.reason, 240, "stale reason") ?? "Session has not reported recent activity."
		} : void 0 : fallback.stale,
		failureCount: typeof record.failureCount === "number" && Number.isFinite(record.failureCount) ? Math.max(0, Math.trunc(record.failureCount)) : fallback.failureCount
	});
}
function normalizeExecution(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) return;
	const record = value;
	const now = Date.now();
	const model = normalizeOptionalString(record.model);
	const id = normalizeOptionalString(record.id) ?? randomUUID();
	if (!model) return;
	const startedAt = normalizeTimestamp(record.startedAt, now);
	const updatedAt = normalizeTimestamp(record.updatedAt, startedAt);
	const sessionKey = normalizeOptionalString(record.sessionKey);
	const runId = normalizeOptionalString(record.runId);
	return {
		id,
		kind: "agent-session",
		engine: normalizeExecutionEngine(record.engine, "codex"),
		mode: normalizeExecutionMode(record.mode, "autonomous"),
		status: normalizeExecutionStatus(record.status, "idle"),
		model,
		startedAt,
		updatedAt,
		...sessionKey ? { sessionKey } : {},
		...runId ? { runId } : {}
	};
}
function syncExecutionSessionKey(execution, sessionKey) {
	if (!execution) return;
	return removeUndefinedExecutionFields({
		...execution,
		sessionKey,
		updatedAt: Date.now()
	});
}
function removeUndefinedExecutionFields(execution) {
	const next = { ...execution };
	if (next.sessionKey === void 0) delete next.sessionKey;
	if (next.runId === void 0) delete next.runId;
	return next;
}
function removeUndefinedMetadataFields(metadata) {
	const next = { ...metadata };
	for (const key of [
		"attempts",
		"comments",
		"links",
		"proof",
		"artifacts",
		"claim",
		"diagnostics",
		"notifications",
		"templateId",
		"archivedAt",
		"stale",
		"failureCount"
	]) {
		const value = next[key];
		if (value === void 0 || Array.isArray(value) && value.length === 0 || typeof value === "number" && value === 0 && key === "failureCount") delete next[key];
	}
	return next;
}
function clearDiagnostics(metadata, kinds) {
	if (!metadata?.diagnostics) return metadata ?? {};
	return {
		...metadata,
		diagnostics: metadata.diagnostics.filter((entry) => !kinds.includes(entry.kind))
	};
}
function metadataIsEmpty(metadata) {
	return !metadata || Object.keys(metadata).length === 0;
}
function metadataByteSize(metadata) {
	return Buffer.byteLength(JSON.stringify(metadata), "utf8");
}
function dropFirst(items) {
	if (!items?.length) return;
	const next = items.slice(1);
	return next.length ? next : void 0;
}
function trimMetadataToBudget(metadata) {
	let next = removeUndefinedMetadataFields(metadata);
	while (metadataByteSize(next) > MAX_CARD_METADATA_BYTES) {
		const currentSize = metadataByteSize(next);
		if (next.attempts?.length) next = removeUndefinedMetadataFields({
			...next,
			attempts: dropFirst(next.attempts)
		});
		else if (next.diagnostics?.length) next = removeUndefinedMetadataFields({
			...next,
			diagnostics: dropFirst(next.diagnostics)
		});
		else if (next.notifications?.length) next = removeUndefinedMetadataFields({
			...next,
			notifications: dropFirst(next.notifications)
		});
		else if (next.proof?.length) next = removeUndefinedMetadataFields({
			...next,
			proof: dropFirst(next.proof)
		});
		else if (next.artifacts?.length) next = removeUndefinedMetadataFields({
			...next,
			artifacts: dropFirst(next.artifacts)
		});
		else if (next.links?.length) next = removeUndefinedMetadataFields({
			...next,
			links: dropFirst(next.links)
		});
		else if (next.comments?.length) next = removeUndefinedMetadataFields({
			...next,
			comments: dropFirst(next.comments)
		});
		if (metadataByteSize(next) >= currentSize) break;
	}
	return next;
}
function compareCards(left, right) {
	if (left.status !== right.status) return WORKBOARD_STATUSES.indexOf(left.status) - WORKBOARD_STATUSES.indexOf(right.status);
	if (left.position !== right.position) return left.position - right.position;
	return left.createdAt - right.createdAt;
}
function cardSessionKey(card) {
	return card.sessionKey ?? card.execution?.sessionKey;
}
function cardRunId(card) {
	return card.runId ?? card.execution?.runId;
}
function executionAttemptStatus(execution) {
	if (execution.status === "running") return "running";
	if (execution.status === "blocked") return "blocked";
	if (execution.status === "done" || execution.status === "review") return "succeeded";
	return "stopped";
}
function syncExecutionAttemptMetadata(metadata, execution, now) {
	if (!execution) return metadata;
	const attemptStatus = executionAttemptStatus(execution);
	const attempts = [...metadata.attempts ?? []];
	const key = execution.runId ?? execution.sessionKey ?? execution.id;
	const existingIndex = attempts.findIndex((attempt) => execution.runId && attempt.runId === execution.runId || !execution.runId && attempt.id === key);
	const existingAttempt = existingIndex >= 0 ? attempts[existingIndex] : void 0;
	const nextAttempt = {
		id: existingAttempt?.id ?? key,
		status: attemptStatus,
		startedAt: existingAttempt?.startedAt ?? execution.startedAt,
		engine: execution.engine,
		mode: execution.mode,
		model: execution.model,
		...execution.sessionKey ? { sessionKey: execution.sessionKey } : {},
		...execution.runId ? { runId: execution.runId } : {},
		...attemptStatus !== "running" && { endedAt: execution.updatedAt || now }
	};
	if (existingIndex >= 0) attempts[existingIndex] = nextAttempt;
	else attempts.push(nextAttempt);
	const previousFailed = existingAttempt?.status === "blocked" || existingAttempt?.status === "failed";
	const failureCount = attemptStatus === "blocked" || attemptStatus === "failed" ? previousFailed ? metadata.failureCount : (metadata.failureCount ?? 0) + 1 : attemptStatus === "succeeded" ? 0 : metadata.failureCount;
	return removeUndefinedMetadataFields({
		...metadata,
		attempts: attempts.slice(-30),
		failureCount
	});
}
function appendEvent(card, event, at = Date.now()) {
	return [...normalizeEvents(card.events), {
		id: randomUUID(),
		at,
		...event
	}].slice(-50);
}
function latestMetadataIdChanged(existing, next) {
	const latestId = next?.at(-1)?.id;
	return Boolean(latestId && latestId !== existing?.at(-1)?.id);
}
function updateEvent(existing, next) {
	if (existing.status !== next.status || existing.position !== next.position) return {
		kind: "moved",
		fromStatus: existing.status,
		toStatus: next.status
	};
	if (cardSessionKey(existing) !== cardSessionKey(next)) return {
		kind: "linked",
		...cardSessionKey(next) ? { sessionKey: cardSessionKey(next) } : {}
	};
	if (existing.metadata?.claim?.token !== next.metadata?.claim?.token) return { kind: "claimed" };
	if (existing.metadata?.claim?.lastHeartbeatAt !== next.metadata?.claim?.lastHeartbeatAt) return { kind: "heartbeat" };
	if (existing.execution?.status !== next.execution?.status || existing.execution?.engine !== next.execution?.engine || cardRunId(existing) !== cardRunId(next)) {
		const existingAttempts = existing.metadata?.attempts ?? [];
		const nextAttempts = next.metadata?.attempts ?? [];
		const latestAttempt = nextAttempts.at(-1);
		if (nextAttempts.length > existingAttempts.length) return {
			kind: "attempt_started",
			...latestAttempt?.sessionKey ? { sessionKey: latestAttempt.sessionKey } : {},
			...latestAttempt?.runId ? { runId: latestAttempt.runId } : {}
		};
		const previousAttempt = latestAttempt ? existingAttempts.find((attempt) => attempt.id === latestAttempt.id) : void 0;
		if (latestAttempt && previousAttempt?.status !== latestAttempt.status) return {
			kind: "attempt_updated",
			...latestAttempt.sessionKey ? { sessionKey: latestAttempt.sessionKey } : {},
			...latestAttempt.runId ? { runId: latestAttempt.runId } : {}
		};
		return {
			kind: "execution_updated",
			...cardSessionKey(next) ? { sessionKey: cardSessionKey(next) } : {},
			...cardRunId(next) ? { runId: cardRunId(next) } : {}
		};
	}
	if ((existing.metadata?.comments?.length ?? 0) !== (next.metadata?.comments?.length ?? 0) || latestMetadataIdChanged(existing.metadata?.comments, next.metadata?.comments)) return { kind: "comment_added" };
	if ((existing.metadata?.links?.length ?? 0) !== (next.metadata?.links?.length ?? 0) || latestMetadataIdChanged(existing.metadata?.links, next.metadata?.links)) return { kind: "link_added" };
	if ((existing.metadata?.proof?.length ?? 0) !== (next.metadata?.proof?.length ?? 0) || latestMetadataIdChanged(existing.metadata?.proof, next.metadata?.proof)) return { kind: "proof_added" };
	if ((existing.metadata?.artifacts?.length ?? 0) !== (next.metadata?.artifacts?.length ?? 0) || latestMetadataIdChanged(existing.metadata?.artifacts, next.metadata?.artifacts)) return { kind: "artifact_added" };
	if ((existing.metadata?.diagnostics?.length ?? 0) !== (next.metadata?.diagnostics?.length ?? 0)) return { kind: "diagnostic" };
	if ((existing.metadata?.notifications?.length ?? 0) !== (next.metadata?.notifications?.length ?? 0) || latestMetadataIdChanged(existing.metadata?.notifications, next.metadata?.notifications)) return { kind: "notification" };
	if (!existing.metadata?.archivedAt && next.metadata?.archivedAt) return { kind: "archived" };
	if (existing.metadata?.archivedAt && !next.metadata?.archivedAt) return { kind: "unarchived" };
	if (!existing.metadata?.stale && next.metadata?.stale) return { kind: "stale" };
	return { kind: "edited" };
}
function removeUndefinedCardFields(card) {
	const next = { ...card };
	for (const key of [
		"notes",
		"agentId",
		"sessionKey",
		"runId",
		"taskId",
		"sourceUrl",
		"execution",
		"startedAt",
		"completedAt",
		"metadata"
	]) if (next[key] === void 0) delete next[key];
	if (metadataIsEmpty(next.metadata)) delete next.metadata;
	return next;
}
function assertCanMutateClaimedCard(card, scope) {
	if (!scope) return;
	const claim = card.metadata?.claim;
	if (!claim) return;
	const ownerId = normalizeOptionalString(scope.ownerId);
	const token = normalizeOptionalString(scope.token);
	if (claim.ownerId === ownerId || token && claim.token === token) return;
	throw new Error(`card is claimed by ${claim.ownerId}.`);
}
function diagnostic(params, now) {
	return {
		...params,
		firstSeenAt: now,
		lastSeenAt: now,
		count: 1
	};
}
function mergeDiagnostics(previous, next) {
	const byKind = new Map(previous?.map((entry) => [entry.kind, entry]));
	return next.map((entry) => {
		const prior = byKind.get(entry.kind);
		return prior ? {
			...entry,
			firstSeenAt: prior.firstSeenAt,
			count: prior.count + 1
		} : entry;
	});
}
function computeCardDiagnostics(card, now) {
	const diagnostics = [];
	const lastHeartbeatAt = (card.metadata?.claim)?.lastHeartbeatAt ?? card.execution?.updatedAt ?? card.updatedAt;
	if ((card.status === "todo" || card.status === "backlog") && card.agentId && now - card.updatedAt > READY_STRANDED_MS) diagnostics.push(diagnostic({
		kind: "stranded_ready",
		severity: "warning",
		title: "Assigned card is waiting",
		detail: "The card has an assigned agent but has not been claimed recently.",
		actions: [{
			kind: "claim",
			label: "Claim card"
		}]
	}, now));
	if (card.status === "running" && now - lastHeartbeatAt > RUNNING_HEARTBEAT_STALE_MS) diagnostics.push(diagnostic({
		kind: "running_without_heartbeat",
		severity: "error",
		title: "Running card has no recent heartbeat",
		detail: "The linked run or claim has not reported recent activity.",
		actions: [{
			kind: "open_session",
			label: "Open session"
		}, {
			kind: "reassign",
			label: "Reassign card"
		}]
	}, now));
	if (card.status === "blocked" && now - card.updatedAt > BLOCKED_TOO_LONG_MS) diagnostics.push(diagnostic({
		kind: "blocked_too_long",
		severity: "warning",
		title: "Blocked card needs attention",
		detail: "The card has been blocked for more than a day.",
		actions: [{
			kind: "unblock",
			label: "Move to todo"
		}]
	}, now));
	if ((card.metadata?.failureCount ?? 0) >= 2) diagnostics.push(diagnostic({
		kind: "repeated_failures",
		severity: "error",
		title: "Repeated run failures",
		detail: "Multiple attempts failed or blocked on this card.",
		actions: [{
			kind: "reassign",
			label: "Reassign card"
		}]
	}, now));
	if (card.status === "done" && !(card.metadata?.proof?.length || card.metadata?.artifacts?.length)) diagnostics.push(diagnostic({
		kind: "missing_proof",
		severity: "warning",
		title: "Done card has no proof",
		detail: "The card is marked done without proof or an attached artifact.",
		actions: [{
			kind: "add_proof",
			label: "Add proof"
		}]
	}, now));
	if (card.sessionKey && !card.execution && card.status === "running") diagnostics.push(diagnostic({
		kind: "orphaned_session",
		severity: "warning",
		title: "Running card has only a loose session link",
		detail: "The card is running but has no execution record for lifecycle handoff.",
		actions: [{
			kind: "open_session",
			label: "Open session"
		}]
	}, now));
	return diagnostics;
}
function capText(value, max) {
	if (!value) return;
	return value.length <= max ? value : `${value.slice(0, Math.max(0, max - 1))}…`;
}
function buildWorkerContext(card) {
	const lines = [
		`# Workboard card ${card.id}`,
		`Title: ${card.title}`,
		`Status: ${card.status}`,
		`Priority: ${card.priority}`,
		`Agent: ${card.agentId ?? "(default)"}`
	];
	if (card.notes) lines.push("", "## Notes", capText(card.notes, 4e3) ?? "");
	const attempts = card.metadata?.attempts?.slice(-8) ?? [];
	if (attempts.length) {
		lines.push("", "## Recent attempts");
		for (const attempt of attempts) lines.push(`- ${attempt.status} ${attempt.model ?? ""} ${attempt.error ? `error=${capText(attempt.error, 240)}` : ""}`.trim());
	}
	const comments = card.metadata?.comments?.slice(-12) ?? [];
	if (comments.length) {
		lines.push("", "## Recent comments");
		for (const comment of comments) lines.push(`- ${capText(comment.body, 400)}`);
	}
	const proof = card.metadata?.proof?.slice(-8) ?? [];
	if (proof.length) {
		lines.push("", "## Proof");
		for (const entry of proof) lines.push(`- ${entry.status}: ${capText(entry.label ?? entry.command ?? entry.url ?? entry.note, 400)}`);
	}
	const artifacts = card.metadata?.artifacts?.slice(-8) ?? [];
	if (artifacts.length) {
		lines.push("", "## Artifacts");
		for (const artifact of artifacts) lines.push(`- ${capText(artifact.label ?? artifact.url ?? artifact.path, 400)}`);
	}
	const links = card.metadata?.links?.slice(-8) ?? [];
	if (links.length) {
		lines.push("", "## Links");
		for (const link of links) lines.push(`- ${link.type}: ${link.title ?? link.url ?? link.targetCardId ?? ""}`);
	}
	const diagnostics = computeCardDiagnostics(card, Date.now());
	if (diagnostics.length) {
		lines.push("", "## Active diagnostics");
		for (const entry of diagnostics) lines.push(`- ${entry.severity}: ${entry.title}`);
	}
	return lines.join("\n");
}
var WorkboardStore = class WorkboardStore {
	constructor(store) {
		this.store = store;
		this.mutationQueue = Promise.resolve();
	}
	async enqueueMutation(run) {
		const result = this.mutationQueue.then(run, run);
		this.mutationQueue = result.then(() => void 0, () => void 0);
		return await result;
	}
	async updateMetadata(id, mutate) {
		return await this.enqueueMutation(async () => {
			const existing = await this.get(id);
			if (!existing) throw new Error(`card not found: ${id}`);
			return await this.updateCard(id, { metadata: mutate(existing) });
		});
	}
	async list() {
		return (await this.store.entries()).map((entry) => entry.value).filter((entry) => entry?.version === 1 && Boolean(entry.card?.id)).map((entry) => entry.card).toSorted(compareCards);
	}
	async get(id) {
		const entry = await this.store.lookup(id.trim());
		return entry?.version === 1 ? entry.card : void 0;
	}
	async create(input) {
		const now = Date.now();
		const status = normalizeStatus(input.status, "todo");
		const cards = await this.list();
		const normalizedPosition = normalizePosition(input.position, NaN);
		const position = Number.isFinite(normalizedPosition) ? normalizedPosition : Math.max(0, ...cards.filter((card) => card.status === status).map((card) => card.position)) + POSITION_STEP;
		const notes = normalizeNotes(input.notes);
		const agentId = normalizeOptionalString(input.agentId);
		const sessionKey = normalizeOptionalString(input.sessionKey);
		const runId = normalizeOptionalString(input.runId);
		const taskId = normalizeOptionalString(input.taskId);
		const sourceUrl = normalizeOptionalString(input.sourceUrl);
		const execution = normalizeExecution(input.execution);
		const syncedMetadata = trimMetadataToBudget(syncExecutionAttemptMetadata(normalizeMetadata(input.metadata, { templateId: normalizeTemplateId(input.templateId) }), execution, now));
		const card = {
			id: randomUUID(),
			title: normalizeTitle(input.title),
			status,
			priority: normalizePriority(input.priority, "normal"),
			labels: normalizeLabels(input.labels),
			position,
			createdAt: now,
			updatedAt: now,
			events: [{
				id: randomUUID(),
				kind: "created",
				at: now,
				toStatus: status,
				...sessionKey ? { sessionKey } : {},
				...runId ? { runId } : {}
			}],
			...notes ? { notes } : {},
			...agentId ? { agentId } : {},
			...sessionKey ? { sessionKey } : {},
			...runId ? { runId } : {},
			...taskId ? { taskId } : {},
			...sourceUrl ? { sourceUrl } : {},
			...execution ? { execution } : {},
			...!metadataIsEmpty(syncedMetadata) ? { metadata: syncedMetadata } : {}
		};
		await this.store.register(card.id, {
			version: 1,
			card
		});
		return card;
	}
	async update(id, patch) {
		return await this.enqueueMutation(async () => await this.updateCard(id, patch));
	}
	async updateCard(id, patch) {
		const existing = await this.get(id);
		if (!existing) throw new Error(`card not found: ${id}`);
		const status = normalizeStatus(patch.status, existing.status);
		const now = Date.now();
		const completedAt = status === "done" ? existing.completedAt ?? now : void 0;
		const startedAt = status === "running" ? existing.startedAt ?? now : existing.startedAt;
		const sessionKey = patch.sessionKey === void 0 ? existing.sessionKey : normalizeOptionalString(patch.sessionKey);
		const execution = patch.execution === void 0 ? patch.sessionKey === void 0 ? existing.execution : syncExecutionSessionKey(existing.execution, sessionKey) : normalizeExecution(patch.execution);
		const metadata = normalizeMetadata(patch.metadata, existing.metadata);
		const next = removeUndefinedCardFields({
			...existing,
			title: patch.title === void 0 ? existing.title : normalizeTitle(patch.title),
			notes: patch.notes === void 0 ? existing.notes : normalizeNotes(patch.notes),
			status,
			priority: patch.priority === void 0 ? existing.priority : normalizePriority(patch.priority, existing.priority),
			labels: patch.labels === void 0 ? existing.labels : normalizeLabels(patch.labels),
			agentId: patch.agentId === void 0 ? existing.agentId : normalizeOptionalString(patch.agentId),
			sessionKey,
			runId: patch.runId === void 0 ? existing.runId : normalizeOptionalString(patch.runId),
			taskId: patch.taskId === void 0 ? existing.taskId : normalizeOptionalString(patch.taskId),
			sourceUrl: patch.sourceUrl === void 0 ? existing.sourceUrl : normalizeOptionalString(patch.sourceUrl),
			execution,
			metadata: patch.templateId === void 0 ? metadata : {
				...metadata,
				templateId: normalizeTemplateId(patch.templateId)
			},
			position: patch.position === void 0 ? existing.position : normalizePosition(patch.position, existing.position),
			updatedAt: now,
			...startedAt ? { startedAt } : {},
			...completedAt ? { completedAt } : {}
		});
		next.metadata = trimMetadataToBudget(syncExecutionAttemptMetadata(next.metadata ?? {}, execution, now));
		next.events = appendEvent(next, updateEvent(existing, next), now);
		if (status !== "done") delete next.completedAt;
		if (metadataIsEmpty(next.metadata)) delete next.metadata;
		await this.store.register(next.id, {
			version: 1,
			card: next
		});
		return next;
	}
	async move(id, status, position) {
		return await this.update(id, {
			status,
			position
		});
	}
	async delete(id) {
		return { deleted: await this.store.delete(id.trim()) };
	}
	async addComment(id, input, scope) {
		const now = Date.now();
		const body = normalizeBoundedString(input.body, void 0, 2e3, "comment body");
		if (!body) throw new Error("comment body is required.");
		const comment = {
			id: randomUUID(),
			body,
			createdAt: now
		};
		return await this.updateMetadata(id, (existing) => {
			assertCanMutateClaimedCard(existing, scope);
			return {
				...existing.metadata,
				comments: [...existing.metadata?.comments ?? [], comment].slice(-50)
			};
		});
	}
	async addLink(id, input) {
		const now = Date.now();
		const targetCardId = normalizeBoundedString(input.targetCardId, void 0, 120, "link target");
		const url = normalizeBoundedString(input.url, void 0, 2e3, "link URL");
		const title = normalizeBoundedString(input.title, void 0, 180, "link title");
		if (!targetCardId && !url) throw new Error("link targetCardId or url is required.");
		const link = {
			id: randomUUID(),
			type: normalizeLinkType(input.type, "relates_to"),
			createdAt: now,
			...targetCardId ? { targetCardId } : {},
			...title ? { title } : {},
			...url ? { url } : {}
		};
		return await this.updateMetadata(id, (existing) => ({
			...existing.metadata,
			links: [...existing.metadata?.links ?? [], link].slice(-50)
		}));
	}
	async addProof(id, input, scope) {
		const proof = normalizeProofInput(input, Date.now());
		return await this.updateMetadata(id, (existing) => {
			assertCanMutateClaimedCard(existing, scope);
			const metadata = clearDiagnostics(existing.metadata, ["missing_proof"]);
			return {
				...metadata,
				proof: [...metadata.proof ?? [], proof].slice(-40)
			};
		});
	}
	async addProofWithArtifact(id, proofInput, artifactInput, scope) {
		const now = Date.now();
		const proof = normalizeProofInput(proofInput, now);
		const artifact = normalizeArtifact({
			...artifactInput,
			createdAt: now
		});
		if (!artifact) throw new Error("artifact url or path is required.");
		return await this.updateMetadata(id, (existing) => {
			assertCanMutateClaimedCard(existing, scope);
			const metadata = clearDiagnostics(existing.metadata, ["missing_proof"]);
			return {
				...metadata,
				proof: [...metadata.proof ?? [], proof].slice(-40),
				artifacts: [...metadata.artifacts ?? [], artifact].slice(-40)
			};
		});
	}
	async addArtifact(id, input, scope) {
		const artifact = normalizeArtifact({
			...input,
			createdAt: Date.now()
		});
		if (!artifact) throw new Error("artifact url or path is required.");
		return await this.updateMetadata(id, (existing) => {
			assertCanMutateClaimedCard(existing, scope);
			const metadata = clearDiagnostics(existing.metadata, ["missing_proof"]);
			return {
				...metadata,
				artifacts: [...metadata.artifacts ?? [], artifact].slice(-40)
			};
		});
	}
	async claim(id, input) {
		const now = Date.now();
		const ownerId = normalizeBoundedString(input.ownerId, void 0, 120, "claim owner");
		if (!ownerId) throw new Error("claim ownerId is required.");
		const ttlSeconds = typeof input.ttlSeconds === "number" && Number.isFinite(input.ttlSeconds) ? Math.max(1, Math.trunc(input.ttlSeconds)) : void 0;
		const token = normalizeBoundedString(input.token, void 0, 160, "claim token") ?? randomUUID();
		const expiresAt = now + (ttlSeconds ? ttlSeconds * 1e3 : DEFAULT_CLAIM_TTL_MS);
		const card = await this.updateMetadata(id, (existing) => {
			const existingClaim = existing.metadata?.claim;
			if (existingClaim && existingClaim.expiresAt && existingClaim.expiresAt > now) throw new Error(`card already claimed by ${existingClaim.ownerId}.`);
			return {
				...clearDiagnostics(existing.metadata, ["stranded_ready"]),
				claim: {
					ownerId,
					token,
					claimedAt: now,
					lastHeartbeatAt: now,
					expiresAt
				}
			};
		});
		return {
			card: await this.update(card.id, {
				status: card.status === "backlog" || card.status === "todo" ? "running" : card.status,
				agentId: card.agentId ?? ownerId
			}),
			token
		};
	}
	async heartbeat(id, input) {
		const note = normalizeBoundedString(input.note, void 0, 400, "heartbeat note");
		return await this.updateMetadata(id, (existing) => {
			const claim = existing.metadata?.claim;
			if (!claim) throw new Error("card is not claimed.");
			const now = Math.max(Date.now(), claim.lastHeartbeatAt + 1);
			const token = normalizeOptionalString(input.token);
			const ownerId = normalizeOptionalString(input.ownerId);
			if (token && token !== claim.token) throw new Error("claim token does not match.");
			if (!token && ownerId && ownerId !== claim.ownerId) throw new Error("claim owner does not match.");
			const nextClaim = {
				...claim,
				lastHeartbeatAt: now,
				expiresAt: claim.expiresAt ? now + Math.max(1, claim.expiresAt > claim.claimedAt ? claim.expiresAt - claim.lastHeartbeatAt : DEFAULT_CLAIM_TTL_MS) : void 0
			};
			const metadata = clearDiagnostics(existing.metadata, ["running_without_heartbeat"]);
			return {
				...metadata,
				claim: removeUndefinedMetadataFields({ claim: nextClaim }).claim,
				comments: note ? [...metadata.comments ?? [], {
					id: randomUUID(),
					body: note,
					createdAt: now
				}].slice(-50) : metadata.comments
			};
		});
	}
	async releaseClaim(id, input = {}) {
		return await this.enqueueMutation(async () => {
			const existing = await this.get(id);
			if (!existing) throw new Error(`card not found: ${id}`);
			const status = input.status === void 0 ? existing.status : normalizeStatus(input.status, existing.status);
			const claim = existing.metadata?.claim;
			if (claim) {
				const token = normalizeOptionalString(input.token);
				const ownerId = normalizeOptionalString(input.ownerId);
				if (token && token !== claim.token) throw new Error("claim token does not match.");
				if (!token && ownerId && ownerId !== claim.ownerId) throw new Error("claim owner does not match.");
			}
			return await this.updateCard(id, {
				status,
				metadata: {
					...existing.metadata,
					claim: void 0
				}
			});
		});
	}
	async unblock(id, scope) {
		return await this.enqueueMutation(async () => {
			const existing = await this.get(id);
			if (!existing) throw new Error(`card not found: ${id}`);
			assertCanMutateClaimedCard(existing, scope);
			const metadata = clearDiagnostics(existing.metadata, ["blocked_too_long"]);
			return await this.updateCard(id, {
				status: "todo",
				metadata: {
					...metadata,
					stale: null
				}
			});
		});
	}
	async bulkUpdate(input) {
		const ids = Array.isArray(input.ids) ? input.ids.filter((id) => typeof id === "string" && id.trim() !== "") : [];
		if (ids.length === 0) throw new Error("ids are required.");
		const patch = input.patch && typeof input.patch === "object" && !Array.isArray(input.patch) ? input.patch : {};
		const cards = [];
		for (const id of ids) {
			const updated = input.archived === void 0 ? await this.update(id, patch) : await this.archive(id, input.archived);
			cards.push(updated);
		}
		return { cards };
	}
	async archive(id, archived) {
		const shouldArchive = archived !== false;
		return await this.updateMetadata(id, (existing) => ({
			...existing.metadata,
			archivedAt: shouldArchive ? Date.now() : 0
		}));
	}
	async exportCards() {
		return {
			cards: await this.list(),
			exportedAt: Date.now()
		};
	}
	async diagnostics(now = Date.now()) {
		const rows = (await this.list()).flatMap((card) => {
			const diagnostics = computeCardDiagnostics(card, now);
			return diagnostics.length ? [{
				card,
				diagnostics
			}] : [];
		});
		return {
			diagnostics: rows,
			count: rows.reduce((total, row) => total + row.diagnostics.length, 0)
		};
	}
	async refreshDiagnostics(now = Date.now()) {
		return await this.enqueueMutation(async () => {
			const cards = await this.list();
			const rows = [];
			for (const card of cards) {
				const latest = await this.get(card.id);
				if (!latest) continue;
				const diagnostics = mergeDiagnostics(latest.metadata?.diagnostics, computeCardDiagnostics(latest, now));
				if (diagnostics.length === 0 && !latest.metadata?.diagnostics?.length) continue;
				const metadata = trimMetadataToBudget({
					...latest.metadata,
					diagnostics
				});
				const next = removeUndefinedCardFields({
					...latest,
					metadata: metadataIsEmpty(metadata) ? void 0 : metadata
				});
				await this.store.register(next.id, {
					version: 1,
					card: next
				});
				if (diagnostics.length > 0) rows.push({
					card: next,
					diagnostics
				});
			}
			return {
				diagnostics: rows,
				count: rows.reduce((total, row) => total + row.diagnostics.length, 0)
			};
		});
	}
	async buildWorkerContext(id) {
		const card = await this.get(id);
		if (!card) throw new Error(`card not found: ${id}`);
		return buildWorkerContext(card);
	}
	static open(openKeyedStore) {
		return new WorkboardStore(openKeyedStore({
			namespace: "workboard.cards",
			maxEntries: MAX_CARDS
		}));
	}
};
//#endregion
//#region extensions/workboard/src/gateway.ts
const READ_SCOPE = "operator.read";
const WRITE_SCOPE = "operator.write";
function respondError(respond, error) {
	respond(false, void 0, {
		code: "workboard_error",
		message: formatErrorMessage(error)
	});
}
function readId(params) {
	const value = params.id;
	if (typeof value === "string" && value.trim()) return value.trim();
	throw new Error("id is required.");
}
function readPatch(params) {
	const patch = params.patch;
	if (patch && typeof patch === "object" && !Array.isArray(patch)) return patch;
	return params;
}
function redactClaimToken(card) {
	const claim = card.metadata?.claim;
	if (!claim) return card;
	return {
		...card,
		metadata: {
			...card.metadata,
			claim: {
				...claim,
				token: "[redacted]"
			}
		}
	};
}
function redactDiagnosticsRows(result) {
	return {
		...result,
		diagnostics: result.diagnostics.map((row) => ({
			...row,
			card: redactClaimToken(row.card)
		}))
	};
}
function registerWorkboardGatewayMethods(params) {
	const { api } = params;
	const store = params.store ?? WorkboardStore.open((options) => api.runtime.state.openKeyedStore(options));
	api.registerGatewayMethod("workboard.cards.list", async ({ respond }) => {
		try {
			respond(true, {
				cards: (await store.list()).map(redactClaimToken),
				statuses: WORKBOARD_STATUSES
			});
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: READ_SCOPE });
	api.registerGatewayMethod("workboard.cards.create", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.create(requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.update", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.update(readId(requestParams), readPatch(requestParams))) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.move", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.move(readId(requestParams), requestParams.status, requestParams.position)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.delete", async ({ params: requestParams, respond }) => {
		try {
			respond(true, await store.delete(readId(requestParams)));
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.comment", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.addComment(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.link", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.addLink(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.proof", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.addProof(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.artifact", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.addArtifact(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.claim", async ({ params: requestParams, respond }) => {
		try {
			const claimed = await store.claim(readId(requestParams), requestParams);
			respond(true, {
				...claimed,
				card: redactClaimToken(claimed.card)
			});
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.heartbeat", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.heartbeat(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.release", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.releaseClaim(readId(requestParams), requestParams)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.unblock", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.unblock(readId(requestParams))) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.bulk", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { cards: (await store.bulkUpdate(requestParams)).cards.map(redactClaimToken) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.diagnostics", async ({ respond }) => {
		try {
			respond(true, redactDiagnosticsRows(await store.diagnostics()));
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: READ_SCOPE });
	api.registerGatewayMethod("workboard.cards.diagnostics.refresh", async ({ respond }) => {
		try {
			respond(true, redactDiagnosticsRows(await store.refreshDiagnostics()));
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.archive", async ({ params: requestParams, respond }) => {
		try {
			respond(true, { card: redactClaimToken(await store.archive(readId(requestParams), requestParams.archived)) });
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: WRITE_SCOPE });
	api.registerGatewayMethod("workboard.cards.export", async ({ respond }) => {
		try {
			const exported = await store.exportCards();
			respond(true, {
				...exported,
				cards: exported.cards.map(redactClaimToken)
			});
		} catch (error) {
			respondError(respond, error);
		}
	}, { scope: READ_SCOPE });
}
//#endregion
export { WorkboardStore as n, registerWorkboardGatewayMethods as t };
