// openclaw/dist/tool-payload-BwXNuCsb.js
//#region src/plugin-sdk/tool-payload.ts
function isToolPayloadTextBlock(block) {
	return !!block && typeof block === "object" && block.type === "text" && typeof block.text === "string";
}
/**
* Extract the most useful payload from tool result-like objects shared across
* outbound core flows and bundled plugin helpers.
*/
function extractToolPayload(result) {
	if (!result) return;
	if (result.details !== void 0) return result.details;
	const text = (Array.isArray(result.content) ? result.content.find(isToolPayloadTextBlock) : void 0)?.text;
	if (!text) return result.content ?? result;
	try {
		return JSON.parse(text);
	} catch {
		return text;
	}
}
const DEFAULT_MAX_PLAIN_TEXT_TOOL_PAYLOAD_BYTES = 256e3;
const END_TOOL_REQUEST = "[END_TOOL_REQUEST]";
const HARMONY_CHANNEL_MARKER = "<|channel|>";
const HARMONY_MESSAGE_MARKER = "<|message|>";
const HARMONY_CALL_MARKER = "<|call|>";
const XMLISH_PARAMETER_CLOSE = "</parameter>";
function isToolNameChar(char) {
	return Boolean(char && /[A-Za-z0-9_-]/.test(char));
}
function skipHorizontalWhitespace(text, start) {
	let index = start;
	while (index < text.length && (text[index] === " " || text[index] === "	")) index += 1;
	return index;
}
function skipWhitespace(text, start) {
	let index = start;
	while (index < text.length && /\s/.test(text[index] ?? "")) index += 1;
	return index;
}
function consumeLineBreak(text, start) {
	if (text[start] === "\r") return text[start + 1] === "\n" ? start + 2 : start + 1;
	if (text[start] === "\n") return start + 1;
	return null;
}
function parseBracketOpening(text, start) {
	if (text[start] !== "[") return null;
	let cursor = start + 1;
	if (text.startsWith("tool:", cursor)) {
		cursor += 5;
		const nameStart = cursor;
		while (isToolNameChar(text[cursor])) cursor += 1;
		if (cursor === nameStart || text[cursor] !== "]") return null;
		return {
			end: cursor + 1,
			name: text.slice(nameStart, cursor),
			requiresClosing: false
		};
	}
	const nameStart = cursor;
	while (isToolNameChar(text[cursor])) cursor += 1;
	if (cursor === nameStart || text[cursor] !== "]") return null;
	const name = text.slice(nameStart, cursor);
	cursor += 1;
	cursor = skipHorizontalWhitespace(text, cursor);
	const afterLineBreak = consumeLineBreak(text, cursor);
	if (afterLineBreak === null) return null;
	return {
		end: afterLineBreak,
		name,
		requiresClosing: true
	};
}
function parseHarmonyOpening(text, start) {
	let cursor = start;
	if (text.startsWith(HARMONY_CHANNEL_MARKER, cursor)) cursor += 11;
	const channelStart = cursor;
	while (/[A-Za-z_]/.test(text[cursor] ?? "")) cursor += 1;
	const channel = text.slice(channelStart, cursor);
	if (channel !== "commentary" && channel !== "analysis" && channel !== "final") return null;
	cursor = skipHorizontalWhitespace(text, cursor);
	if (!text.startsWith("to=", cursor)) return null;
	cursor += 3;
	const nameStart = cursor;
	while (isToolNameChar(text[cursor])) cursor += 1;
	if (cursor === nameStart) return null;
	const name = text.slice(nameStart, cursor);
	cursor = skipHorizontalWhitespace(text, cursor);
	if (!text.startsWith("code", cursor)) return null;
	cursor += 4;
	cursor = skipWhitespace(text, cursor);
	if (text.startsWith(HARMONY_MESSAGE_MARKER, cursor)) cursor = skipWhitespace(text, cursor + 11);
	return {
		end: cursor,
		name,
		requiresClosing: false
	};
}
function parseXmlishFunctionOpening(text, start) {
	const match = /^<function=([A-Za-z0-9_.:-]{1,120})>\s*/i.exec(text.slice(start));
	if (!match?.[1]) return null;
	return {
		end: start + match[0].length,
		name: match[1],
		requiresClosing: false
	};
}
function parseOpening(text, start) {
	return parseBracketOpening(text, start) ?? parseHarmonyOpening(text, start);
}
function consumeJsonObject(text, start, maxPayloadBytes) {
	const cursor = skipWhitespace(text, start);
	if (text[cursor] !== "{") return null;
	let depth = 0;
	let inString = false;
	let escaped = false;
	for (let index = cursor; index < text.length; index += 1) {
		const char = text[index];
		if (index + 1 - cursor > maxPayloadBytes) return null;
		if (inString) {
			if (escaped) escaped = false;
			else if (char === "\\") escaped = true;
			else if (char === "\"") inString = false;
			continue;
		}
		if (char === "\"") {
			inString = true;
			continue;
		}
		if (char === "{") depth += 1;
		else if (char === "}") {
			depth -= 1;
			if (depth === 0) {
				const rawJson = text.slice(cursor, index + 1);
				try {
					const parsed = JSON.parse(rawJson);
					if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return null;
					return {
						end: index + 1,
						value: parsed
					};
				} catch {
					return null;
				}
			}
		}
	}
	return null;
}
function parseClosing(text, start, name) {
	const cursor = skipWhitespace(text, start);
	if (text.startsWith(END_TOOL_REQUEST, cursor)) return cursor + 18;
	const namedClosing = `[/${name}]`;
	if (text.startsWith(namedClosing, cursor)) return cursor + namedClosing.length;
	return null;
}
function parseOptionalHarmonyClosing(text, start) {
	const cursor = skipWhitespace(text, start);
	if (text.startsWith(HARMONY_CALL_MARKER, cursor)) return cursor + 8;
	return start;
}
function parsePlainTextToolCallBlockAt(text, start, options) {
	const opening = parseOpening(text, start);
	if (!opening) return null;
	const allowedToolNames = options?.allowedToolNames ? new Set(options.allowedToolNames) : void 0;
	if (allowedToolNames && !allowedToolNames.has(opening.name)) return null;
	const payload = consumeJsonObject(text, opening.end, options?.maxPayloadBytes ?? DEFAULT_MAX_PLAIN_TEXT_TOOL_PAYLOAD_BYTES);
	if (!payload) return null;
	const closingEnd = opening.requiresClosing ? parseClosing(text, payload.end, opening.name) : parseOptionalHarmonyClosing(text, payload.end);
	if (closingEnd === null) return null;
	return {
		arguments: payload.value,
		end: closingEnd,
		name: opening.name,
		raw: text.slice(start, closingEnd),
		start
	};
}
function consumeXmlishParameterBlock(text, start, maxPayloadBytes) {
	const cursor = skipWhitespace(text, start);
	const openMatch = /^<parameter=[A-Za-z0-9_.:-]{1,120}>\s*/i.exec(text.slice(cursor));
	if (!openMatch) return null;
	const payloadStart = cursor + openMatch[0].length;
	const closeStart = text.toLowerCase().indexOf(XMLISH_PARAMETER_CLOSE, payloadStart);
	if (closeStart === -1 || closeStart + 12 - cursor > maxPayloadBytes) return null;
	return closeStart + 12;
}
function consumeXmlishParameterBlocks(text, start, maxPayloadBytes) {
	let cursor = start;
	let consumed = false;
	while (true) {
		const next = consumeXmlishParameterBlock(text, cursor, maxPayloadBytes);
		if (next === null) break;
		if (next - start > maxPayloadBytes) return null;
		cursor = next;
		consumed = true;
	}
	return consumed ? cursor : null;
}
function consumeOptionalXmlishFunctionClose(text, start) {
	const cursor = skipWhitespace(text, start);
	return text.slice(cursor).toLowerCase().startsWith("</function>") ? cursor + 11 : start;
}
function parseXmlishPlainTextToolCallBlockEndAt(text, start, options) {
	const opening = parseBracketOpening(text, start) ?? parseXmlishFunctionOpening(text, start);
	if (!opening) return null;
	const allowedToolNames = options?.allowedToolNames ? new Set(options.allowedToolNames) : void 0;
	if (allowedToolNames && !allowedToolNames.has(opening.name)) return null;
	const payloadEnd = consumeXmlishParameterBlocks(text, opening.end, options?.maxPayloadBytes ?? DEFAULT_MAX_PLAIN_TEXT_TOOL_PAYLOAD_BYTES);
	if (payloadEnd === null) return null;
	return consumeOptionalXmlishFunctionClose(text, payloadEnd);
}
function parseStandalonePlainTextToolCallBlocks(text, options) {
	const blocks = [];
	let cursor = skipWhitespace(text, 0);
	while (cursor < text.length) {
		const block = parsePlainTextToolCallBlockAt(text, cursor, options);
		if (!block) return null;
		blocks.push(block);
		cursor = skipWhitespace(text, block.end);
	}
	return blocks.length > 0 ? blocks : null;
}
function stripPlainTextToolCallBlocks(text) {
	if (!text || !/\[(?:tool:)?[A-Za-z0-9_-]+\]/.test(text) && !/(?:^|\n)\s*(?:<\|channel\|>)?(?:commentary|analysis|final)\s+to=/.test(text) && !/(?:^|\n)\s*<function=[A-Za-z0-9_.:-]{1,120}>/i.test(text)) return text;
	let result = "";
	let cursor = 0;
	let index = 0;
	while (index < text.length) {
		if (!(index === 0 || text[index - 1] === "\n")) {
			index += 1;
			continue;
		}
		const blockStart = skipHorizontalWhitespace(text, index);
		const blockEnd = parsePlainTextToolCallBlockAt(text, blockStart)?.end ?? parseXmlishPlainTextToolCallBlockEndAt(text, blockStart);
		if (blockEnd === null) {
			index += 1;
			continue;
		}
		result += text.slice(cursor, index);
		cursor = blockEnd;
		const afterBlockLineBreak = consumeLineBreak(text, cursor);
		if (afterBlockLineBreak !== null) cursor = afterBlockLineBreak;
		index = cursor;
	}
	result += text.slice(cursor);
	return result;
}
//#endregion
export { parseStandalonePlainTextToolCallBlocks as n, stripPlainTextToolCallBlocks as r, extractToolPayload as t };
