// openclaw/dist/web-search-shared-DGp5_ylu.js
import { t as createWebSearchProviderContractFields } from "./provider-web-search-contract-Dc6JzeJZ.js";
//#region extensions/tavily/web-search-shared.ts
const TAVILY_CREDENTIAL_PATH = "plugins.entries.tavily.config.webSearch.apiKey";
function buildTavilyWebSearchProviderBase() {
	return {
		id: "tavily",
		label: "Tavily Search",
		hint: "Structured results with domain filters and AI answer summaries",
		onboardingScopes: ["text-inference"],
		credentialLabel: "Tavily API key",
		envVars: ["TAVILY_API_KEY"],
		placeholder: "tvly-...",
		signupUrl: "https://tavily.com/",
		docsUrl: "https://docs.openclaw.ai/tools/tavily",
		autoDetectOrder: 70,
		credentialPath: TAVILY_CREDENTIAL_PATH,
		...createWebSearchProviderContractFields({
			credentialPath: TAVILY_CREDENTIAL_PATH,
			searchCredential: {
				type: "scoped",
				scopeId: "tavily"
			},
			configuredCredential: { pluginId: "tavily" },
			selectionPluginId: "tavily"
		})
	};
}
//#endregion
export { buildTavilyWebSearchProviderBase as n, TAVILY_CREDENTIAL_PATH as t };
