// openclaw-sdk/dist/cjs/connection/policies.js
"use strict";
/**
 * Policy Manager
 *
 * Handles server policies received in hello-ok response.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyManager = exports.DEFAULT_POLICY = void 0;
exports.createPolicyManager = createPolicyManager;
/** Default policy values */
exports.DEFAULT_POLICY = {
    maxPayload: 1048576, // 1MB
    maxBufferedBytes: 65536, // 64KB
    tickIntervalMs: 30000, // 30 seconds
};
// ============================================================================
// Policy Manager
// ============================================================================
/**
 * Policy manager for storing and accessing server policies.
 *
 * Handles policies received in hello-ok response from the server.
 * Provides default values when policies are not set.
 */
class PolicyManager {
    constructor() {
        this.policy = exports.DEFAULT_POLICY;
        this.hasSetPolicy = false;
    }
    /**
     * Store policies from hello-ok.
     *
     * @param policy Policy from hello-ok response
     */
    setPolicies(policy) {
        this.policy = { ...policy };
        this.hasSetPolicy = true;
    }
    /**
     * Check if policy has been explicitly set.
     */
    hasPolicy() {
        return this.hasSetPolicy;
    }
    /**
     * Get maximum payload size.
     */
    getMaxPayload() {
        return this.policy.maxPayload;
    }
    /**
     * Get maximum buffered bytes.
     */
    getMaxBufferedBytes() {
        return this.policy.maxBufferedBytes;
    }
    /**
     * Get tick interval in milliseconds.
     */
    getTickIntervalMs() {
        return this.policy.tickIntervalMs;
    }
    /**
     * Get tick interval in seconds.
     */
    getTickIntervalSeconds() {
        return Math.floor(this.policy.tickIntervalMs / 1000);
    }
}
exports.PolicyManager = PolicyManager;
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a policy manager.
 *
 * @returns A new PolicyManager instance
 *
 * @example
 * ```ts
 * import { createPolicyManager, DEFAULT_POLICY } from './connection/policies.js';
 *
 * const policyMgr = createPolicyManager();
 * const policy = policyMgr.getPolicy();
 * console.log("Max frame size:", policy.maxFrameSize);
 * ```
 */
function createPolicyManager() {
    return new PolicyManager();
}
