// openclaw-sdk/dist/cjs/connection/protocol.js
"use strict";
/**
 * Protocol Version Negotiation
 *
 * Handles protocol version negotiation with the server.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtocolNegotiator = exports.DEFAULT_PROTOCOL_VERSION = void 0;
exports.createProtocolNegotiator = createProtocolNegotiator;
// Default protocol version
exports.DEFAULT_PROTOCOL_VERSION = {
    min: 1,
    max: 1,
};
// ============================================================================
// Protocol Negotiator
// ============================================================================
/**
 * Protocol version negotiator.
 *
 * Handles client-server protocol version negotiation:
 * - Client sends minProtocol/maxProtocol in connect
 * - Server returns negotiated protocol in hello-ok
 * - Client handles version mismatch gracefully
 */
class ProtocolNegotiator {
    constructor(range = {}) {
        this.negotiatedVersion = null;
        this.range = { ...exports.DEFAULT_PROTOCOL_VERSION, ...range };
    }
    /**
     * Get the protocol version range to send in connect frame.
     */
    getRange() {
        return { ...this.range };
    }
    /**
     * Get the negotiated protocol version.
     */
    getNegotiatedVersion() {
        return this.negotiatedVersion;
    }
    /**
     * Check if negotiation is complete.
     */
    isNegotiated() {
        return this.negotiatedVersion !== null;
    }
    /**
     * Process server response and negotiate version.
     *
     * @param helloOk Server hello-ok response
     * @returns Negotiated protocol version
     * @throws Error if version is out of supported range
     */
    negotiate(helloOk) {
        const serverVersion = helloOk.protocol;
        // Check if server version is within our supported range
        if (serverVersion < this.range.min || serverVersion > this.range.max) {
            throw new Error(`Protocol version ${serverVersion} is out of supported range [${this.range.min}, ${this.range.max}]`);
        }
        this.negotiatedVersion = serverVersion;
        return {
            version: serverVersion,
            min: this.range.min,
            max: this.range.max,
        };
    }
    /**
     * Reset negotiation state.
     */
    reset() {
        this.negotiatedVersion = null;
    }
}
exports.ProtocolNegotiator = ProtocolNegotiator;
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a protocol negotiator.
 *
 * @param range - Optional protocol version range
 * @returns A new ProtocolNegotiator instance
 *
 * @example
 * ```ts
 * import { createProtocolNegotiator } from './connection/protocol.js';
 *
 * const negotiator = createProtocolNegotiator({
 *   minProtocol: 1,
 *   maxProtocol: 2
 * });
 * ```
 */
function createProtocolNegotiator(range) {
    return new ProtocolNegotiator(range);
}
