// openclaw-sdk/dist/cjs/connection/state.js
"use strict";
/**
 * Connection State Types
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionStateMachine = void 0;
exports.createConnectionStateMachine = createConnectionStateMachine;
/** Valid state transitions map */
const VALID_TRANSITIONS = {
    disconnected: ['connecting'],
    connecting: ['handshaking', 'disconnected', 'closed'],
    handshaking: ['authenticating', 'reconnecting', 'disconnected', 'closed'],
    authenticating: ['ready', 'reconnecting', 'disconnected', 'closed'],
    ready: ['reconnecting', 'disconnected', 'closed'],
    reconnecting: ['connecting', 'handshaking', 'authenticating', 'ready', 'disconnected', 'closed'],
    closed: ['disconnected'],
};
// ============================================================================
// State Machine
// ============================================================================
/**
 * Connection state machine with valid transition enforcement.
 *
 * States:
 * - `disconnected`: Initial state, no connection
 * - `connecting`: TCP/TLS connection in progress
 * - `handshaking`: Protocol handshake in progress
 * - `authenticating`: Authentication in progress
 * - `ready`: Fully connected and authenticated
 * - `reconnecting`: Attempting to reconnect after disconnect
 * - `closed`: Connection closed (terminal state)
 */
class ConnectionStateMachine {
    constructor() {
        this.state = 'disconnected';
        this.listeners = new Set();
        this.listenerErrorHandler = null;
    }
    /**
     * Get current state.
     */
    getState() {
        return this.state;
    }
    /**
     * Check if currently connected (handshaking or later).
     */
    isConnected() {
        return (this.state === 'connecting' ||
            this.state === 'handshaking' ||
            this.state === 'authenticating' ||
            this.state === 'ready');
    }
    /**
     * Check if ready for requests.
     */
    isReady() {
        return this.state === 'ready';
    }
    /**
     * Check if can transition to a new state.
     */
    canTransitionTo(newState) {
        const valid = VALID_TRANSITIONS[this.state];
        return valid.includes(newState);
    }
    /**
     * Transition to a new state.
     *
     * @param newState Target state
     * @throws Error if transition is invalid
     */
    transitionTo(newState) {
        if (!this.canTransitionTo(newState)) {
            throw new Error(`Invalid state transition from '${this.state}' to '${newState}'`);
        }
        const previous = this.state;
        this.state = newState;
        this.emit({ previous, current: newState });
    }
    /**
     * Transition to connecting (convenience method).
     */
    connect() {
        this.transitionTo('connecting');
    }
    /**
     * Transition to handshaking (convenience method).
     */
    startHandshake() {
        this.transitionTo('handshaking');
    }
    /**
     * Transition to authenticating (convenience method).
     */
    startAuth() {
        this.transitionTo('authenticating');
    }
    /**
     * Transition to ready (convenience method).
     */
    ready() {
        this.transitionTo('ready');
    }
    /**
     * Transition to reconnecting (convenience method).
     */
    startReconnect() {
        this.transitionTo('reconnecting');
    }
    /**
     * Transition to disconnected (convenience method).
     */
    disconnect() {
        this.transitionTo('disconnected');
    }
    /**
     * Close connection (terminal state).
     */
    close() {
        this.transitionTo('closed');
    }
    /**
     * Reset to disconnected state.
     */
    reset() {
        this.state = 'disconnected';
    }
    /**
     * Add a state change listener.
     */
    onChange(listener) {
        this.listeners.add(listener);
        return () => this.listeners.delete(listener);
    }
    /**
     * Set a callback for handling errors thrown by state change listeners.
     *
     * By default, listener errors are silently ignored to prevent one buggy
     * listener from breaking the entire state machine flow. Use this method to
     * customize error handling (e.g., logging to error tracking service).
     *
     * @param handler - Callback function to handle listener errors
     */
    onListenerError(handler) {
        this.listenerErrorHandler = handler;
    }
    /**
     * Emit state change event.
     */
    emit(event) {
        for (const listener of this.listeners) {
            try {
                listener(event);
            }
            catch (error) {
                if (this.listenerErrorHandler) {
                    this.listenerErrorHandler({ error, event });
                }
                else {
                    // Default: silent to prevent cascading failures during state transitions
                    // Users can set onListenerError to capture these
                }
            }
        }
    }
}
exports.ConnectionStateMachine = ConnectionStateMachine;
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a connection state machine.
 *
 * @returns A new ConnectionStateMachine instance
 *
 * @example
 * ```ts
 * import { createConnectionStateMachine } from './connection/state.js';
 *
 * const stateMachine = createConnectionStateMachine();
 *
 * stateMachine.onStateChange((event) => {
 *   console.log(`State: ${event.previousState} -> ${event.newState}`);
 * });
 *
 * stateMachine.transition('connecting');
 * ```
 */
function createConnectionStateMachine() {
    return new ConnectionStateMachine();
}
