// openclaw-sdk/dist/cjs/connection/tls.js
"use strict";
/**
 * TLS Validator
 *
 * Validates TLS certificate fingerprints for secure connections.
 * Supports both Node.js (with socket inspection) and browser (skipped).
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TlsValidator = void 0;
exports.createTlsValidator = createTlsValidator;
const errors_js_1 = require("../errors.js");
// ============================================================================
// TLS Validator
// ============================================================================
/**
 * TLS certificate fingerprint validator.
 *
 * Validates TLS certificates using SHA-256 SPKI fingerprints.
 * Provides constant-time comparison to prevent timing attacks.
 *
 * Platform support:
 * - Node.js: Full validation with socket inspection
 * - Browser: Validation skipped (browser handles TLS)
 */
class TlsValidator {
    /**
     * Check if TLS validation is supported on this platform.
     *
     * Returns false in browser environments where TLS socket
     * inspection is not available.
     */
    static get isSupported() {
        // Check if we're in Node.js by verifying tls module exists
        try {
            require('tls');
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Create a TLS validator.
     *
     * @param config Configuration with expected fingerprints
     */
    constructor(config) {
        this.expectedFingerprints = new Set(config.expectedFingerprints.map(normalizeFingerprint));
        this.requireValidation = config.requireValidation;
    }
    /**
     * Validate TLS certificate.
     *
     * @param socket TLS socket to validate
     * @returns Validation result
     * @throws ConnectionError if fingerprint doesn't match and requireValidation is true
     */
    validate(socket) {
        // If no fingerprints configured
        if (this.expectedFingerprints.size === 0) {
            if (this.requireValidation) {
                throw new errors_js_1.ConnectionError({
                    code: 'TLS_FINGERPRINT_MISMATCH',
                    message: 'TLS validation required but no fingerprints configured',
                    retryable: false,
                });
            }
            return { valid: true };
        }
        const cert = socket.getPeerCertificate();
        const fingerprint = cert.fingerprint256 ? normalizeFingerprint(cert.fingerprint256) : undefined;
        if (!fingerprint) {
            const error = 'Certificate has no fingerprint';
            if (this.requireValidation) {
                throw new errors_js_1.ConnectionError({
                    code: 'TLS_FINGERPRINT_MISMATCH',
                    message: error,
                    retryable: false,
                });
            }
            return { valid: false, error };
        }
        // Check if fingerprint matches any expected
        const isValid = this.constantTimeCompare(fingerprint);
        const result = {
            valid: isValid,
            fingerprint: cert.fingerprint256 ?? undefined,
            subject: cert.subject ?? undefined,
            issuer: cert.issuer ?? undefined,
            validFrom: cert.valid_from ? new Date(cert.valid_from) : undefined,
            validTo: cert.valid_to ? new Date(cert.valid_to) : undefined,
        };
        if (!isValid) {
            result.error = 'Fingerprint does not match expected';
            if (this.requireValidation) {
                throw new errors_js_1.ConnectionError({
                    code: 'TLS_FINGERPRINT_MISMATCH',
                    message: `TLS fingerprint mismatch: expected ${[...this.expectedFingerprints][0]}, got ${fingerprint}`,
                    retryable: false,
                });
            }
        }
        return result;
    }
    /**
     * Add an expected fingerprint.
     *
     * @param fingerprint Fingerprint to add
     */
    addExpectedFingerprint(fingerprint) {
        this.expectedFingerprints.add(normalizeFingerprint(fingerprint));
    }
    /**
     * Clear all expected fingerprints.
     */
    clearExpectedFingerprints() {
        this.expectedFingerprints.clear();
    }
    /**
     * Constant-time comparison to prevent timing attacks.
     */
    constantTimeCompare(input) {
        for (const expected of this.expectedFingerprints) {
            if (constantTimeEqual(input, expected)) {
                return true;
            }
        }
        return false;
    }
}
exports.TlsValidator = TlsValidator;
// ============================================================================
// Helper Functions
// ============================================================================
/**
 * Normalize fingerprint format (colon-separated hex to uppercase).
 */
function normalizeFingerprint(fingerprint) {
    return fingerprint.toUpperCase().replace(/:/g, '');
}
/**
 * Constant-time string comparison.
 * Prevents timing attacks by always comparing all bytes.
 */
function constantTimeEqual(a, b) {
    if (a.length !== b.length) {
        return false;
    }
    let result = 0;
    for (let i = 0; i < a.length; i++) {
        result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }
    return result === 0;
}
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a TLS validator.
 *
 * @param config - TLS validator configuration
 * @returns A new TlsValidator instance
 *
 * @example
 * ```ts
 * import { createTlsValidator } from './connection/tls.js';
 *
 * const validator = createTlsValidator({
 *   enabled: true,
 *   expectedFingerprint: "sha256:..."
 * });
 * ```
 */
function createTlsValidator(config) {
    return new TlsValidator(config);
}
