// openclaw-sdk/dist/esm/auth/provider.js
/**
 * Credentials Provider
 *
 * Interface for secure credential access with support for multiple
 * authentication methods (token, device, password, bootstrap).
 *
 * @module
 */
/**
 * Static credentials provider for simple use cases.
 *
 * Suitable for development but not production for device keys.
 */
export class StaticCredentialsProvider {
    constructor(config) {
        this.config = config;
    }
    async getToken() {
        return this.config.token ?? null;
    }
    async refreshToken(currentToken) {
        if (!this.config.refreshToken) {
            return { token: null, success: false, errorCode: 'REFRESH_NOT_SUPPORTED' };
        }
        try {
            const newToken = await this.config.refreshToken(currentToken);
            return { token: newToken, success: true };
        }
        catch (error) {
            return {
                token: null,
                success: false,
                errorCode: 'REFRESH_FAILED',
                retryAfterMs: error?.retryAfterMs,
            };
        }
    }
    async getDeviceCredentials() {
        return this.config.device ?? null;
    }
    async getBootstrapToken() {
        return this.config.bootstrapToken ?? null;
    }
    async getPassword() {
        return this.config.password ?? null;
    }
    async signChallenge(nonce, timestamp) {
        if (!this.config.device?.privateKey) {
            throw new Error('No private key available for signing');
        }
        // Node.js crypto implementation
        const crypto = await import('crypto');
        const payload = `${nonce}:${timestamp}`;
        const sign = crypto.createSign('SHA256');
        sign.update(payload);
        sign.end();
        const signature = sign.sign(this.config.device.privateKey, 'base64');
        return signature;
    }
}
const DEFAULT_AUTH_CHAIN = ['bootstrapToken', 'deviceToken', 'token', 'password'];
/**
 * Auth handler that integrates with ConnectionManager.
 */
export class AuthHandler {
    constructor(provider, config = {}) {
        this.provider = provider;
        this.config = config;
        this.currentMethod = null;
    }
    /**
     * Prepare auth data for connect request.
     */
    async prepareAuth() {
        const methods = this.config.primaryAuth
            ? [this.config.primaryAuth]
            : (this.config.fallbackChain ?? DEFAULT_AUTH_CHAIN);
        for (const method of methods) {
            const data = await this.tryAuthMethod(method);
            if (data) {
                this.currentMethod = method;
                return { method, data };
            }
        }
        return null;
    }
    /**
     * Handle incoming challenge from server.
     */
    async handleChallenge(challenge) {
        const now = Date.now();
        // Check expiration
        if (now >= challenge.expiresAt) {
            const { AuthError } = await import('../errors.js');
            throw new AuthError({
                code: 'CHALLENGE_EXPIRED',
                message: 'Challenge expired',
                retryable: true,
            });
        }
        // Sign the challenge
        if (!this.provider.signChallenge) {
            const { AuthError } = await import('../errors.js');
            throw new AuthError({
                code: 'CHALLENGE_FAILED',
                message: 'Credentials provider does not support challenge signing',
                retryable: false,
            });
        }
        try {
            const signature = await this.provider.signChallenge(challenge.nonce, challenge.timestamp);
            return { signature, timestamp: challenge.timestamp };
        }
        catch (error) {
            const { AuthError } = await import('../errors.js');
            throw new AuthError({
                code: 'CHALLENGE_FAILED',
                message: `Failed to sign challenge: ${error instanceof Error ? error.message : 'Unknown'}`,
                retryable: false,
            });
        }
    }
    /**
     * Get the current auth method being used.
     */
    getAuthMethod() {
        return this.currentMethod;
    }
    /**
     * Refresh token if supported (for reconnection).
     */
    async refreshToken() {
        if (!this.provider.refreshToken) {
            return null;
        }
        const currentToken = this.currentMethod === 'token'
            ? ((await this.provider.getToken?.()) ?? undefined)
            : undefined;
        return this.provider.refreshToken(currentToken ?? null);
    }
    async tryAuthMethod(method) {
        switch (method) {
            case 'bootstrapToken': {
                const token = await this.provider.getBootstrapToken?.();
                return token ? { bootstrapToken: token } : null;
            }
            case 'deviceToken': {
                const device = await this.provider.getDeviceCredentials?.();
                if (device) {
                    return {
                        deviceToken: device.publicKey,
                    };
                }
                return null;
            }
            case 'token': {
                const token = await this.provider.getToken?.();
                return token ? { token } : null;
            }
            case 'password': {
                const password = await this.provider.getPassword?.();
                return password ?? null;
            }
        }
    }
}
// ============================================================================
// Factory
// ============================================================================
/**
 * Create an auth handler.
 *
 * @param provider - Credentials provider instance
 * @param config - Optional auth configuration
 * @returns AuthHandler instance
 *
 * @example
 * ```ts
 * const provider = new StaticCredentialsProvider({
 *   token: "my-token"
 * });
 *
 * const auth = createAuthHandler(provider, {
 *   primaryAuth: "token",
 *   fallbackChain: ["bootstrapToken", "password"]
 * });
 * ```
 */
export function createAuthHandler(provider, config) {
    return new AuthHandler(provider, config);
}
