// openclaw-sdk/dist/cjs/client.js
"use strict";
/**
 * OpenClaw SDK Client
 *
 * Main entry point for interacting with the OpenClaw Gateway via WebSocket.
 * This class provides a facade over the connection and request managers.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenClawClient = void 0;
exports.createClient = createClient;
const websocket_js_1 = require("./transport/websocket.js");
const connection_js_1 = require("./managers/connection.js");
const request_js_1 = require("./managers/request.js");
// ============================================================================
// Default Configuration Constants
// ============================================================================
const DEFAULT_REQUEST_TIMEOUT_MS = 30000;
const DEFAULT_CONNECT_TIMEOUT_MS = 30000;
const DEFAULT_MAX_RECONNECT_ATTEMPTS = 10;
const DEFAULT_RECONNECT_DELAY_MS = 1000;
// ============================================================================
// Main Client Class
// ============================================================================
/**
 * OpenClaw Client
 *
 * Main client class for connecting to and interacting with an OpenClaw Gateway.
 * Provides methods for connection lifecycle, request/response handling, and state queries.
 */
class OpenClawClient {
    /**
     * Create a new OpenClaw client instance.
     *
     * @param config - Client configuration
     */
    constructor(config) {
        // Event handler storage - using Set for O(1) add/remove
        this.stateChangeHandlers = new Set();
        this.errorHandlers = new Set();
        this.messageHandlers = new Set();
        this.closedHandlers = new Set();
        // Normalize connection config - supports both flat and nested style
        const normalizedConfig = this.normalizeConnectionConfig(config);
        this.config = config;
        // Create WebSocket transport
        const transport = new websocket_js_1.WebSocketTransport({
            connectTimeoutMs: normalizedConfig.connectTimeoutMs,
        });
        // Create connection manager with event handlers
        this.connectionManager = (0, connection_js_1.createConnectionManager)(transport, {
            defaultRequestTimeout: normalizedConfig.requestTimeoutMs,
            autoReconnect: normalizedConfig.autoReconnect,
            reconnectDelayMs: normalizedConfig.reconnectDelayMs,
            maxReconnectAttempts: normalizedConfig.maxReconnectAttempts,
        });
        // Set up connection manager event handlers
        this.setupConnectionHandlers();
        // Create request manager
        this.requestManager = (0, request_js_1.createRequestManager)();
    }
    /**
     * Normalize connection config - supports both flat and nested style.
     * Nested config takes precedence over flat config.
     */
    normalizeConnectionConfig(config) {
        const conn = config.connection ?? {};
        return {
            requestTimeoutMs: conn.requestTimeoutMs ?? config.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS,
            connectTimeoutMs: conn.connectTimeoutMs ?? config.connectTimeoutMs ?? DEFAULT_CONNECT_TIMEOUT_MS,
            autoReconnect: conn.autoReconnect ?? config.autoReconnect ?? true,
            maxReconnectAttempts: conn.maxReconnectAttempts ?? config.maxReconnectAttempts ?? DEFAULT_MAX_RECONNECT_ATTEMPTS,
            reconnectDelayMs: conn.reconnectDelayMs ?? config.reconnectDelayMs ?? DEFAULT_RECONNECT_DELAY_MS,
        };
    }
    /**
     * Set up connection manager event handlers.
     */
    setupConnectionHandlers() {
        const handlers = {
            onStateChange: event => {
                this.stateChangeHandlers.forEach(handler => {
                    try {
                        handler(event.newState);
                    }
                    catch (error) {
                        // Silently ignore handler errors to prevent cascading failures
                        console.error('Error in stateChange handler:', error);
                    }
                });
            },
            onError: event => {
                const error = new Error(event.message);
                this.errorHandlers.forEach(handler => {
                    try {
                        handler(error);
                    }
                    catch (err) {
                        console.error('Error in error handler:', err);
                    }
                });
            },
            onMessage: frame => {
                this.messageHandlers.forEach(handler => {
                    try {
                        handler(frame);
                    }
                    catch (error) {
                        console.error('Error in message handler:', error);
                    }
                });
                // Also handle request responses
                if (frame.type === 'res') {
                    this.requestManager.resolveRequest(frame.id, frame);
                }
            },
            onClosed: () => {
                this.closedHandlers.forEach(handler => {
                    try {
                        handler();
                    }
                    catch (error) {
                        console.error('Error in closed handler:', error);
                    }
                });
            },
        };
        this.connectionManager.setHandlers(handlers);
    }
    /**
     * Check if the client is currently connected.
     *
     * @returns True if the connection state is 'ready'
     *
     * @example
     * ```ts
     * if (client.isConnected) {
     *   console.log("Ready to send requests");
     * }
     * ```
     */
    get isConnected() {
        return this.connectionManager.getState() === 'ready';
    }
    /**
     * Get the current connection state.
     *
     * @returns The current ConnectionState ('disconnected' | 'connecting' | 'ready' | 'closing')
     *
     * @example
     * ```ts
     * console.log("State:", client.connectionState);
     * // Output: State: ready
     * ```
     */
    get connectionState() {
        return this.connectionManager.getState();
    }
    /**
     * Connect to the OpenClaw Gateway.
     *
     * @returns Promise that resolves when the connection is established
     * @throws Error if connection fails
     *
     * @example
     * ```ts
     * const client = createClient({
     *   url: "ws://localhost:8080",
     *   clientId: "my-app"
     * });
     *
     * await client.connect();
     * console.log("Connected:", client.isConnected);
     * ```
     */
    async connect() {
        // Build connection parameters
        const connectParams = {
            minProtocol: 1,
            maxProtocol: 1,
            client: {
                id: this.config.clientId,
                displayName: this.config.clientId,
                version: this.config.clientVersion ?? '1.0.0',
                platform: this.config.platform ?? 'typescript-sdk',
                deviceFamily: this.config.deviceFamily,
                modelIdentifier: this.config.modelIdentifier,
                mode: (this.config.mode ?? 'node'),
                instanceId: this.config.instanceId,
            },
            auth: this.config.auth,
            device: this.config.device,
        };
        // Connect using connection manager
        await this.connectionManager.connect(this.config.url, connectParams);
    }
    /**
     * Disconnect from the OpenClaw Gateway.
     *
     * @example
     * ```ts
     * client.disconnect();
     * console.log("Disconnected:", client.isConnected);
     * ```
     */
    disconnect() {
        this.connectionManager.disconnect();
    }
    /**
     * Send a request to the OpenClaw Gateway and wait for a response.
     *
     * @param method - The method name to invoke
     * @param params - Optional parameters for the method
     * @param options - Optional request options
     * @returns Promise that resolves with the response payload
     * @throws Error if the request fails or times out
     */
    async request(method, params, options) {
        // Check if signal is already aborted
        if (options?.signal?.aborted) {
            throw new Error('Request aborted');
        }
        // Generate a unique request ID
        const requestId = this.generateRequestId();
        // Create the request frame
        const requestFrame = {
            type: 'req',
            id: requestId,
            method,
            params,
        };
        // Determine timeout
        let timeout = this.config.requestTimeoutMs ?? 30000;
        if (options?.expectFinal && options.expectFinalTimeoutMs) {
            timeout = options.expectFinalTimeoutMs;
        }
        // Add pending request
        const responsePromise = this.requestManager.addRequest(requestId, {
            timeout,
        });
        // Set up abort handler
        let abortHandler;
        let cleanupAbortHandler;
        try {
            if (options?.signal) {
                abortHandler = () => {
                    this.requestManager.abortRequest(requestId);
                };
                options.signal.addEventListener('abort', abortHandler);
                // Store cleanup function
                cleanupAbortHandler = () => {
                    if (abortHandler && options.signal) {
                        options.signal.removeEventListener('abort', abortHandler);
                        abortHandler = undefined;
                    }
                };
                // Also clean up on abort
                options.signal.addEventListener('abort', cleanupAbortHandler, {
                    once: true,
                });
            }
            // Send the request
            this.connectionManager.send(requestFrame);
            // Wait for response
            const response = await responsePromise;
            if (!response.ok) {
                throw new Error(`Request failed: ${response.error?.code ?? 'UNKNOWN'} - ${response.error?.message ?? 'Unknown error'}`);
            }
            return response.payload;
        }
        finally {
            // Always clean up abort handler
            if (abortHandler && options?.signal) {
                options.signal.removeEventListener('abort', abortHandler);
            }
        }
    }
    /**
     * Abort a pending request by ID.
     *
     * @param requestId - The request ID to abort
     *
     * @example
     * ```ts
     * // Abort a specific request
     * client.abort("req-123-456");
     * ```
     */
    abort(requestId) {
        this.requestManager.abortRequest(requestId);
    }
    /**
     * Generate a unique request ID.
     *
     * @returns A unique request ID string
     */
    generateRequestId() {
        return `req-${Date.now()}-${Math.random().toString(36).substring(2, 15)}`;
    }
    /**
     * Register a handler for connection state changes.
     *
     * @param handler - Function to call when connection state changes
     * @returns Unsubscribe function to remove the handler
     *
     * @example
     * ```ts
     * const unsub = client.onStateChange((state) => {
     *   console.log("State changed to:", state);
     * });
     *
     * // Later: unsub();
     * ```
     */
    onStateChange(handler) {
        this.stateChangeHandlers.add(handler);
        return () => this.stateChangeHandlers.delete(handler);
    }
    /**
     * Register a handler for connection errors.
     *
     * @param handler - Function to call when an error occurs
     * @returns Unsubscribe function to remove the handler
     *
     * @example
     * ```ts
     * const unsub = client.onError((error) => {
     *   console.error("Connection error:", error.message);
     * });
     * ```
     */
    onError(handler) {
        this.errorHandlers.add(handler);
        return () => this.errorHandlers.delete(handler);
    }
    /**
     * Register a handler for incoming gateway frames.
     *
     * @param handler - Function to call when a frame is received
     * @returns Unsubscribe function to remove the handler
     *
     * @example
     * ```ts
     * const unsub = client.onMessage((frame) => {
     *   console.log("Received frame:", frame.type);
     * });
     * ```
     */
    onMessage(handler) {
        this.messageHandlers.add(handler);
        return () => this.messageHandlers.delete(handler);
    }
    /**
     * Register a handler for connection close events.
     *
     * @param handler - Function to call when the connection closes
     * @returns Unsubscribe function to remove the handler
     *
     * @example
     * ```ts
     * const unsub = client.onClosed(() => {
     *   console.log("Connection closed");
     * });
     * ```
     */
    onClosed(handler) {
        this.closedHandlers.add(handler);
        return () => this.closedHandlers.delete(handler);
    }
}
exports.OpenClawClient = OpenClawClient;
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a new OpenClaw client instance.
 *
 * @param config - Client configuration
 * @returns A new OpenClaw client instance
 *
 * @example
 * ```ts
 * const client = createClient({
 *   url: "ws://localhost:8080",
 *   clientId: "my-app",
 *   auth: { token: "my-token" }
 * });
 *
 * await client.connect();
 * const result = await client.request("ping");
 * ```
 */
function createClient(config) {
    return new OpenClawClient(config);
}
// ============================================================================
// Default Export
// ============================================================================
exports.default = OpenClawClient;
