// openclaw-sdk/dist/cjs/errors.js
"use strict";
/**
 * OpenClaw SDK Error Types
 *
 * Centralized error definitions for the SDK including authentication,
 * connection, protocol, and request errors.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReconnectError = exports.GatewayError = exports.AbortError = exports.CancelledError = exports.TimeoutError = exports.RequestError = exports.ProtocolError = exports.ConnectionError = exports.AuthError = exports.OpenClawError = void 0;
exports.createErrorFromResponse = createErrorFromResponse;
exports.isOpenClawError = isOpenClawError;
exports.isAuthError = isAuthError;
exports.isConnectionError = isConnectionError;
exports.isTimeoutError = isTimeoutError;
exports.isCancelledError = isCancelledError;
exports.isAbortError = isAbortError;
/**
 * Base SDK error class.
 *
 * @example
 * ```ts
 * try {
 *   await client.request("someMethod");
 * } catch (error) {
 *   if (error instanceof OpenClawError) {
 *     console.log(`Error ${error.code}: ${error.message}`);
 *     console.log("Retryable:", error.retryable);
 *   }
 * }
 * ```
 */
class OpenClawError extends Error {
    constructor(message, code, retryable = false, details) {
        super(message);
        this.code = code;
        this.retryable = retryable;
        this.details = details;
        this.name = this.constructor.name;
        Error.captureStackTrace(this, this.constructor);
    }
    /**
     * Convert to plain object for serialization.
     *
     * @returns Plain object representation of the error
     *
     * @example
     * ```ts
     * const error = new OpenClawError("Something went wrong", "TEST_ERROR");
     * const json = error.toJSON();
     * console.log(json);
     * // { name: 'OpenClawError', message: 'Something went wrong', code: 'TEST_ERROR', retryable: false, details: undefined }
     * ```
     */
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            code: this.code,
            retryable: this.retryable,
            details: this.details,
        };
    }
}
exports.OpenClawError = OpenClawError;
/**
 * Authentication error.
 *
 * @example
 * ```ts
 * throw new AuthError({
 *   code: 'AUTH_TOKEN_EXPIRED',
 *   message: 'Token has expired',
 *   retryable: true
 * });
 * ```
 */
class AuthError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? isAuthErrorRetryable(config.code), config.details);
        this.name = 'AuthError';
    }
}
exports.AuthError = AuthError;
/**
 * Check if an auth error code is retryable
 */
function isAuthErrorRetryable(code) {
    const retryableCodes = [
        'CHALLENGE_EXPIRED',
        'AUTH_TOKEN_EXPIRED',
        'AUTH_RATE_LIMITED',
    ];
    return retryableCodes.includes(code);
}
/**
 * Connection error.
 *
 * @example
 * ```ts
 * throw new ConnectionError({
 *   code: 'CONNECT_TIMEOUT',
 *   message: 'Connection timed out',
 *   retryable: true
 * });
 * ```
 */
class ConnectionError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? true, // Most connection errors are retryable
        config.details);
        this.name = 'ConnectionError';
    }
}
exports.ConnectionError = ConnectionError;
/**
 * Protocol error.
 *
 * @example
 * ```ts
 * throw new ProtocolError({
 *   code: 'PROTOCOL_UNSUPPORTED',
 *   message: 'Protocol version not supported'
 * });
 * ```
 */
class ProtocolError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? false, config.details);
        this.name = 'ProtocolError';
    }
}
exports.ProtocolError = ProtocolError;
/**
 * Request error.
 *
 * @example
 * ```ts
 * throw new RequestError({
 *   code: 'METHOD_NOT_FOUND',
 *   message: 'Method not found'
 * });
 * ```
 */
class RequestError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? isRequestErrorRetryable(config.code), config.details);
        this.name = 'RequestError';
    }
}
exports.RequestError = RequestError;
function isRequestErrorRetryable(code) {
    const retryableCodes = ['REQUEST_TIMEOUT', 'INTERNAL_ERROR'];
    return retryableCodes.includes(code);
}
/**
 * Timeout error (shorthand for request timeout).
 *
 * @example
 * ```ts
 * throw new TimeoutError('Request timed out after 30s');
 * ```
 */
class TimeoutError extends RequestError {
    constructor(message = 'Request timed out', details) {
        super({
            code: 'REQUEST_TIMEOUT',
            message,
            retryable: true,
            details,
        });
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/**
 * Request cancelled error.
 *
 * @example
 * ```ts
 * throw new CancelledError('Request was cancelled');
 * ```
 */
class CancelledError extends RequestError {
    constructor(message = 'Request was cancelled', details) {
        super({
            code: 'REQUEST_CANCELLED',
            message,
            retryable: false,
            details,
        });
        this.name = 'CancelledError';
    }
}
exports.CancelledError = CancelledError;
/**
 * Request aborted error (AbortController).
 *
 * @example
 * ```ts
 * throw new AbortError('Request was aborted via AbortController');
 * ```
 */
class AbortError extends RequestError {
    constructor(message = 'Request was aborted', details) {
        super({
            code: 'REQUEST_ABORTED',
            message,
            retryable: false,
            details,
        });
        this.name = 'AbortError';
    }
}
exports.AbortError = AbortError;
/**
 * Gateway business logic error.
 *
 * @example
 * ```ts
 * throw new GatewayError({
 *   code: 'AGENT_NOT_FOUND',
 *   message: 'Agent not found'
 * });
 * ```
 */
class GatewayError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? false, config.details);
        this.name = 'GatewayError';
    }
}
exports.GatewayError = GatewayError;
/**
 * Reconnection error.
 *
 * @example
 * ```ts
 * throw new ReconnectError({
 *   code: 'MAX_RECONNECT_ATTEMPTS',
 *   message: 'Max reconnection attempts exceeded',
 *   retryable: false
 * });
 * ```
 */
class ReconnectError extends OpenClawError {
    constructor(config) {
        super(config.message, config.code, config.retryable ?? false, config.details);
        this.name = 'ReconnectError';
    }
}
exports.ReconnectError = ReconnectError;
// ============================================================================
// Error Factory
// ============================================================================
/**
 * Create an error from a response frame error shape.
 *
 * @param error - The error object from the response frame
 * @returns Appropriate error instance based on error code
 *
 * @example
 * ```ts
 * const frame = await client.request("someMethod");
 * if (!frame.ok && frame.error) {
 *   const error = createErrorFromResponse(frame.error);
 *   // Throws AuthError, ConnectionError, ProtocolError, RequestError, GatewayError based on code
 * }
 * ```
 */
function createErrorFromResponse(error) {
    const code = error.code.toUpperCase();
    // Auth errors
    if (code.startsWith('AUTH_') || code.startsWith('CHALLENGE_')) {
        return new AuthError({
            code: code,
            message: error.message,
            retryable: error.retryable,
            details: error.details,
        });
    }
    // Connection errors
    if (code.startsWith('CONNECTION_') || code === 'TLS_FINGERPRINT_MISMATCH') {
        return new ConnectionError({
            code: code,
            message: error.message,
            retryable: error.retryable,
            details: error.details,
        });
    }
    // Protocol errors
    if (code.startsWith('PROTOCOL_')) {
        return new ProtocolError({
            code: code,
            message: error.message,
            retryable: error.retryable,
            details: error.details,
        });
    }
    // Request errors
    if (code === 'METHOD_NOT_FOUND' || code === 'INVALID_PARAMS' || code === 'INTERNAL_ERROR') {
        return new RequestError({
            code: code,
            message: error.message,
            retryable: error.retryable,
            details: error.details,
        });
    }
    // Gateway errors - treat unknown codes as gateway errors
    return new GatewayError({
        code: code,
        message: error.message,
        retryable: error.retryable,
        details: error.details,
    });
}
// ============================================================================
// Error Guards
// ============================================================================
/**
 * Type guard for OpenClawError
 *
 * @param error - Value to check
 * @returns True if the value is an OpenClawError
 *
 * @example
 * ```ts
 * try {
 *   await client.request("method");
 * } catch (error) {
 *   if (isOpenClawError(error)) {
 *     console.log(error.code, error.retryable);
 *   }
 * }
 * ```
 */
function isOpenClawError(error) {
    return error instanceof OpenClawError;
}
/**
 * Type guard for AuthError
 *
 * @param error - Value to check
 * @returns True if the value is an AuthError
 *
 * @example
 * ```ts
 * if (isAuthError(error)) {
 *   console.log("Auth error:", error.code);
 * }
 * ```
 */
function isAuthError(error) {
    return error instanceof AuthError;
}
/**
 * Type guard for ConnectionError
 *
 * @param error - Value to check
 * @returns True if the value is a ConnectionError
 *
 * @example
 * ```ts
 * if (isConnectionError(error)) {
 *   console.log("Connection error:", error.code);
 * }
 * ```
 */
function isConnectionError(error) {
    return error instanceof ConnectionError;
}
/**
 * Type guard for TimeoutError
 *
 * @param error - Value to check
 * @returns True if the value is a TimeoutError
 *
 * @example
 * ```ts
 * if (isTimeoutError(error)) {
 *   console.log("Request timed out");
 * }
 * ```
 */
function isTimeoutError(error) {
    return error instanceof TimeoutError;
}
/**
 * Type guard for CancelledError
 *
 * @param error - Value to check
 * @returns True if the value is a CancelledError
 *
 * @example
 * ```ts
 * if (isCancelledError(error)) {
 *   console.log("Request was cancelled");
 * }
 * ```
 */
function isCancelledError(error) {
    return error instanceof CancelledError;
}
/**
 * Type guard for AbortError
 *
 * @param error - Value to check
 * @returns True if the value is an AbortError
 *
 * @example
 * ```ts
 * if (isAbortError(error)) {
 *   console.log("Request was aborted");
 * }
 * ```
 */
function isAbortError(error) {
    return error instanceof AbortError;
}
// ============================================================================
// Default Export
// ============================================================================
exports.default = OpenClawError;
