// openclaw-sdk/dist/cjs/index.js
"use strict";
/**
 * OpenClaw SDK
 *
 * TypeScript SDK for connecting to OpenClaw Gateway via WebSocket.
 * Provides a high-level client interface with connection management,
 * request/response handling, and event emission.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TickMonitor = exports.createTlsValidator = exports.TlsValidator = exports.DEFAULT_POLICY = exports.createPolicyManager = exports.PolicyManager = exports.createConnectionStateMachine = exports.ConnectionStateMachine = exports.createProtocolNegotiator = exports.ProtocolNegotiator = exports.createReconnectManager = exports.ReconnectManager = exports.MAX_EVENT_NAME_LENGTH = exports.createEventManager = exports.EventManager = exports.createAuthHandler = exports.AuthHandler = exports.StaticCredentialsProvider = exports.isErrorResponse = exports.isSuccessfulResponse = exports.isEventFrame = exports.isResponseFrame = exports.isRequestFrame = exports.validateConnectParams = exports.validateErrorShape = exports.validateEventFrame = exports.validateResponseFrame = exports.validateRequestFrame = exports.validateFrame = exports.validateRequestId = exports.ValidationError = exports.createErrorFromResponse = exports.isAbortError = exports.isCancelledError = exports.isTimeoutError = exports.isConnectionError = exports.isAuthError = exports.isOpenClawError = exports.ReconnectError = exports.GatewayError = exports.AbortError = exports.CancelledError = exports.TimeoutError = exports.RequestError = exports.ProtocolError = exports.ConnectionError = exports.AuthError = exports.OpenClawError = exports.createClient = exports.OpenClawClient = void 0;
exports.default = exports.isLogger = exports.createTimeoutManager = exports.TimeoutManager = exports.createGapDetector = exports.GapDetector = exports.createTickMonitor = void 0;
// Main client exports
var client_js_1 = require("./client.js");
Object.defineProperty(exports, "OpenClawClient", { enumerable: true, get: function () { return client_js_1.OpenClawClient; } });
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return client_js_1.createClient; } });
// ============================================================================
// Error Types
// ============================================================================
// Error classes
var errors_js_1 = require("./errors.js");
Object.defineProperty(exports, "OpenClawError", { enumerable: true, get: function () { return errors_js_1.OpenClawError; } });
Object.defineProperty(exports, "AuthError", { enumerable: true, get: function () { return errors_js_1.AuthError; } });
Object.defineProperty(exports, "ConnectionError", { enumerable: true, get: function () { return errors_js_1.ConnectionError; } });
Object.defineProperty(exports, "ProtocolError", { enumerable: true, get: function () { return errors_js_1.ProtocolError; } });
Object.defineProperty(exports, "RequestError", { enumerable: true, get: function () { return errors_js_1.RequestError; } });
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return errors_js_1.TimeoutError; } });
Object.defineProperty(exports, "CancelledError", { enumerable: true, get: function () { return errors_js_1.CancelledError; } });
Object.defineProperty(exports, "AbortError", { enumerable: true, get: function () { return errors_js_1.AbortError; } });
Object.defineProperty(exports, "GatewayError", { enumerable: true, get: function () { return errors_js_1.GatewayError; } });
Object.defineProperty(exports, "ReconnectError", { enumerable: true, get: function () { return errors_js_1.ReconnectError; } });
// Error type guards
var errors_js_2 = require("./errors.js");
Object.defineProperty(exports, "isOpenClawError", { enumerable: true, get: function () { return errors_js_2.isOpenClawError; } });
Object.defineProperty(exports, "isAuthError", { enumerable: true, get: function () { return errors_js_2.isAuthError; } });
Object.defineProperty(exports, "isConnectionError", { enumerable: true, get: function () { return errors_js_2.isConnectionError; } });
Object.defineProperty(exports, "isTimeoutError", { enumerable: true, get: function () { return errors_js_2.isTimeoutError; } });
Object.defineProperty(exports, "isCancelledError", { enumerable: true, get: function () { return errors_js_2.isCancelledError; } });
Object.defineProperty(exports, "isAbortError", { enumerable: true, get: function () { return errors_js_2.isAbortError; } });
// Error factory
var errors_js_3 = require("./errors.js");
Object.defineProperty(exports, "createErrorFromResponse", { enumerable: true, get: function () { return errors_js_3.createErrorFromResponse; } });
// Validation functions
var validation_js_1 = require("./protocol/validation.js");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return validation_js_1.ValidationError; } });
Object.defineProperty(exports, "validateRequestId", { enumerable: true, get: function () { return validation_js_1.validateRequestId; } });
Object.defineProperty(exports, "validateFrame", { enumerable: true, get: function () { return validation_js_1.validateFrame; } });
Object.defineProperty(exports, "validateRequestFrame", { enumerable: true, get: function () { return validation_js_1.validateRequestFrame; } });
Object.defineProperty(exports, "validateResponseFrame", { enumerable: true, get: function () { return validation_js_1.validateResponseFrame; } });
Object.defineProperty(exports, "validateEventFrame", { enumerable: true, get: function () { return validation_js_1.validateEventFrame; } });
Object.defineProperty(exports, "validateErrorShape", { enumerable: true, get: function () { return validation_js_1.validateErrorShape; } });
Object.defineProperty(exports, "validateConnectParams", { enumerable: true, get: function () { return validation_js_1.validateConnectParams; } });
// Type guards
Object.defineProperty(exports, "isRequestFrame", { enumerable: true, get: function () { return validation_js_1.isRequestFrame; } });
Object.defineProperty(exports, "isResponseFrame", { enumerable: true, get: function () { return validation_js_1.isResponseFrame; } });
Object.defineProperty(exports, "isEventFrame", { enumerable: true, get: function () { return validation_js_1.isEventFrame; } });
Object.defineProperty(exports, "isSuccessfulResponse", { enumerable: true, get: function () { return validation_js_1.isSuccessfulResponse; } });
Object.defineProperty(exports, "isErrorResponse", { enumerable: true, get: function () { return validation_js_1.isErrorResponse; } });
// Authentication
var provider_js_1 = require("./auth/provider.js");
Object.defineProperty(exports, "StaticCredentialsProvider", { enumerable: true, get: function () { return provider_js_1.StaticCredentialsProvider; } });
Object.defineProperty(exports, "AuthHandler", { enumerable: true, get: function () { return provider_js_1.AuthHandler; } });
Object.defineProperty(exports, "createAuthHandler", { enumerable: true, get: function () { return provider_js_1.createAuthHandler; } });
// Event Manager
var event_js_1 = require("./managers/event.js");
Object.defineProperty(exports, "EventManager", { enumerable: true, get: function () { return event_js_1.EventManager; } });
Object.defineProperty(exports, "createEventManager", { enumerable: true, get: function () { return event_js_1.createEventManager; } });
Object.defineProperty(exports, "MAX_EVENT_NAME_LENGTH", { enumerable: true, get: function () { return event_js_1.MAX_EVENT_NAME_LENGTH; } });
// ============================================================================
// Reconnection Manager (Advanced/Stand-alone)
// ============================================================================
/**
 * Reconnection Manager - Advanced Stand-alone Reconnection Control
 *
 * **Use this when you need custom reconnection behavior** outside of
 * ConnectionManager's built-in reconnection.
 *
 * ### Decision Guide: Which Reconnection Approach?
 *
 * **Use ConnectionManager's built-in reconnection if you need:**
 * - Simple automatic reconnection on disconnect
 * - Configurable max attempts and delay
 * - Standard Fibonacci backoff
 * - Integration with connection lifecycle
 *
 * **Use ReconnectManager stand-alone if you need:**
 * - Custom reconnection logic separate from ConnectionManager
 * - Reconnection state tracking for external management
 * - Event-driven reconnection workflows
 * - Advanced retry strategies beyond standard backoff
 *
 * @example
 * ```ts
 * // Stand-alone usage for custom reconnection flow
 * const reconnectMgr = createReconnectManager({
 *   maxAttempts: 10,
 *   initialDelayMs: 1000,
 *   maxDelayMs: 30000,
 * });
 *
 * reconnectMgr.on("stateChange", (state) => {
 *   if (state.phase === "waiting") {
 *     console.log(`Reconnecting in ${state.delayMs}ms...`);
 *   }
 * });
 * ```
 */
var reconnect_js_1 = require("./managers/reconnect.js");
Object.defineProperty(exports, "ReconnectManager", { enumerable: true, get: function () { return reconnect_js_1.ReconnectManager; } });
Object.defineProperty(exports, "createReconnectManager", { enumerable: true, get: function () { return reconnect_js_1.createReconnectManager; } });
// Connection
var protocol_js_1 = require("./connection/protocol.js");
Object.defineProperty(exports, "ProtocolNegotiator", { enumerable: true, get: function () { return protocol_js_1.ProtocolNegotiator; } });
Object.defineProperty(exports, "createProtocolNegotiator", { enumerable: true, get: function () { return protocol_js_1.createProtocolNegotiator; } });
var state_js_1 = require("./connection/state.js");
Object.defineProperty(exports, "ConnectionStateMachine", { enumerable: true, get: function () { return state_js_1.ConnectionStateMachine; } });
Object.defineProperty(exports, "createConnectionStateMachine", { enumerable: true, get: function () { return state_js_1.createConnectionStateMachine; } });
var policies_js_1 = require("./connection/policies.js");
Object.defineProperty(exports, "PolicyManager", { enumerable: true, get: function () { return policies_js_1.PolicyManager; } });
Object.defineProperty(exports, "createPolicyManager", { enumerable: true, get: function () { return policies_js_1.createPolicyManager; } });
Object.defineProperty(exports, "DEFAULT_POLICY", { enumerable: true, get: function () { return policies_js_1.DEFAULT_POLICY; } });
var tls_js_1 = require("./connection/tls.js");
Object.defineProperty(exports, "TlsValidator", { enumerable: true, get: function () { return tls_js_1.TlsValidator; } });
Object.defineProperty(exports, "createTlsValidator", { enumerable: true, get: function () { return tls_js_1.createTlsValidator; } });
// Events
var tick_js_1 = require("./events/tick.js");
Object.defineProperty(exports, "TickMonitor", { enumerable: true, get: function () { return tick_js_1.TickMonitor; } });
Object.defineProperty(exports, "createTickMonitor", { enumerable: true, get: function () { return tick_js_1.createTickMonitor; } });
var gap_js_1 = require("./events/gap.js");
Object.defineProperty(exports, "GapDetector", { enumerable: true, get: function () { return gap_js_1.GapDetector; } });
Object.defineProperty(exports, "createGapDetector", { enumerable: true, get: function () { return gap_js_1.createGapDetector; } });
// ============================================================================
// Timeout Manager
// ============================================================================
var timeoutManager_js_1 = require("./utils/timeoutManager.js");
Object.defineProperty(exports, "TimeoutManager", { enumerable: true, get: function () { return timeoutManager_js_1.TimeoutManager; } });
Object.defineProperty(exports, "createTimeoutManager", { enumerable: true, get: function () { return timeoutManager_js_1.createTimeoutManager; } });
var logger_js_1 = require("./types/logger.js");
Object.defineProperty(exports, "isLogger", { enumerable: true, get: function () { return logger_js_1.isLogger; } });
// Re-export default
var client_js_2 = require("./client.js");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(client_js_2).default; } });
