// openclaw-sdk/dist/esm/types/logger.js
/**
 * Logger Types (Reserved for Future Use)
 *
 * @module types/logger
 * @beta
 *
 * This module defines logging interfaces reserved for future use.
 * Currently, the SDK uses `console.log()` directly for the few logging
 * needs it has. A full logging implementation violates YAGNI at this stage.
 *
 * **DO NOT** implement a logger class unless:
 * - There are 10+ logging calls in the codebase, OR
 * - User requirements explicitly request configurable logging
 *
 * When implementing in the future:
 * - Consider: Winston, Pino, or similar battle-tested libraries
 * - Add `logger?: Logger` to `ClientConfig`
 * - Replace `console.log()` calls with logger calls
 */
/**
 * Log level enumeration.
 *
 * Defines standard severity levels for log messages.
 *
 * @beta
 */
export var LogLevel;
(function (LogLevel) {
    /** Detailed information for debugging purposes */
    LogLevel["Debug"] = "debug";
    /** General informational messages */
    LogLevel["Info"] = "info";
    /** Warning conditions that don't prevent operation */
    LogLevel["Warn"] = "warn";
    /** Error events that might still allow the application to continue */
    LogLevel["Error"] = "error";
})(LogLevel || (LogLevel = {}));
/**
 * Type guard to check if a value implements Logger.
 *
 * @param value - Value to check
 * @returns True if value implements Logger interface
 *
 * @beta
 *
 * @example
 * ```ts
 * import { isLogger } from './types/logger.js';
 *
 * const customLogger = {
 *   name: "myLogger",
 *   level: "info",
 *   debug: () => {},
 *   info: () => {},
 *   warn: () => {},
 *   error: () => {}
 * };
 *
 * if (isLogger(customLogger)) {
 *   customLogger.info("Logger detected!");
 * }
 * ```
 */
export function isLogger(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'name' in value &&
        'level' in value &&
        'debug' in value &&
        'info' in value &&
        'warn' in value &&
        'error' in value &&
        typeof value.debug === 'function' &&
        typeof value.info === 'function' &&
        typeof value.warn === 'function' &&
        typeof value.error === 'function');
}
