// openclaw-sdk/dist/cjs/transport/browser.js
"use strict";
/**
 * Browser WebSocket Transport
 *
 * WebSocket transport implementation for browsers using the native `WebSocket` API.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserWebSocketTransport = void 0;
exports.createBrowserWebSocketTransport = createBrowserWebSocketTransport;
const websocket_js_1 = require("./websocket.js");
// ============================================================================
// Browser WebSocket Transport
// ============================================================================
/**
 * WebSocket transport implementation for browsers.
 *
 * Uses the native browser `WebSocket` API for WebSocket connectivity.
 */
class BrowserWebSocketTransport {
    /**
     * Create a browser WebSocket transport.
     *
     * @param config - Transport configuration
     */
    constructor(config) {
        this.ws = null;
        // Ready state
        this._readyState = websocket_js_1.ReadyState.CLOSED;
        this.config = config;
    }
    /**
     * Get the WebSocket URL.
     */
    get url() {
        return this.config.url;
    }
    /**
     * Get the current ready state.
     */
    get readyState() {
        return this._readyState;
    }
    /**
     * Get the ready state as a string.
     */
    get readyStateString() {
        return (0, websocket_js_1.readyStateToString)(this._readyState);
    }
    /**
     * Connect to a WebSocket server.
     *
     * @param url - The WebSocket URL to connect to
     * @returns Promise that resolves when connected
     */
    async connect(url) {
        if (this._readyState !== websocket_js_1.ReadyState.CLOSED) {
            throw new Error('Cannot connect: transport is not closed');
        }
        this._readyState = websocket_js_1.ReadyState.CONNECTING;
        return new Promise((resolve, reject) => {
            try {
                this.ws = new WebSocket(url);
                // Set up event handlers
                this.ws.onopen = () => {
                    this._readyState = websocket_js_1.ReadyState.OPEN;
                    if (this.config.onopen) {
                        this.config.onopen({ target: url });
                    }
                    resolve();
                };
                this.ws.onclose = (event) => {
                    this._readyState = websocket_js_1.ReadyState.CLOSED;
                    if (this.config.onclose) {
                        this.config.onclose({
                            code: event.code,
                            reason: event.reason,
                            wasClean: event.wasClean,
                        });
                    }
                };
                this.ws.onerror = (_event) => {
                    if (this._readyState === websocket_js_1.ReadyState.CONNECTING) {
                        reject(new Error('WebSocket connection failed'));
                    }
                    if (this.config.onerror) {
                        this.config.onerror({
                            message: 'WebSocket error',
                            error: new Error('WebSocket error'),
                        });
                    }
                };
                this.ws.onmessage = (event) => {
                    if (event.data instanceof ArrayBuffer) {
                        if (this.config.onbinary) {
                            this.config.onbinary(event.data);
                        }
                    }
                    else if (typeof event.data === 'string') {
                        if (this.config.onmessage) {
                            this.config.onmessage(event.data);
                        }
                    }
                };
            }
            catch (error) {
                reject(error);
            }
        });
    }
    /**
     * Send data through the WebSocket connection.
     *
     * @param data - Data to send (string or ArrayBuffer)
     * @throws Error if the connection is not open
     */
    send(data) {
        if (this._readyState !== websocket_js_1.ReadyState.OPEN || !this.ws) {
            throw new Error('Cannot send: transport is not open');
        }
        this.ws.send(data);
    }
    /**
     * Close the WebSocket connection.
     *
     * @param code - Optional close status code (default: 1000)
     * @param reason - Optional close reason
     */
    close(code, reason) {
        if (this.ws) {
            this.ws.close(code, reason);
            this._readyState = websocket_js_1.ReadyState.CLOSING;
        }
    }
}
exports.BrowserWebSocketTransport = BrowserWebSocketTransport;
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a browser WebSocket transport instance.
 *
 * @param config - Transport configuration
 * @returns A new transport instance
 *
 * @example
 * ```ts
 * import { createBrowserWebSocketTransport } from './transport/browser.js';
 *
 * const transport = createBrowserWebSocketTransport({
 *   url: "ws://localhost:8080"
 * });
 * ```
 */
function createBrowserWebSocketTransport(config) {
    return new BrowserWebSocketTransport(config);
}
// ============================================================================
// Re-exports
// ============================================================================
exports.default = BrowserWebSocketTransport;
