// openclaw-sdk/dist/cjs/transport/index.js
"use strict";
/**
 * Transport Module - Platform Detection and Exports
 *
 * Automatically detects the runtime environment (Node.js vs browser)
 * and exports the appropriate WebSocket transport implementation.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketTransport = void 0;
exports.createWebSocketTransport = createWebSocketTransport;
exports.default = createWebSocketTransport;
exports.createNodeWebSocketTransport = createNodeWebSocketTransport;
exports.createBrowserWebSocketTransport = createBrowserWebSocketTransport;
// ============================================================================
// Platform Detection
// ============================================================================
/**
 * Detect if running in browser environment.
 *
 * Uses runtime detection (not build-time) to determine platform.
 * This allows the same bundle to work in both Node.js and browser.
 */
const isBrowser = typeof globalThis.window !== "undefined";
/**
 * Get the current platform.
 *
 * Respects FORCE_PLATFORM environment variable for testing purposes.
 * - "node": Force Node.js transport
 * - "browser": Force browser transport
 * - undefined: Auto-detect based on runtime
 */
const platform = (process.env.FORCE_PLATFORM || (isBrowser ? "browser" : "node"));
// ============================================================================
// Factory Functions (Dynamic Import)
// ============================================================================
/**
 * Create a WebSocket transport for the current platform.
 *
 * This function automatically selects the appropriate transport implementation
 * based on the runtime environment (Node.js or browser).
 *
 * @param config - Transport configuration (platform-specific)
 * @returns Promise resolving to a new transport instance
 *
 * @example
 * ```ts
 * // Works in both Node.js and browser
 * const transport = await createWebSocketTransport({
 *   url: "wss://example.com",
 *   onopen: (event) => console.log("Connected"),
 * });
 * ```
 */
async function createWebSocketTransport(config) {
    if (platform === "node") {
        const { createNodeWebSocketTransport } = await Promise.resolve().then(() => __importStar(require("./node.js")));
        return createNodeWebSocketTransport(config);
    }
    else {
        const { createBrowserWebSocketTransport } = await Promise.resolve().then(() => __importStar(require("./browser.js")));
        return createBrowserWebSocketTransport(config);
    }
}
/**
 * Create a Node.js WebSocket transport.
 *
 * Use this function when you explicitly need the Node.js implementation.
 *
 * @param config - Node.js transport configuration
 * @returns Promise resolving to a new NodeWebSocketTransport instance
 */
async function createNodeWebSocketTransport(config) {
    const { createNodeWebSocketTransport: factory } = await Promise.resolve().then(() => __importStar(require("./node.js")));
    return factory(config);
}
/**
 * Create a browser WebSocket transport.
 *
 * Use this function when you explicitly need the browser implementation.
 *
 * @param config - Browser transport configuration
 * @returns Promise resolving to a new BrowserWebSocketTransport instance
 */
async function createBrowserWebSocketTransport(config) {
    const { createBrowserWebSocketTransport: factory } = await Promise.resolve().then(() => __importStar(require("./browser.js")));
    return factory(config);
}
// ============================================================================
// Backward Compatibility
// ============================================================================
/**
 * Re-export the generic WebSocketTransport from websocket.ts for backward compatibility.
 *
 * This allows existing code that imports from the old location to continue working.
 */
var websocket_js_1 = require("./websocket.js");
Object.defineProperty(exports, "WebSocketTransport", { enumerable: true, get: function () { return websocket_js_1.WebSocketTransport; } });
