// openclaw-sdk/dist/cjs/transport/websocket.js
"use strict";
/**
 * OpenClaw WebSocket Transport
 *
 * This module provides a WebSocket transport abstraction that works across
 * different environments (Node.js and browser) with a type-safe event emitter pattern.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketTransport = exports.ReadyState = void 0;
exports.readyStateToString = readyStateToString;
exports.createWebSocketTransport = createWebSocketTransport;
const timeoutManager_js_1 = require("../utils/timeoutManager.js");
/**
 * WebSocket ready states as defined in the WebSocket specification
 */
var ReadyState;
(function (ReadyState) {
    /** The connection is not yet open */
    ReadyState[ReadyState["CONNECTING"] = 0] = "CONNECTING";
    /** The connection is open and ready to communicate */
    ReadyState[ReadyState["OPEN"] = 1] = "OPEN";
    /** The connection is in the process of closing */
    ReadyState[ReadyState["CLOSING"] = 2] = "CLOSING";
    /** The connection is closed or couldn't be opened */
    ReadyState[ReadyState["CLOSED"] = 3] = "CLOSED";
})(ReadyState || (exports.ReadyState = ReadyState = {}));
/**
 * Convert numeric ready state to string representation
 *
 * @param state - Numeric ready state
 * @returns String representation of the ready state
 *
 * @example
 * ```ts
 * import { readyStateToString, ReadyState } from './transport/websocket.js';
 *
 * console.log(readyStateToString(ReadyState.OPEN)); // "open"
 * ```
 */
function readyStateToString(state) {
    switch (state) {
        case ReadyState.CONNECTING:
            return 'connecting';
        case ReadyState.OPEN:
            return 'open';
        case ReadyState.CLOSING:
            return 'closing';
        case ReadyState.CLOSED:
            return 'closed';
    }
}
// ============================================================================
// Base WebSocket Transport Implementation
// ============================================================================
/**
 * Base WebSocket transport implementation
 *
 * This class provides a cross-platform WebSocket transport that works in both
 * Node.js (with the 'ws' library) and browser environments (using native WebSocket).
 */
class WebSocketTransport {
    constructor(config = {}) {
        this.ws = null;
        this._url = '';
        this._readyState = ReadyState.CLOSED;
        this.timeoutManager = new timeoutManager_js_1.TimeoutManager();
        // Event handlers
        this.onopen = null;
        this.onclose = null;
        this.onerror = null;
        this.onmessage = null;
        this.onbinary = null;
        this.config = {
            connectTimeoutMs: config.connectTimeoutMs ?? 30000,
            useBinary: config.useBinary ?? false,
            WebSocketImpl: config.WebSocketImpl ?? this.getDefaultWebSocket(),
        };
    }
    /**
     * Get the default WebSocket implementation for the current environment
     */
    getDefaultWebSocket() {
        // Browser environment
        if (typeof WebSocket !== 'undefined') {
            return WebSocket;
        }
        // Node.js environment - require 'ws' dynamically
        try {
            const ws = require('ws');
            return ws.WebSocket || ws.default;
        }
        catch {
            throw new Error("WebSocket is not available. Please install the 'ws' package for Node.js environments.");
        }
    }
    /**
     * Get the current URL
     */
    get url() {
        return this._url;
    }
    /**
     * Get the current ready state
     */
    get readyState() {
        return this._readyState;
    }
    /**
     * Get the current ready state as a string
     */
    get readyStateString() {
        return readyStateToString(this._readyState);
    }
    /**
     * Connect to a WebSocket server
     *
     * @param url - The WebSocket URL to connect to
     * @returns Promise that resolves when the connection is established
     */
    async connect(url) {
        if (this._readyState === ReadyState.OPEN || this._readyState === ReadyState.CONNECTING) {
            throw new Error(`Already connected or connecting to ${this._url}`);
        }
        this._url = url;
        this._readyState = ReadyState.CONNECTING;
        return new Promise((resolve, reject) => {
            try {
                // Create WebSocket instance
                const WebSocketImpl = this.config.WebSocketImpl;
                this.ws = new WebSocketImpl(url);
                // Set up connection timeout
                this.timeoutManager.set(() => {
                    if (this._readyState === ReadyState.CONNECTING) {
                        this.cleanup();
                        const error = {
                            message: `Connection timeout after ${this.config.connectTimeoutMs}ms`,
                            recoverable: true,
                        };
                        this.handleError(error);
                        reject(new Error(error.message));
                    }
                }, this.config.connectTimeoutMs, 'ws-connect');
                // Set up event handlers
                this.ws.onopen = () => {
                    this.timeoutManager.clear('ws-connect');
                    this._readyState = ReadyState.OPEN;
                    if (this.onopen) {
                        this.onopen({
                            url: this._url,
                            timestamp: Date.now(),
                        });
                    }
                    resolve();
                };
                this.ws.onclose = (event) => {
                    this.timeoutManager.clear('ws-connect');
                    // Check if we were still connecting before updating state
                    const wasConnecting = this._readyState === ReadyState.CONNECTING;
                    this._readyState = ReadyState.CLOSED;
                    const closeEvent = {
                        code: event.code,
                        reason: event.reason,
                        wasClean: event.wasClean,
                    };
                    if (this.onclose) {
                        this.onclose(closeEvent);
                    }
                    // If we were still connecting, reject the promise
                    if (wasConnecting) {
                        reject(new Error(`Connection closed: ${event.reason} (${event.code})`));
                    }
                };
                this.ws.onerror = (event) => {
                    this.timeoutManager.clear('ws-connect');
                    const error = {
                        message: 'WebSocket error occurred',
                        original: event,
                        recoverable: true,
                    };
                    this.handleError(error);
                    // If we were connecting, reject the promise and update state
                    if (this._readyState === ReadyState.CONNECTING) {
                        this._readyState = ReadyState.CLOSED;
                        this.cleanup(); // Clean up WebSocket reference to prevent memory leak
                        reject(new Error(error.message));
                    }
                };
                this.ws.onmessage = (event) => {
                    if (typeof event.data === 'string') {
                        if (this.onmessage) {
                            this.onmessage(event.data);
                        }
                    }
                    else if (event.data instanceof ArrayBuffer || ArrayBuffer.isView(event.data)) {
                        let buffer;
                        if (event.data instanceof ArrayBuffer) {
                            buffer = event.data;
                        }
                        else {
                            // Use buffer.slice to extract the relevant portion
                            const sourceBuffer = event.data.buffer;
                            buffer = sourceBuffer.slice(event.data.byteOffset, event.data.byteOffset + event.data.byteLength);
                        }
                        if (this.onbinary) {
                            this.onbinary(buffer);
                        }
                    }
                };
            }
            catch (err) {
                this.cleanup();
                reject(new Error(`Failed to create WebSocket: ${err instanceof Error ? err.message : String(err)}`));
            }
        });
    }
    /**
     * Send data through the WebSocket connection
     *
     * @param data - Data to send (string or ArrayBuffer)
     * @throws Error if the connection is not open
     */
    send(data) {
        if (this._readyState !== ReadyState.OPEN || !this.ws) {
            throw new Error(`Cannot send data: connection is ${this.readyStateString} (expected: open)`);
        }
        try {
            this.ws.send(data);
        }
        catch (err) {
            const error = {
                message: `Failed to send data: ${err instanceof Error ? err.message : String(err)}`,
                original: err instanceof Error ? err : undefined,
                recoverable: false,
            };
            this.handleError(error);
            throw error;
        }
    }
    /**
     * Close the WebSocket connection
     *
     * @param code - Optional close status code (default: 1000)
     * @param reason - Optional close reason
     */
    close(code, reason) {
        if (this._readyState === ReadyState.CLOSED) {
            return;
        }
        this.timeoutManager.clear('ws-connect');
        if (this.ws) {
            this._readyState = ReadyState.CLOSING;
            this.ws.close(code ?? 1000, reason ?? '');
        }
    }
    /**
     * Handle errors consistently
     */
    handleError(error) {
        if (this.onerror) {
            this.onerror(error);
        }
    }
    /**
     * Clean up resources
     */
    cleanup() {
        this._readyState = ReadyState.CLOSED;
        this.ws = null;
    }
}
exports.WebSocketTransport = WebSocketTransport;
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a WebSocket transport instance
 *
 * @param config - Optional configuration
 * @returns A new WebSocket transport instance
 *
 * @example
 * ```ts
 * import { createWebSocketTransport } from './transport/websocket.js';
 *
 * const transport = createWebSocketTransport({
 *   connectTimeoutMs: 30000
 * });
 * ```
 */
function createWebSocketTransport(config) {
    return new WebSocketTransport(config);
}
// ============================================================================
// Re-exports
// ============================================================================
exports.default = WebSocketTransport;
