// openclaw-sdk/dist/esm/transport/index.js
/**
 * Transport Module - Platform Detection and Exports
 *
 * Automatically detects the runtime environment (Node.js vs browser)
 * and exports the appropriate WebSocket transport implementation.
 */
// ============================================================================
// Platform Detection
// ============================================================================
/**
 * Detect if running in browser environment.
 *
 * Uses runtime detection (not build-time) to determine platform.
 * This allows the same bundle to work in both Node.js and browser.
 */
const isBrowser = typeof globalThis.window !== "undefined";
/**
 * Get the current platform.
 *
 * Respects FORCE_PLATFORM environment variable for testing purposes.
 * - "node": Force Node.js transport
 * - "browser": Force browser transport
 * - undefined: Auto-detect based on runtime
 */
const platform = (process.env.FORCE_PLATFORM || (isBrowser ? "browser" : "node"));
// ============================================================================
// Factory Functions (Dynamic Import)
// ============================================================================
/**
 * Create a WebSocket transport for the current platform.
 *
 * This function automatically selects the appropriate transport implementation
 * based on the runtime environment (Node.js or browser).
 *
 * @param config - Transport configuration (platform-specific)
 * @returns Promise resolving to a new transport instance
 *
 * @example
 * ```ts
 * // Works in both Node.js and browser
 * const transport = await createWebSocketTransport({
 *   url: "wss://example.com",
 *   onopen: (event) => console.log("Connected"),
 * });
 * ```
 */
export async function createWebSocketTransport(config) {
    if (platform === "node") {
        const { createNodeWebSocketTransport } = await import("./node.js");
        return createNodeWebSocketTransport(config);
    }
    else {
        const { createBrowserWebSocketTransport } = await import("./browser.js");
        return createBrowserWebSocketTransport(config);
    }
}
/**
 * Create a Node.js WebSocket transport.
 *
 * Use this function when you explicitly need the Node.js implementation.
 *
 * @param config - Node.js transport configuration
 * @returns Promise resolving to a new NodeWebSocketTransport instance
 */
export async function createNodeWebSocketTransport(config) {
    const { createNodeWebSocketTransport: factory } = await import("./node.js");
    return factory(config);
}
/**
 * Create a browser WebSocket transport.
 *
 * Use this function when you explicitly need the browser implementation.
 *
 * @param config - Browser transport configuration
 * @returns Promise resolving to a new BrowserWebSocketTransport instance
 */
export async function createBrowserWebSocketTransport(config) {
    const { createBrowserWebSocketTransport: factory } = await import("./browser.js");
    return factory(config);
}
// ============================================================================
// Backward Compatibility
// ============================================================================
/**
 * Re-export the generic WebSocketTransport from websocket.ts for backward compatibility.
 *
 * This allows existing code that imports from the old location to continue working.
 */
export { WebSocketTransport } from "./websocket.js";
// ============================================================================
// Default Export
// ============================================================================
/**
 * Default export is the factory function for the current platform.
 */
export { createWebSocketTransport as default };
