// openclaw-sdk/dist/esm/transport/node.js
/**
 * Node.js WebSocket Transport
 *
 * WebSocket transport implementation for Node.js using the `ws` package.
 *
 * @module
 */
import { WebSocket as WS } from 'ws';
import { ReadyState, readyStateToString } from './websocket.js';
// ============================================================================
// Node.js WebSocket Transport
// ============================================================================
/**
 * WebSocket transport implementation for Node.js.
 *
 * Uses the `ws` package for WebSocket connectivity.
 */
export class NodeWebSocketTransport {
    /**
     * Create a Node.js WebSocket transport.
     *
     * @param config - Transport configuration
     */
    constructor(config) {
        this.ws = null;
        // Ready state
        this._readyState = ReadyState.CLOSED;
        this.config = config;
    }
    /**
     * Get the WebSocket URL.
     */
    get url() {
        return this.config.url;
    }
    /**
     * Get the current ready state.
     */
    get readyState() {
        return this._readyState;
    }
    /**
     * Get the ready state as a string.
     */
    get readyStateString() {
        return readyStateToString(this._readyState);
    }
    /**
     * Connect to a WebSocket server.
     *
     * @param url - The WebSocket URL to connect to
     * @returns Promise that resolves when connected
     */
    async connect(url) {
        if (this._readyState !== ReadyState.CLOSED) {
            throw new Error('Cannot connect: transport is not closed');
        }
        this._readyState = ReadyState.CONNECTING;
        return new Promise((resolve, reject) => {
            try {
                // Prepare WebSocket options
                const options = {};
                // Add TLS validator if provided
                if (this.config.tlsValidator) {
                    options.rejectUnauthorized = false;
                }
                this.ws = new WS(url, options);
                // Set up event handlers
                this.ws.on('open', () => {
                    this._readyState = ReadyState.OPEN;
                    if (this.config.onopen) {
                        this.config.onopen({ target: url });
                    }
                    resolve();
                });
                this.ws.on('close', (code, reason) => {
                    this._readyState = ReadyState.CLOSED;
                    if (this.config.onclose) {
                        this.config.onclose({
                            code,
                            reason: reason.toString(),
                            wasClean: code === 1000,
                        });
                    }
                });
                this.ws.on('error', (error) => {
                    if (this._readyState === ReadyState.CONNECTING) {
                        reject(error);
                    }
                    if (this.config.onerror) {
                        this.config.onerror({ message: error.message, error });
                    }
                });
                this.ws.on('message', (data, isBinary) => {
                    if (isBinary && this.config.onbinary) {
                        // Convert Buffer to ArrayBuffer
                        const arrayBuffer = data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength);
                        this.config.onbinary(arrayBuffer);
                    }
                    else if (!isBinary && this.config.onmessage) {
                        this.config.onmessage(data.toString('utf-8'));
                    }
                });
            }
            catch (error) {
                reject(error);
            }
        });
    }
    /**
     * Send data through the WebSocket connection.
     *
     * @param data - Data to send (string or ArrayBuffer)
     * @throws Error if the connection is not open
     */
    send(data) {
        if (this._readyState !== ReadyState.OPEN || !this.ws) {
            throw new Error('Cannot send: transport is not open');
        }
        if (typeof data === 'string') {
            this.ws.send(data);
        }
        else {
            // Convert ArrayBuffer to Buffer for ws package
            this.ws.send(Buffer.from(data));
        }
    }
    /**
     * Close the WebSocket connection.
     *
     * @param code - Optional close status code (default: 1000)
     * @param reason - Optional close reason
     */
    close(code, reason) {
        if (this.ws) {
            this.ws.close(code, reason);
            this._readyState = ReadyState.CLOSING;
        }
    }
}
// ============================================================================
// Factory Function
// ============================================================================
/**
 * Create a Node.js WebSocket transport instance.
 *
 * @param config - Transport configuration
 * @returns A new transport instance
 *
 * @example
 * ```ts
 * import { createNodeWebSocketTransport } from './transport/node.js';
 *
 * const transport = createNodeWebSocketTransport({
 *   url: "ws://localhost:8080"
 * });
 * ```
 */
export function createNodeWebSocketTransport(config) {
    return new NodeWebSocketTransport(config);
}
// ============================================================================
// Re-exports
// ============================================================================
export default NodeWebSocketTransport;
