// openclaw-sdk/dist/cjs/protocol/types.js
"use strict";
/**
 * OpenClaw Protocol Types
 *
 * This module defines TypeScript types for the OpenClaw WebSocket protocol.
 * These types are derived from the TypeBox schemas in the openclaw package
 * but defined as native TypeScript types for better IDE support.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameTypes = exports.ErrorCodes = void 0;
// ============================================================================
// Error Codes
// ============================================================================
exports.ErrorCodes = {
    NOT_LINKED: 'NOT_LINKED',
    NOT_PAIRED: 'NOT_PAIRED',
    AGENT_TIMEOUT: 'AGENT_TIMEOUT',
    INVALID_REQUEST: 'INVALID_REQUEST',
    UNAVAILABLE: 'UNAVAILABLE',
};
// ============================================================================
// Frame Type Constants
// ============================================================================
/** Frame type constants to avoid magic strings */
exports.FrameTypes = {
    REQUEST: 'req',
    RESPONSE: 'res',
    EVENT: 'event',
};
