// openclaw-sdk/dist/cjs/protocol/validation.js
"use strict";
/**
 * OpenClaw Protocol Validation
 *
 * This module provides validation functions for protocol frames and payloads.
 * It re-exports validators from the openclaw package and adds SDK-specific
 * validation utilities.
 *
 * @see https://github.com/openclaw/openclaw
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = void 0;
exports.validateRequestId = validateRequestId;
exports.validateFrame = validateFrame;
exports.validateRequestFrame = validateRequestFrame;
exports.validateResponseFrame = validateResponseFrame;
exports.validateEventFrame = validateEventFrame;
exports.validateErrorShape = validateErrorShape;
exports.validateConnectParams = validateConnectParams;
exports.isRequestFrame = isRequestFrame;
exports.isResponseFrame = isResponseFrame;
exports.isEventFrame = isEventFrame;
exports.isSuccessfulResponse = isSuccessfulResponse;
exports.isErrorResponse = isErrorResponse;
/**
 * Validation error thrown when frame validation fails.
 *
 * @example
 * ```ts
 * throw new ValidationError('Invalid frame type', 'INVALID_FRAME_TYPE');
 * ```
 */
class ValidationError extends Error {
    constructor(message, code = 'VALIDATION_ERROR', details) {
        super(message);
        this.name = 'ValidationError';
        this.code = code;
        this.details = details;
    }
}
exports.ValidationError = ValidationError;
/**
 * Default request ID validation options.
 */
const DEFAULT_REQUEST_ID_OPTIONS = {
    allowNumeric: true,
    minLength: 1,
    maxLength: 255,
};
/**
 * Validates a request ID format.
 *
 * Request IDs should be non-empty strings or numbers (if allowed).
 *
 * @param id - The request ID to validate
 * @param options - Validation options
 * @returns True if the request ID is valid
 *
 * @example
 * ```ts
 * validateRequestId('abc-123'); // true
 * validateRequestId(123); // true
 * validateRequestId(''); // false
 * validateRequestId(null); // false
 * ```
 */
function validateRequestId(id, options = {}) {
    const opts = { ...DEFAULT_REQUEST_ID_OPTIONS, ...options };
    if (typeof id === 'string') {
        return id.length >= opts.minLength && id.length <= opts.maxLength;
    }
    if (opts.allowNumeric && typeof id === 'number') {
        return Number.isFinite(id) && id >= 0;
    }
    return false;
}
/**
 * Validates and parses a WebSocket frame.
 *
 * This function checks that the input is a valid GatewayFrame and returns
 * the parsed frame. Throws ValidationError if the input is invalid.
 *
 * @param data - Unknown data to validate as a frame
 * @returns The validated frame
 * @throws ValidationError if the frame is invalid
 *
 * @example
 * ```ts
 * const frame = validateFrame({ type: 'req', id: '1', method: 'ping', params: {} });
 * // frame is typed as RequestFrame
 * ```
 */
function validateFrame(data) {
    if (!data || typeof data !== 'object') {
        throw new ValidationError('Invalid frame: must be an object', 'INVALID_FRAME_TYPE');
    }
    const frame = data;
    // Check type discriminator
    if (!frame.type || typeof frame.type !== 'string') {
        throw new ValidationError('Invalid frame: missing or invalid type field', 'MISSING_FRAME_TYPE');
    }
    const validTypes = ['req', 'res', 'event'];
    if (!validTypes.includes(frame.type)) {
        throw new ValidationError(`Invalid frame type: ${frame.type}. Must be one of: ${validTypes.join(', ')}`, 'INVALID_FRAME_TYPE');
    }
    // Type-specific validation (exhaustive switch)
    switch (frame.type) {
        case 'req':
            return validateRequestFrame(frame);
        case 'res':
            return validateResponseFrame(frame);
        case 'event':
            return validateEventFrame(frame);
        default: {
            // Exhaustive check - this ensures all frame types are handled
            const _exhaustive = frame.type;
            throw new ValidationError(`Unknown frame type: ${_exhaustive}`, 'UNKNOWN_FRAME_TYPE');
        }
    }
}
/**
 * Validates a request frame.
 *
 * @param frame - Partial frame to validate
 * @returns The validated request frame
 * @throws ValidationError if validation fails
 *
 * @example
 * ```ts
 * const frame = validateRequestFrame({
 *   id: "req-1",
 *   method: "ping",
 *   params: {}
 * });
 * // frame is typed as RequestFrame
 * ```
 */
function validateRequestFrame(frame) {
    if (!frame.id) {
        throw new ValidationError('Invalid request frame: missing id field', 'MISSING_REQUEST_ID');
    }
    if (!validateRequestId(frame.id)) {
        throw new ValidationError(`Invalid request frame: invalid id format`, 'INVALID_REQUEST_ID');
    }
    if (!frame.method || typeof frame.method !== 'string') {
        throw new ValidationError('Invalid request frame: missing or invalid method field', 'MISSING_METHOD');
    }
    if (frame.method.length === 0) {
        throw new ValidationError('Invalid request frame: method cannot be empty', 'INVALID_METHOD');
    }
    return {
        type: 'req',
        id: frame.id,
        method: frame.method,
        params: frame.params,
    };
}
/**
 * Validates a response frame.
 *
 * @param frame - Partial frame to validate
 * @returns The validated response frame
 * @throws ValidationError if validation fails
 *
 * @example
 * ```ts
 * const frame = validateResponseFrame({
 *   id: "req-1",
 *   ok: true,
 *   payload: { result: "pong" }
 * });
 * // frame is typed as ResponseFrame
 * ```
 */
function validateResponseFrame(frame) {
    if (!frame.id) {
        throw new ValidationError('Invalid response frame: missing id field', 'MISSING_RESPONSE_ID');
    }
    if (!validateRequestId(frame.id)) {
        throw new ValidationError('Invalid response frame: invalid id format', 'INVALID_RESPONSE_ID');
    }
    if (typeof frame.ok !== 'boolean') {
        throw new ValidationError('Invalid response frame: missing or invalid ok field', 'MISSING_OK_FIELD');
    }
    // If response is not OK, error field should be present
    if (!frame.ok && !frame.error) {
        throw new ValidationError('Invalid response frame: error field required when ok is false', 'MISSING_ERROR_FIELD');
    }
    // Validate error shape if present
    if (frame.error) {
        validateErrorShape(frame.error);
    }
    return {
        type: 'res',
        id: frame.id,
        ok: frame.ok,
        payload: frame.payload,
        error: frame.error,
    };
}
/**
 * Validates an event frame.
 *
 * @param frame - Partial frame to validate
 * @returns The validated event frame
 * @throws ValidationError if validation fails
 *
 * @example
 * ```ts
 * const frame = validateEventFrame({
 *   event: "agent.created",
 *   payload: { agentId: "agent-123" }
 * });
 * // frame is typed as EventFrame
 * ```
 */
function validateEventFrame(frame) {
    if (!frame.event || typeof frame.event !== 'string') {
        throw new ValidationError('Invalid event frame: missing or invalid event field', 'MISSING_EVENT_NAME');
    }
    if (frame.event.length === 0) {
        throw new ValidationError('Invalid event frame: event name cannot be empty', 'INVALID_EVENT_NAME');
    }
    // Validate optional sequence number
    if (frame.seq !== undefined) {
        if (typeof frame.seq !== 'number' || !Number.isFinite(frame.seq) || frame.seq < 0) {
            throw new ValidationError('Invalid event frame: seq must be a non-negative number', 'INVALID_SEQUENCE');
        }
    }
    // Validate optional state version
    if (frame.stateVersion !== undefined) {
        if (typeof frame.stateVersion !== 'object' || frame.stateVersion === null) {
            throw new ValidationError('Invalid event frame: stateVersion must be an object', 'INVALID_STATE_VERSION');
        }
        const sv = frame.stateVersion;
        if (typeof sv.presence !== 'number' || !Number.isFinite(sv.presence) || sv.presence < 0) {
            throw new ValidationError('Invalid event frame: stateVersion.presence must be a non-negative number', 'INVALID_STATE_VERSION_PRESENCE');
        }
        if (typeof sv.health !== 'number' || !Number.isFinite(sv.health) || sv.health < 0) {
            throw new ValidationError('Invalid event frame: stateVersion.health must be a non-negative number', 'INVALID_STATE_VERSION_HEALTH');
        }
    }
    return {
        type: 'event',
        event: frame.event,
        payload: frame.payload,
        seq: frame.seq,
        stateVersion: frame.stateVersion,
    };
}
/**
 * Validates an error shape object.
 *
 * @param error - Error object to validate
 * @throws ValidationError if validation fails
 */
function validateErrorShape(error) {
    if (!error || typeof error !== 'object') {
        throw new ValidationError('Invalid error: must be an object', 'INVALID_ERROR_TYPE');
    }
    const err = error;
    if (!err.code || typeof err.code !== 'string') {
        throw new ValidationError('Invalid error: missing or invalid code field', 'MISSING_ERROR_CODE');
    }
    if (!err.message || typeof err.message !== 'string') {
        throw new ValidationError('Invalid error: missing or invalid message field', 'MISSING_ERROR_MESSAGE');
    }
    if (err.retryable !== undefined && typeof err.retryable !== 'boolean') {
        throw new ValidationError('Invalid error: retryable must be a boolean', 'INVALID_RETRYABLE');
    }
    if (err.retryAfterMs !== undefined) {
        if (typeof err.retryAfterMs !== 'number' ||
            !Number.isFinite(err.retryAfterMs) ||
            err.retryAfterMs < 0) {
            throw new ValidationError('Invalid error: retryAfterMs must be a non-negative number', 'INVALID_RETRY_AFTER_MS');
        }
    }
}
/**
 * Validates connection parameters.
 *
 * @param params - Parameters to validate
 * @returns The validated connection parameters
 * @throws ValidationError if validation fails
 *
 * @example
 * ```ts
 * const params = validateConnectParams({
 *   minProtocol: 1,
 *   maxProtocol: 1,
 *   client: {
 *     id: "my-app",
 *     version: "1.0.0",
 *     platform: "typescript-sdk",
 *     mode: "node"
 *   }
 * });
 * // params is typed as ConnectParams
 * ```
 */
function validateConnectParams(params) {
    if (!params || typeof params !== 'object') {
        throw new ValidationError('Invalid connect params: must be an object', 'INVALID_CONNECT_PARAMS');
    }
    const p = params;
    // Validate protocol version range
    if (typeof p.minProtocol !== 'number' || !Number.isFinite(p.minProtocol) || p.minProtocol < 1) {
        throw new ValidationError('Invalid connect params: minProtocol must be a number >= 1', 'INVALID_MIN_PROTOCOL');
    }
    if (typeof p.maxProtocol !== 'number' || !Number.isFinite(p.maxProtocol) || p.maxProtocol < 1) {
        throw new ValidationError('Invalid connect params: maxProtocol must be a number >= 1', 'INVALID_MAX_PROTOCOL');
    }
    if (p.minProtocol > p.maxProtocol) {
        throw new ValidationError('Invalid connect params: minProtocol cannot be greater than maxProtocol', 'INVALID_PROTOCOL_RANGE');
    }
    // Validate client object
    if (!p.client || typeof p.client !== 'object') {
        throw new ValidationError('Invalid connect params: missing or invalid client field', 'MISSING_CLIENT_INFO');
    }
    const client = p.client;
    if (!client.id || typeof client.id !== 'string') {
        throw new ValidationError('Invalid connect params: client.id is required', 'MISSING_CLIENT_ID');
    }
    if (!client.version || typeof client.version !== 'string') {
        throw new ValidationError('Invalid connect params: client.version is required', 'MISSING_CLIENT_VERSION');
    }
    if (!client.platform || typeof client.platform !== 'string') {
        throw new ValidationError('Invalid connect params: client.platform is required', 'MISSING_CLIENT_PLATFORM');
    }
    if (!client.mode || typeof client.mode !== 'string') {
        throw new ValidationError('Invalid connect params: client.mode is required', 'MISSING_CLIENT_MODE');
    }
    return params;
}
/**
 * Type guard to check if a value is a RequestFrame.
 *
 * @param frame - Value to check
 * @returns True if the value is a RequestFrame
 *
 * @example
 * ```ts
 * const data = JSON.parse(someJson);
 * if (isRequestFrame(data)) {
 *   console.log("Method:", data.method); // TypeScript knows it's RequestFrame
 * }
 * ```
 */
function isRequestFrame(frame) {
    try {
        return validateRequestFrame(frame).type === 'req';
    }
    catch {
        return false;
    }
}
/**
 * Type guard to check if a value is a ResponseFrame.
 *
 * @param frame - Value to check
 * @returns True if the value is a ResponseFrame
 *
 * @example
 * ```ts
 * const data = JSON.parse(someJson);
 * if (isResponseFrame(data)) {
 *   console.log("OK:", data.ok); // TypeScript knows it's ResponseFrame
 * }
 * ```
 */
function isResponseFrame(frame) {
    try {
        return validateResponseFrame(frame).type === 'res';
    }
    catch {
        return false;
    }
}
/**
 * Type guard to check if a value is an EventFrame.
 *
 * @param frame - Value to check
 * @returns True if the value is an EventFrame
 *
 * @example
 * ```ts
 * const data = JSON.parse(someJson);
 * if (isEventFrame(data)) {
 *   console.log("Event:", data.event); // TypeScript knows it's EventFrame
 * }
 * ```
 */
function isEventFrame(frame) {
    try {
        return validateEventFrame(frame).type === 'event';
    }
    catch {
        return false;
    }
}
/**
 * Type guard to check if a frame is a successful response.
 *
 * @param frame - Value to check
 * @returns True if the value is a successful ResponseFrame
 *
 * @example
 * ```ts
 * const data = JSON.parse(someJson);
 * if (isSuccessfulResponse(data)) {
 *   console.log("Result:", data.payload); // TypeScript knows it's successful
 * }
 * ```
 */
function isSuccessfulResponse(frame) {
    return isResponseFrame(frame) && frame.ok === true;
}
/**
 * Type guard to check if a frame is an error response.
 *
 * @param frame - Value to check
 * @returns True if the value is an error ResponseFrame
 *
 * @example
 * ```ts
 * const data = JSON.parse(someJson);
 * if (isErrorResponse(data)) {
 *   console.log("Error:", data.error); // TypeScript knows it's an error
 * }
 * ```
 */
function isErrorResponse(frame) {
    return isResponseFrame(frame) && frame.ok === false;
}
